{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/ZKingQ/CS598-DLH-SP24/blob/main/DLH_Team_71_Tensorflow.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "nfyi6FdlRDyg"
      },
      "source": [
        "# CS598 Deep Learning for Healthcare - Final Project - DeepMicro: Deep Representation Learning for Disease Prediction based on Microbiome Data\n",
        "\n",
        "### Lotte Zhu, Kaiqing Zhang, Matthew Trueblood (Team ID: 71)\n",
        "#### GitHub Link: https://github.com/ZKingQ/CS598-DLH-SP24"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "MQ0sNuMePBXx"
      },
      "source": [
        "# 1. Introduction\n",
        "<!-- This is an introduction to your report, you should edit this text/mardown section to compose. In this text/markdown, you should introduce:\n",
        "\n",
        "*   Background of the problem\n",
        "  * what type of problem: disease/readmission/mortality prediction,  feature engineeing, data processing, etc\n",
        "  * what is the importance/meaning of solving the problem\n",
        "  * what is the difficulty of the problem\n",
        "  * the state of the art methods and effectiveness.\n",
        "*   Paper explanation\n",
        "  * what did the paper propose\n",
        "  * what is the innovations of the method\n",
        "  * how well the proposed method work (in its own metrics)\n",
        "  * what is the contribution to the reasearch regime (referring the Background above, how important the paper is to the problem). -->\n",
        "\n",
        "\n",
        "The expanding knowledge of microbiota uncovers its crucial role in human health [1]. It plays an important role in immune system, metabolism functions and even carcinognesis of certain cancers, hence microbiota can be used to predict various disease with emerging sequencing technologies [1,2,3,4,5,6]. However, there are three major challenges to realize the predictions in practice [7]. First, the low number of samples, together with the large number of features, leads to the curse of dimensionality. Second, there is a research gap in using strain-level profiles to classify samples into patient and healthy control groups across different diseases. Third, a rigorous validation framework is essential. Prior research has shown that tuning hyperparameters on the test set without a separate validation set may lead to an overestimation of model performance [8,9,10].\n",
        "\n",
        "The DeepMicro paper proposes the deployment of autoencoders to learn low-dimensional representations from microbiota data and then predict disease by another classification model based on the learned representations, with both trainings having thorough validation schemes [7]. The authors hypothesize that these innovations can contribute to the followings:\n",
        "\n",
        "1. The appropriate autoencoders can effectively solve the curse of dimensionality.\n",
        "\n",
        "2. They also reduce latency compared with alternative models without representation learing, in the mean time maintaining favorable training metrics.\n",
        "\n",
        "In this project, our primary objective revolves around the implementation and evaluation of various autoencoder architectures. We aim to validate the disease prediction capabilities as outlined in the paper. Furthermore, we discuss the ablations inherent in the innovation of autoencoder. The rationale for choosing this paper stems from our pursuit of knowledge in deep representation learning.\n"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 2. Github Repository\n",
        "\n",
        "Our project is published on [CS598-DLH-SP24](https://github.com/ZKingQ/CS598-DLH-SP24/tree/main)."
      ],
      "metadata": {
        "id": "2bwDxCqqbwhb"
      }
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "uygL9tTPSVHB"
      },
      "source": [
        "# 3. Scope of Reproducibility\n",
        "\n",
        "<!-- List hypotheses from the paper you will test and the corresponding experiments you will run.\n",
        "\n",
        "\n",
        "1.   Hypothesis 1: xxxxxxx\n",
        "2.   Hypothesis 2: xxxxxxx\n",
        "\n",
        "You can insert images in this notebook text, [see this link](https://stackoverflow.com/questions/50670920/how-to-insert-an-inline-image-in-google-colaboratory-from-google-drive) and example below:\n",
        "\n",
        "![sample_image.png](https://drive.google.com/uc?export=view&id=1g2efvsRJDxTxKz-OY3loMhihrEUdBxbc)\n",
        "\n",
        "\n",
        "You can also use code to display images, see the code below.\n",
        "\n",
        "The images must be saved in Google Drive first.\n",
        "\n",
        "-->\n",
        "\n",
        "Our team achieved successful replication of the model by leveraging its open source codebase, resulting in metrics that are comparable to those showcased in the DeepMicro paper. In our endeavor, we thoroughly examined and addressed the claims below put forth in the original paper.\n",
        "\n",
        "1. **Dimensionality reduction engineering with traditional statistical techniques including Principal Component Analysis (PCA) and Gaussian Random Projection (GRP)**\n",
        "\n",
        "- Principal Component Analysis (PCA) aims to capture the most significant patterns and variations in a dataset by identifying orthogonal axes, known as principal components, that maximize the data variance.\n",
        "- Gaussian Random Projection (GRP) seeks to preserve pairwise distances between data points by projecting them onto a lower-dimensional space using random projections drawn from a Gaussian distribution.\n",
        "\n",
        "2. **Innovated representation learning employing with four different autoencoders including Shallow Autoencoder (SAE), Deep Autoencoder (DAE), Variational Autoencoder (VAE), and Convolutional Autoencoder (CAE)**\n",
        "\n",
        "- Shallow Autoencoder (SAE): This is the simplest form of an autoencoder, consisting of a fully\n",
        "connected encoder layer and a decoder layer. The latent representation is obtained from the encoder\n",
        "layer, which is a lower-dimensional space compared to the original input.\n",
        "- Deep Autoencoder (DAE): In addition to the encoder and decoder layers, DAE introduces hidden\n",
        "layers between the input and latent layers and between the latent and output layers. Rectified Linear\n",
        "Unit (ReLU) activation functions are used in the hidden layers.\n",
        "- Variational Autoencoder (VAE): VAE learns probabilistic representations by approximating the\n",
        "true posterior distribution of latent embeddings. It assumes that the posterior distribution follows\n",
        "a Gaussian distribution. VAE uses an encoder network to encode the means and variances of the\n",
        "Gaussian distribution and samples the latent representation from this distribution. The decoder\n",
        "network then reconstructs the input based on the sampled latent representation.\n",
        "- Convolutional Autoencoder (CAE): Instead of fully connected layers, CAE incorporates convo-\n",
        "lutional layers, where each unit is connected to local regions of the previous layer. Convolutional layers use filters (kernels) to perform convolution operations. CAE employs convolutional transpose layers (deconvolutional layers) to make the decoder symmetric to the encoder. No pooling layers\n",
        "are used in CAE.\n",
        "\n",
        "3. **Classification learning including including Support Vector Machine (SVM), Random Forest (RF), and Multi-Layer Perceptron (MLP)**\n",
        "- Support Vector Machine (SVM) is a supervised learning algorithm that aims to find an optimal hyperplane to classify data by maximizing the margin between different classes.\n",
        "- Random Forest (RF) is an ensemble learning method that constructs a multitude of decision trees and combines their predictions to make classifications.\n",
        "- Multi-Layer Perceptron (MLP) is a type of neural network that consists of multiple layers of interconnected nodes, enabling it to learn non-linear relationships and perform classification tasks.\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "xWAHJ_1CdtaA"
      },
      "source": [
        "# 4. Methodology\n",
        "\n",
        "<!-- This methodology is the core of your project. It consists of run-able codes with necessary annotations to show the expeiment you executed for testing the hypotheses.\n",
        "\n",
        "The methodology at least contains two subsections **data** and **model** in your experiment. -->"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "zPov0WCKWbPh"
      },
      "source": [
        "## 4.1. Environment"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### 4.1.1. Python Version\n",
        "\n",
        "We use colab default python environment which is `python==3.10.12`."
      ],
      "metadata": {
        "id": "0dxJ28Zaccme"
      }
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "oTXyVLh8YFl7",
        "outputId": "9157c559-1b7c-488f-ba1c-81abe32682e2"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Python 3.10.12\n"
          ]
        }
      ],
      "source": [
        "# Python version\n",
        "!python --version"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### 4.1.2. Packages\n",
        "\n",
        "We utilize PyTorch to build our models so this project primarily depends on `sklearn` and `pytorch`."
      ],
      "metadata": {
        "id": "-IMSOXGXcYM9"
      }
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "ExecuteTime": {
          "end_time": "2024-04-29T11:48:38.502933Z",
          "start_time": "2024-04-29T11:48:33.190976Z"
        },
        "id": "Lyf5Jfp7sGWw"
      },
      "outputs": [],
      "source": [
        "# import packages needed in this project\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "import time\n",
        "import json\n",
        "import datetime\n",
        "import math\n",
        "import matplotlib\n",
        "matplotlib.use('agg')\n",
        "import matplotlib.pyplot as plt\n",
        "\n",
        "from sklearn.model_selection import train_test_split\n",
        "from sklearn.decomposition import PCA\n",
        "from sklearn.random_projection import GaussianRandomProjection\n",
        "from sklearn.model_selection import GridSearchCV\n",
        "from sklearn.svm import SVC\n",
        "from sklearn.ensemble import RandomForestClassifier\n",
        "from sklearn.model_selection import StratifiedKFold\n",
        "from sklearn.metrics import roc_auc_score\n",
        "from sklearn.metrics import accuracy_score\n",
        "from sklearn.metrics import precision_score\n",
        "from sklearn.metrics import recall_score\n",
        "from sklearn.metrics import f1_score\n",
        "\n",
        "# importing PyTorch\n",
        "import torch\n",
        "import torch.nn as nn\n",
        "import torch.nn.functional as F\n",
        "import torchvision"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "2NbPHUTMbkD3"
      },
      "source": [
        "##  4.2. Data\n",
        "<!-- Data includes raw data (MIMIC III tables), descriptive statistics (our homework questions), and data processing (feature engineering).\n",
        "  * Source of the data: where the data is collected from; if data is synthetic or self-generated, explain how. If possible, please provide a link to the raw datasets.\n",
        "  * Statistics: include basic descriptive statistics of the dataset like size, cross validation split, label distribution, etc.\n",
        "  * Data process: how do you munipulate the data, e.g., change the class labels, split the dataset to train/valid/test, refining the dataset.\n",
        "  * Illustration: printing results, plotting figures for illustration.\n",
        "  * You can upload your raw dataset to Google Drive and mount this Colab to the same directory. If your raw dataset is too large, you can upload the processed dataset and have a code to load the processed dataset. -->"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "SRRY_dsgYL2Q"
      },
      "source": [
        "### 4.2.1 Data Download Instruction\n",
        "\n",
        "We store two datasets, abundance and marker from the [DeepMicro codebase](https://github.com/minoh0201/DeepMicro/tree/master/data) in the path `raw_data_dir` of the [colab project](https://drive.google.com/file/d/1c6ZPVp346e1wQH6opg4YwlCRjRgfTXCk/view?usp=drive_link).\n",
        "\n",
        "When we run this project locally, we use the function `fetch_data_from_remote` to download the datasets from the google drive to current working directory."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {
        "ExecuteTime": {
          "end_time": "2024-04-29T11:48:47.979130Z",
          "start_time": "2024-04-29T11:48:47.975250Z"
        },
        "id": "yGCbNWqfcRcs",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "8195b56b-26b9-4196-edea-66d9f93c71b4"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Data fetched and unzipped successfully\n",
            "abundance_Cirrhosis.txt   abundance_IBD.txt\t abundance_T2D.txt\n",
            "abundance_Colorectal.txt  abundance_Obesity.txt  abundance_WT2D.txt\n",
            "marker_Cirrhosis.txt   marker_IBD.txt\t   marker_T2D.txt\n",
            "marker_Colorectal.txt  marker_Obesity.txt  marker_WT2D.txt\n"
          ]
        }
      ],
      "source": [
        "# data dir\n",
        "raw_data_dir = 'data/'\n",
        "\n",
        "# Load data\n",
        "import os\n",
        "import requests\n",
        "from zipfile import ZipFile\n",
        "from io import BytesIO\n",
        "\n",
        "def fetch_data_from_remote():\n",
        "    # cloud location where the .zip file of the data used by this project is stored\n",
        "    url = 'https://drive.google.com/file/d/1c6ZPVp346e1wQH6opg4YwlCRjRgfTXCk/view?usp=drive_link'\n",
        "\n",
        "    file_id = url.split('/')[-2]\n",
        "    dwn_url = 'https://drive.google.com/uc?export=download&id=' + file_id\n",
        "\n",
        "    # unzip the data into the local working directory\n",
        "    response = requests.get(dwn_url)\n",
        "    zip_content = BytesIO(response.content)\n",
        "    with ZipFile(zip_content, 'r') as main_zip:\n",
        "        main_zip.extractall()\n",
        "        with main_zip.open('data/abundance.zip') as nested_zip_abundance:\n",
        "            with ZipFile(nested_zip_abundance) as zip_ref_abundance:\n",
        "                zip_ref_abundance.extractall('data/abundance')\n",
        "\n",
        "        with main_zip.open('data/marker.zip') as nested_zip_marker:\n",
        "            with ZipFile(nested_zip_marker) as zip_ref_marker:\n",
        "                zip_ref_marker.extractall('data/marker')\n",
        "\n",
        "if os.path.exists('data'):\n",
        "    print(\"Data already exists locally\")\n",
        "else:\n",
        "    fetch_data_from_remote()\n",
        "    print(\"Data fetched and unzipped successfully\")\n",
        "\n",
        "# show that the data is available locally now\n",
        "!ls data/abundance/\n",
        "!ls data/marker/"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "FJNbhZZ7cTjS"
      },
      "source": [
        "### 4.2.2. Data Description\n",
        "\n",
        "Our reproductivity utilizes the same datasets as the original paper, which include six disease (Table 1). They are inflammatory bowel disease (IBD), type 2 diabetes in European women (EW-T2D), type 2 diabetes in Chinese (C-T2D), obesity (Obesity), liver cirrhosis (Cirrhosis), and colorectal cancer (Colorectal).\n",
        "\n",
        "![](assets/Data_Table1.png)\n",
        "\n",
        "In each dataset, marker profile and abundance profile of microbiome are used to train our models (Table 2).\n",
        "\n",
        "![](assets/Data_Table2.png)\n",
        "\n",
        "All the data are stored in txt format and the data path structure is as following.\n",
        "\n",
        "```\n",
        ".\n",
        "├── ...\n",
        "├── data                              # Data folder\n",
        "│   ├── marker                        # Marker profile data\n",
        "│       ├── marker_IBD.txt            # Inflammatory bowel disease (IBD)\n",
        "│       ├── marker_WT2D.txt           # Type 2 diabetes in European women (EW-T2D)\n",
        "│       ├── marker_T2D.txt            # Type 2 diabetes in Chinese (C-T2D)\n",
        "│       ├── marker_Obesity.txt        # Obesity (Obesity)\n",
        "│       ├── marker_Cirrhosis.txt      # Liver cirrhosis (Cirrhosis)\n",
        "│       ├── marker_Colorectal.txt     # Colorectal cancer (Colorectal)\n",
        "│   ├── abundance                     # Abundance profile data\n",
        "│       ├── abundance_IBD.txt         # Inflammatory bowel disease (IBD)\n",
        "│       ├── abundance_WT2D.txt        # Type 2 diabetes in European women (EW-T2D)\n",
        "│       ├── abundance_T2D.txt         # Type 2 diabetes in Chinese (C-T2D)\n",
        "│       ├── abundance_Obesity.txt     # Obesity (Obesity)\n",
        "│       ├── abundance_Cirrhosis.txt   # Liver cirrhosis (Cirrhosis)\n",
        "│       ├── abundance_Colorectal.txt  # Colorectal cancer (Colorectal)\n",
        "└── ...\n",
        "```\n",
        "In each txt file, it has different number of features and data points as shown in Table 1 and Table 2 above. In the sample demonstration and visualizations below. The colorectal cancer of microbiome abundance profile has 503 features (exluding those dummy labels) and 121 entries.\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {
        "ExecuteTime": {
          "end_time": "2024-04-29T11:51:20.979319Z",
          "start_time": "2024-04-29T11:51:20.828998Z"
        },
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 586
        },
        "id": "BYGSgWXzjdeX",
        "outputId": "15c00de3-bd3f-4875-9eff-74e85e55b89d"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "0                     dataset_name            sampleID subjectID bodysite  \\\n",
              "1   Zeller_fecal_colorectal_cancer  CCIS00146684ST-4-0    fr-726    stool   \n",
              "2   Zeller_fecal_colorectal_cancer  CCIS00281083ST-3-0    fr-060    stool   \n",
              "3   Zeller_fecal_colorectal_cancer  CCIS02124300ST-4-0    fr-568    stool   \n",
              "4   Zeller_fecal_colorectal_cancer  CCIS02379307ST-4-0    fr-828    stool   \n",
              "5   Zeller_fecal_colorectal_cancer  CCIS03473770ST-4-0    fr-192    stool   \n",
              "6   Zeller_fecal_colorectal_cancer  CCIS06260551ST-3-0    fr-200    stool   \n",
              "7   Zeller_fecal_colorectal_cancer  CCIS07539127ST-4-0    fr-460    stool   \n",
              "8   Zeller_fecal_colorectal_cancer  CCIS07648107ST-4-0    fr-053    stool   \n",
              "9   Zeller_fecal_colorectal_cancer  CCIS08668806ST-3-0    fr-214    stool   \n",
              "10  Zeller_fecal_colorectal_cancer  CCIS09568613ST-4-0    fr-400    stool   \n",
              "\n",
              "0         disease age  gender country sequencing_technology  pubmedid  ...  \\\n",
              "1               n  72  female  france              Illumina  25432777  ...   \n",
              "2               n  53    male  france              Illumina  25432777  ...   \n",
              "3               n  35    male  france              Illumina  25432777  ...   \n",
              "4          cancer  67    male  france              Illumina  25432777  ...   \n",
              "5               n  29    male  france              Illumina  25432777  ...   \n",
              "6          cancer  58    male  france              Illumina  25432777  ...   \n",
              "7               n  77  female  france              Illumina  25432777  ...   \n",
              "8               n  62  female  france              Illumina  25432777  ...   \n",
              "9   small_adenoma  63    male  france              Illumina  25432777  ...   \n",
              "10              n  67    male  france              Illumina  25432777  ...   \n",
              "\n",
              "0  k__Eukaryota|p__Ascomycota|c__Saccharomycetes|o__Saccharomycetales|f__Saccharomycetaceae|g__Eremothecium|s__Eremothecium_unclassified  \\\n",
              "1                                                   0                                                                                      \n",
              "2                                                   0                                                                                      \n",
              "3                                                   0                                                                                      \n",
              "4                                                   0                                                                                      \n",
              "5                                                   0                                                                                      \n",
              "6                                                   0                                                                                      \n",
              "7                                                   0                                                                                      \n",
              "8                                                   0                                                                                      \n",
              "9                                                   0                                                                                      \n",
              "10                                                  0                                                                                      \n",
              "\n",
              "0  k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Lactobacillaceae|g__Lactobacillus|s__Lactobacillus_antri  \\\n",
              "1                                                   0                                                                    \n",
              "2                                                   0                                                                    \n",
              "3                                                   0                                                                    \n",
              "4                                                   0                                                                    \n",
              "5                                                   0                                                                    \n",
              "6                                                   0                                                                    \n",
              "7                                                   0                                                                    \n",
              "8                                                   0                                                                    \n",
              "9                                                   0                                                                    \n",
              "10                                                  0                                                                    \n",
              "\n",
              "0  k__Bacteria|p__Firmicutes|c__Bacilli|o__Bacillales|f__Bacillaceae|g__Lysinibacillus|s__Lysinibacillus_fusiformis  \\\n",
              "1                                                   0                                                                 \n",
              "2                                                   0                                                                 \n",
              "3                                                   0                                                                 \n",
              "4                                                   0                                                                 \n",
              "5                                                   0                                                                 \n",
              "6                                             0.31121                                                                 \n",
              "7                                                   0                                                                 \n",
              "8                                                   0                                                                 \n",
              "9                                                   0                                                                 \n",
              "10                                                  0                                                                 \n",
              "\n",
              "0  k__Archaea|p__Euryarchaeota|c__Methanobacteria|o__Methanobacteriales|f__Methanobacteriaceae|g__Methanobacterium|s__Methanobacterium_unclassified  \\\n",
              "1                                                   0                                                                                                 \n",
              "2                                                   0                                                                                                 \n",
              "3                                                   0                                                                                                 \n",
              "4                                                   0                                                                                                 \n",
              "5                                                   0                                                                                                 \n",
              "6                                                   0                                                                                                 \n",
              "7                                                   0                                                                                                 \n",
              "8                                                   0                                                                                                 \n",
              "9                                                   0                                                                                                 \n",
              "10                                                  0                                                                                                 \n",
              "\n",
              "0  k__Bacteria|p__Firmicutes|c__Bacilli|o__Bacillales|f__Bacillaceae|g__Lysinibacillus|s__Lysinibacillus_boronitolerans  \\\n",
              "1                                                   0                                                                     \n",
              "2                                                   0                                                                     \n",
              "3                                                   0                                                                     \n",
              "4                                                   0                                                                     \n",
              "5                                                   0                                                                     \n",
              "6                                             0.03562                                                                     \n",
              "7                                                   0                                                                     \n",
              "8                                                   0                                                                     \n",
              "9                                                   0                                                                     \n",
              "10                                                  0                                                                     \n",
              "\n",
              "0  k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Enterococcaceae|g__Bavariicoccus|s__Bavariicoccus_seileri  \\\n",
              "1                                                   0                                                                     \n",
              "2                                                   0                                                                     \n",
              "3                                                   0                                                                     \n",
              "4                                                   0                                                                     \n",
              "5                                                   0                                                                     \n",
              "6                                                   0                                                                     \n",
              "7                                                   0                                                                     \n",
              "8                                                   0                                                                     \n",
              "9                                                   0                                                                     \n",
              "10                                                  0                                                                     \n",
              "\n",
              "0  k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Enterococcaceae|g__Enterococcus|s__Enterococcus_gilvus  \\\n",
              "1                                                   0                                                                  \n",
              "2                                                   0                                                                  \n",
              "3                                                   0                                                                  \n",
              "4                                                   0                                                                  \n",
              "5                                             0.03756                                                                  \n",
              "6                                                   0                                                                  \n",
              "7                                                   0                                                                  \n",
              "8                                                   0                                                                  \n",
              "9                                                   0                                                                  \n",
              "10                                                  0                                                                  \n",
              "\n",
              "0  k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Lactobacillaceae|g__Lactobacillus|s__Lactobacillus_otakiensis  \\\n",
              "1                                                   0                                                                         \n",
              "2                                                   0                                                                         \n",
              "3                                                   0                                                                         \n",
              "4                                                   0                                                                         \n",
              "5                                                   0                                                                         \n",
              "6                                                   0                                                                         \n",
              "7                                                   0                                                                         \n",
              "8                                                   0                                                                         \n",
              "9                                                   0                                                                         \n",
              "10                                                  0                                                                         \n",
              "\n",
              "0  k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Peptococcaceae|g__Desulfotomaculum|s__Desulfotomaculum_ruminis  \\\n",
              "1                                                   0                                                                           \n",
              "2                                                   0                                                                           \n",
              "3                                                   0                                                                           \n",
              "4                                                   0                                                                           \n",
              "5                                                   0                                                                           \n",
              "6                                                   0                                                                           \n",
              "7                                                   0                                                                           \n",
              "8                                                   0                                                                           \n",
              "9                                                   0                                                                           \n",
              "10                                                  0                                                                           \n",
              "\n",
              "0  k__Bacteria|p__Firmicutes|c__Negativicutes|o__Selenomonadales|f__Veillonellaceae|g__Megasphaera|s__Megasphaera_sp_BV3C16_1  \n",
              "1                                                   0                                                                          \n",
              "2                                                   0                                                                          \n",
              "3                                                   0                                                                          \n",
              "4                                                   0                                                                          \n",
              "5                                                   0                                                                          \n",
              "6                                                   0                                                                          \n",
              "7                                                   0                                                                          \n",
              "8                                                   0                                                                          \n",
              "9                                                   0                                                                          \n",
              "10                                                  0                                                                          \n",
              "\n",
              "[10 rows x 714 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-906e9047-f6ed-4d91-a8b9-d88c6aecf1ff\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>dataset_name</th>\n",
              "      <th>sampleID</th>\n",
              "      <th>subjectID</th>\n",
              "      <th>bodysite</th>\n",
              "      <th>disease</th>\n",
              "      <th>age</th>\n",
              "      <th>gender</th>\n",
              "      <th>country</th>\n",
              "      <th>sequencing_technology</th>\n",
              "      <th>pubmedid</th>\n",
              "      <th>...</th>\n",
              "      <th>k__Eukaryota|p__Ascomycota|c__Saccharomycetes|o__Saccharomycetales|f__Saccharomycetaceae|g__Eremothecium|s__Eremothecium_unclassified</th>\n",
              "      <th>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Lactobacillaceae|g__Lactobacillus|s__Lactobacillus_antri</th>\n",
              "      <th>k__Bacteria|p__Firmicutes|c__Bacilli|o__Bacillales|f__Bacillaceae|g__Lysinibacillus|s__Lysinibacillus_fusiformis</th>\n",
              "      <th>k__Archaea|p__Euryarchaeota|c__Methanobacteria|o__Methanobacteriales|f__Methanobacteriaceae|g__Methanobacterium|s__Methanobacterium_unclassified</th>\n",
              "      <th>k__Bacteria|p__Firmicutes|c__Bacilli|o__Bacillales|f__Bacillaceae|g__Lysinibacillus|s__Lysinibacillus_boronitolerans</th>\n",
              "      <th>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Enterococcaceae|g__Bavariicoccus|s__Bavariicoccus_seileri</th>\n",
              "      <th>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Enterococcaceae|g__Enterococcus|s__Enterococcus_gilvus</th>\n",
              "      <th>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Lactobacillaceae|g__Lactobacillus|s__Lactobacillus_otakiensis</th>\n",
              "      <th>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Peptococcaceae|g__Desulfotomaculum|s__Desulfotomaculum_ruminis</th>\n",
              "      <th>k__Bacteria|p__Firmicutes|c__Negativicutes|o__Selenomonadales|f__Veillonellaceae|g__Megasphaera|s__Megasphaera_sp_BV3C16_1</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Zeller_fecal_colorectal_cancer</td>\n",
              "      <td>CCIS00146684ST-4-0</td>\n",
              "      <td>fr-726</td>\n",
              "      <td>stool</td>\n",
              "      <td>n</td>\n",
              "      <td>72</td>\n",
              "      <td>female</td>\n",
              "      <td>france</td>\n",
              "      <td>Illumina</td>\n",
              "      <td>25432777</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>Zeller_fecal_colorectal_cancer</td>\n",
              "      <td>CCIS00281083ST-3-0</td>\n",
              "      <td>fr-060</td>\n",
              "      <td>stool</td>\n",
              "      <td>n</td>\n",
              "      <td>53</td>\n",
              "      <td>male</td>\n",
              "      <td>france</td>\n",
              "      <td>Illumina</td>\n",
              "      <td>25432777</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>Zeller_fecal_colorectal_cancer</td>\n",
              "      <td>CCIS02124300ST-4-0</td>\n",
              "      <td>fr-568</td>\n",
              "      <td>stool</td>\n",
              "      <td>n</td>\n",
              "      <td>35</td>\n",
              "      <td>male</td>\n",
              "      <td>france</td>\n",
              "      <td>Illumina</td>\n",
              "      <td>25432777</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>Zeller_fecal_colorectal_cancer</td>\n",
              "      <td>CCIS02379307ST-4-0</td>\n",
              "      <td>fr-828</td>\n",
              "      <td>stool</td>\n",
              "      <td>cancer</td>\n",
              "      <td>67</td>\n",
              "      <td>male</td>\n",
              "      <td>france</td>\n",
              "      <td>Illumina</td>\n",
              "      <td>25432777</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>Zeller_fecal_colorectal_cancer</td>\n",
              "      <td>CCIS03473770ST-4-0</td>\n",
              "      <td>fr-192</td>\n",
              "      <td>stool</td>\n",
              "      <td>n</td>\n",
              "      <td>29</td>\n",
              "      <td>male</td>\n",
              "      <td>france</td>\n",
              "      <td>Illumina</td>\n",
              "      <td>25432777</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0.03756</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>Zeller_fecal_colorectal_cancer</td>\n",
              "      <td>CCIS06260551ST-3-0</td>\n",
              "      <td>fr-200</td>\n",
              "      <td>stool</td>\n",
              "      <td>cancer</td>\n",
              "      <td>58</td>\n",
              "      <td>male</td>\n",
              "      <td>france</td>\n",
              "      <td>Illumina</td>\n",
              "      <td>25432777</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0.31121</td>\n",
              "      <td>0</td>\n",
              "      <td>0.03562</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>Zeller_fecal_colorectal_cancer</td>\n",
              "      <td>CCIS07539127ST-4-0</td>\n",
              "      <td>fr-460</td>\n",
              "      <td>stool</td>\n",
              "      <td>n</td>\n",
              "      <td>77</td>\n",
              "      <td>female</td>\n",
              "      <td>france</td>\n",
              "      <td>Illumina</td>\n",
              "      <td>25432777</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>Zeller_fecal_colorectal_cancer</td>\n",
              "      <td>CCIS07648107ST-4-0</td>\n",
              "      <td>fr-053</td>\n",
              "      <td>stool</td>\n",
              "      <td>n</td>\n",
              "      <td>62</td>\n",
              "      <td>female</td>\n",
              "      <td>france</td>\n",
              "      <td>Illumina</td>\n",
              "      <td>25432777</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>Zeller_fecal_colorectal_cancer</td>\n",
              "      <td>CCIS08668806ST-3-0</td>\n",
              "      <td>fr-214</td>\n",
              "      <td>stool</td>\n",
              "      <td>small_adenoma</td>\n",
              "      <td>63</td>\n",
              "      <td>male</td>\n",
              "      <td>france</td>\n",
              "      <td>Illumina</td>\n",
              "      <td>25432777</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>Zeller_fecal_colorectal_cancer</td>\n",
              "      <td>CCIS09568613ST-4-0</td>\n",
              "      <td>fr-400</td>\n",
              "      <td>stool</td>\n",
              "      <td>n</td>\n",
              "      <td>67</td>\n",
              "      <td>male</td>\n",
              "      <td>france</td>\n",
              "      <td>Illumina</td>\n",
              "      <td>25432777</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>10 rows × 714 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-906e9047-f6ed-4d91-a8b9-d88c6aecf1ff')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-906e9047-f6ed-4d91-a8b9-d88c6aecf1ff button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-906e9047-f6ed-4d91-a8b9-d88c6aecf1ff');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-093e2865-3641-4f3d-93f6-da8829dbee82\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-093e2865-3641-4f3d-93f6-da8829dbee82')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-093e2865-3641-4f3d-93f6-da8829dbee82 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "sample"
            }
          },
          "metadata": {},
          "execution_count": 4
        }
      ],
      "source": [
        "# Extract a sample data of colorectal cancer of microbiome abundance profile\n",
        "sample = pd.read_csv(os.path.join(raw_data_dir, \"abundance\", \"abundance_Colorectal.txt\"), sep='\\t', index_col=0, header=None)\n",
        "sample = sample.T\n",
        "sample.head(10)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "5EjqDkrWm0lc",
        "outputId": "15340101-fc13-4b6e-a015-d39312ebdff0"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "503"
            ]
          },
          "metadata": {},
          "execution_count": 5
        }
      ],
      "source": [
        "# Count the features excluding patients information\n",
        "len([ _ for _ in sample.columns if '|' in _ ])"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "i93VorpFlPnl",
        "outputId": "ed802f7e-6598-4294-95c9-5e25a3da9437"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "(121, 714)"
            ]
          },
          "metadata": {},
          "execution_count": 6
        }
      ],
      "source": [
        "# Calculate the dimensions of the sameple dataset\n",
        "sample.shape"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 475
        },
        "id": "PPFi9QXolIoa",
        "outputId": "de7698a4-45fe-4468-cf36-d6ed72ed1b02"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                                                   count unique  \\\n",
              "0                                                                 \n",
              "dataset_name                                         121      1   \n",
              "sampleID                                             121    121   \n",
              "subjectID                                            121    121   \n",
              "bodysite                                             121      1   \n",
              "disease                                              121      3   \n",
              "...                                                  ...    ...   \n",
              "k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactoba...   121      2   \n",
              "k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactoba...   121      5   \n",
              "k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactoba...   121      2   \n",
              "k__Bacteria|p__Firmicutes|c__Clostridia|o__Clos...   121      2   \n",
              "k__Bacteria|p__Firmicutes|c__Negativicutes|o__S...   121      2   \n",
              "\n",
              "                                                                               top  \\\n",
              "0                                                                                    \n",
              "dataset_name                                        Zeller_fecal_colorectal_cancer   \n",
              "sampleID                                                        CCIS98832363ST-4-0   \n",
              "subjectID                                                                   fr-552   \n",
              "bodysite                                                                     stool   \n",
              "disease                                                                     cancer   \n",
              "...                                                                            ...   \n",
              "k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactoba...                               0   \n",
              "k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactoba...                               0   \n",
              "k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactoba...                               0   \n",
              "k__Bacteria|p__Firmicutes|c__Clostridia|o__Clos...                               0   \n",
              "k__Bacteria|p__Firmicutes|c__Negativicutes|o__S...                               0   \n",
              "\n",
              "                                                   freq  \n",
              "0                                                        \n",
              "dataset_name                                        121  \n",
              "sampleID                                              1  \n",
              "subjectID                                             1  \n",
              "bodysite                                            121  \n",
              "disease                                              48  \n",
              "...                                                 ...  \n",
              "k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactoba...  120  \n",
              "k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactoba...  117  \n",
              "k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactoba...  120  \n",
              "k__Bacteria|p__Firmicutes|c__Clostridia|o__Clos...  120  \n",
              "k__Bacteria|p__Firmicutes|c__Negativicutes|o__S...  120  \n",
              "\n",
              "[714 rows x 4 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-d4352922-d963-41e0-823b-f41d91d5d70c\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>count</th>\n",
              "      <th>unique</th>\n",
              "      <th>top</th>\n",
              "      <th>freq</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>dataset_name</th>\n",
              "      <td>121</td>\n",
              "      <td>1</td>\n",
              "      <td>Zeller_fecal_colorectal_cancer</td>\n",
              "      <td>121</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>sampleID</th>\n",
              "      <td>121</td>\n",
              "      <td>121</td>\n",
              "      <td>CCIS98832363ST-4-0</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>subjectID</th>\n",
              "      <td>121</td>\n",
              "      <td>121</td>\n",
              "      <td>fr-552</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>bodysite</th>\n",
              "      <td>121</td>\n",
              "      <td>1</td>\n",
              "      <td>stool</td>\n",
              "      <td>121</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>disease</th>\n",
              "      <td>121</td>\n",
              "      <td>3</td>\n",
              "      <td>cancer</td>\n",
              "      <td>48</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Enterococcaceae|g__Bavariicoccus|s__Bavariicoccus_seileri</th>\n",
              "      <td>121</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>120</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Enterococcaceae|g__Enterococcus|s__Enterococcus_gilvus</th>\n",
              "      <td>121</td>\n",
              "      <td>5</td>\n",
              "      <td>0</td>\n",
              "      <td>117</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Lactobacillaceae|g__Lactobacillus|s__Lactobacillus_otakiensis</th>\n",
              "      <td>121</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>120</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Peptococcaceae|g__Desulfotomaculum|s__Desulfotomaculum_ruminis</th>\n",
              "      <td>121</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>120</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>k__Bacteria|p__Firmicutes|c__Negativicutes|o__Selenomonadales|f__Veillonellaceae|g__Megasphaera|s__Megasphaera_sp_BV3C16_1</th>\n",
              "      <td>121</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>120</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>714 rows × 4 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-d4352922-d963-41e0-823b-f41d91d5d70c')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-d4352922-d963-41e0-823b-f41d91d5d70c button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-d4352922-d963-41e0-823b-f41d91d5d70c');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-f5d003d7-aacc-4cb5-b787-fa73f0f888ff\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-f5d003d7-aacc-4cb5-b787-fa73f0f888ff')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-f5d003d7-aacc-4cb5-b787-fa73f0f888ff button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "summary": "{\n  \"name\": \"sample\",\n  \"rows\": 714,\n  \"fields\": [\n    {\n      \"column\": \"count\",\n      \"properties\": {\n        \"dtype\": \"date\",\n        \"min\": \"121\",\n        \"max\": \"121\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"121\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"unique\",\n      \"properties\": {\n        \"dtype\": \"date\",\n        \"min\": 1,\n        \"max\": 121,\n        \"num_unique_values\": 103,\n        \"samples\": [\n          23\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"top\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 21,\n        \"samples\": [\n          \"Zeller_fecal_colorectal_cancer\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"freq\",\n      \"properties\": {\n        \"dtype\": \"date\",\n        \"min\": \"1\",\n        \"max\": \"121\",\n        \"num_unique_values\": 105,\n        \"samples\": [\n          \"77\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 7
        }
      ],
      "source": [
        "# Have a generate descrption of the sample dataset\n",
        "sample.describe().T"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "QTkZz-HrofL1"
      },
      "source": [
        "### 4.2.3. Data Preprocess\n",
        "\n",
        "According to the original paper, the original data provided in the codebase have been preprocessed and cleaned properly. However, we still need to identify the features columns and labels to extract `X` and `Y`, respectively.\n",
        "\n",
        "In abundance data, we extradct features by `k__`. In marker data, the features are chracterized with `gi|`.\n",
        "\n",
        "For labels of different deseases, we created an dictionary `label_dict` to make the labels into binary numbers (i.e. either 1 or 0)."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {
        "id": "acW06e9MbqgM"
      },
      "outputs": [],
      "source": [
        "\n",
        "label_dict = {\n",
        "    # Controls\n",
        "    'n': 0,\n",
        "    # Chirrhosis\n",
        "    'cirrhosis': 1,\n",
        "    # Colorectal Cancer\n",
        "    'cancer': 1, 'small_adenoma': 0,\n",
        "    # IBD\n",
        "    'ibd_ulcerative_colitis': 1, 'ibd_crohn_disease': 1,\n",
        "    # T2D and WT2D\n",
        "    't2d': 1,\n",
        "    # Obesity\n",
        "    'leaness': 0, 'obesity': 1,\n",
        "}\n",
        "\n",
        "\n",
        "def load_data(data_dir, filename, dtype=None):\n",
        "    feature_string = ''\n",
        "    if filename.split('_')[0] == 'abundance':\n",
        "        feature_string = \"k__\"\n",
        "    if filename.split('_')[0] == 'marker':\n",
        "        feature_string = \"gi|\"\n",
        "    # read file\n",
        "    filename = data_dir + filename\n",
        "    if not os.path.isfile(filename):\n",
        "        print(\"FileNotFoundError: File {} does not exist\".format(filename))\n",
        "        exit()\n",
        "    raw = pd.read_csv(filename, sep='\\t', index_col=0, header=None)\n",
        "\n",
        "    # select rows having feature index identifier string\n",
        "    X = raw.loc[raw.index.str.contains(feature_string, regex=False)].T\n",
        "\n",
        "    # get class labels\n",
        "    Y = raw.loc['disease']\n",
        "    Y = Y.replace(label_dict)\n",
        "\n",
        "    # train and test split\n",
        "    X_train, X_test, y_train, y_test = train_test_split(\n",
        "        X.values.astype(dtype), Y.values.astype('int'), test_size=0.2,\n",
        "        random_state=seed, stratify=Y.values\n",
        "        )\n",
        "    print(\"Train data shape: \", X_train.shape)\n",
        "    print(\"Test data shape: \", X_test.shape)\n",
        "    return X_train, X_test, y_train, y_test"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "GXSpN9sMbqgN"
      },
      "source": [
        "##   4.3. Model\n",
        "The model includes the model definitation which usually is a class, model training, and other necessary parts.\n",
        "  * Model architecture: layer number/size/type, activation function, etc\n",
        "  * Training objectives: loss function, optimizer, weight of each loss term, etc\n",
        "  * Others: whether the model is pretrained, Monte Carlo simulation for uncertainty analysis, etc\n",
        "  * The code of model should have classes of the model, functions of model training, model validation, etc.\n",
        "  * If your model training is done outside of this notebook, please upload the trained model here and develop a function to load and test it.\n",
        "\n",
        "### 4.3.1. Citation to the Original Paper\n",
        "\n",
        "Oh, M., & Zhang, L. (2020). DeepMicro: deep representation learning for disease prediction based on microbiome data. Scientific reports, 10(1), 6026.\n",
        "\n",
        "### 4.3.2. Link to the Original Paper's Repo\n",
        "[DeepMicro Github Repository](https://github.com/minoh0201/DeepMicro)\n",
        "\n",
        "### 4.3.3. Model Descriptions\n",
        "\n",
        "An autoencoder represents a type of neural network designed for the purpose of reconstructing its input data, denoted as $x$. In its fundamental structure, it comprises an encoder function, denoted as $f_\\phi(⋅)$, and a decoder function, denoted as $f_\\theta′(⋅)$, with $\\phi$ and $\\theta$ serving as the parameters associated with the encoder and decoder functions, respectively.\n",
        "The training objective of an autoencoder is to minimize the disparity between the original input $x$ and its reconstructed counterpart $x′$. This discrepancy, typically quantified using a reconstruction loss metric such as squared error, can be mathematically expressed as $L(x, x')=||x-x'||^2=||x-f_\\theta'(f_\\phi(x))||^2$.\n",
        "\n",
        "In this project, we focus on the utilization of a trained autoencoder to obtain a lower-dimensional latent representation $z = f_\\phi(x)$ of the input. There are four autoencoders that we incorporate as below.\n",
        "\n",
        "1. Shallow Autoencoder (SAE): a fully connected encoder connecting the input layer to the latent layer, and a decoder producing the reconstructed input $x′$ by combining the outputs of the latent layer using weighted sums, with both the latent and output layers utilizing a linear activation function.\n",
        "\n",
        "2. Deep Autoencoder (DAE): enhanced SAE model by inserting hidden layers with Rectified Linear Unit (ReLu) activation function and Glorot uniform initializer between the input and latent layers, maintaining an equal number of hidden layers (either one or two layers) in both the encoder and decoder sections.\n",
        "\n",
        "3. Variational autoencoder (VAE): it learns probabilistic representations $z$ by approximating the true posterior distribution with $q_\\phi(z|x)$ assuming a Gaussian distribution. The encoder encodes the means and variances of the Gaussian distribution, allowing sampling of latent representation $z$. This sampled representation is then fed into the decoder network to generate the reconstructed input $x′ \\sim g_\\theta(x|z)$.\n",
        "\n",
        "4. convolutional autoencoder (CAE): equipped with convolutional layers where each unit is connected locally to the previous layer. The layers consist of multiple filters with weights for convolution operations. We employed ReLu activation, Glorot uniform initializer, and avoided pooling layers to prevent excessive generalization. The $n$-dimensional input vector was reshaped into a squared image of size $d \\times d \\times 1$, where $d = ⌊ \\sqrt{n} ⌋ + 1$.\n",
        "\n",
        "After getting the low-dimentional representations of profiles, we build classification models.\n",
        "\n",
        "1. Support Vector Machine (SVM): a grid search using SVM is conducted to explore the hyper-parameter space. We considered both radial basis function (RBF) and linear kernels, adjusting penalty parameter C and kernel coefficient gamma for RBF.\n",
        "\n",
        "2. Random Forest (RF): We examine two criteria, Gini impurity and information gain, for selecting features to split a node in a decision tree. The maximum number of features considered for the best split at each node was determined using the square root of the sample size and the logarithm to base 2 of the sample size.\n",
        "\n",
        "3. Multi-Layer Perceptron (MLP): ReLu activations were used for the hidden layers, while the output layer employed sigmoid activation with a single unit. The number of units in the hidden layers was set to half of the preceding layer, excluding the first hidden layer.\n",
        "\n",
        "The overall workflow is demonstrated in the Figure 1 below.\n",
        "\n",
        "![](assets/Model_Figure1.png)\n",
        "\n",
        "### 4.3.4. Implementation Code\n",
        "\n",
        "Please find the implementations of autoencoders in the section [Autoencoders Definitions](#Autoencoders-Definitions) below."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "c3XEF8AmtwSI"
      },
      "source": [
        "#### Autoencoders Definitions\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {
        "id": "pWkNlD2ORZoG"
      },
      "outputs": [],
      "source": [
        "class ShallowAutoEncoder(nn.Module):\n",
        "    def __init__(self, input_dim=28 * 28):\n",
        "        super(ShallowAutoEncoder, self).__init__()\n",
        "        self.input_dim = input_dim\n",
        "        self.encoder1 = nn.Linear(input_dim, 32)\n",
        "        self.decoder1 = nn.Linear(32, input_dim)\n",
        "\n",
        "    def encoder(self, x):\n",
        "        return torch.tanh(self.encoder1(x))\n",
        "\n",
        "    def decoder(self, z):\n",
        "        return torch.sigmoid(self.decoder1(z))\n",
        "\n",
        "    def forward(self, x):\n",
        "        z = self.encoder(x.view(-1, self.input_dim))\n",
        "        return self.decoder(z)\n",
        "\n",
        "    @staticmethod\n",
        "    def loss_func(x_hat, x):\n",
        "        return F.mse_loss(x_hat, x, reduction='sum')\n",
        "\n",
        "\n",
        "class AutoEncoder(nn.Module):\n",
        "    def __init__(self, input_dim=28*28):\n",
        "        super(AutoEncoder, self).__init__()\n",
        "        self.input_dim = input_dim\n",
        "\n",
        "        # DNN as encoder\n",
        "        # 512-256-128\n",
        "        self.encoder1 = nn.Linear(input_dim, 512)\n",
        "        self.encoder2 = nn.Linear(512, 256)\n",
        "        self.encoder3 = nn.Linear(256, 128)\n",
        "\n",
        "        # DNN as decoder\n",
        "        self.decoder1 = nn.Linear(128, 256)\n",
        "        self.decoder2 = nn.Linear(256, 512)\n",
        "        self.decoder3 = nn.Linear(512, input_dim)\n",
        "\n",
        "    def encoder(self, x):\n",
        "        h = torch.tanh(self.encoder1(x))\n",
        "        h = torch.tanh(self.encoder2(h))\n",
        "        return torch.tanh(self.encoder3(h))\n",
        "\n",
        "    def decoder(self, z):\n",
        "        h = torch.tanh(self.decoder1(z))\n",
        "        h = torch.tanh(self.decoder2(h))\n",
        "        return torch.sigmoid(self.decoder3(h))\n",
        "\n",
        "    def forward(self, x):\n",
        "        z = self.encoder(x.view(-1, self.input_dim))\n",
        "        return self.decoder(z)\n",
        "\n",
        "    @staticmethod\n",
        "    def loss_func(x_hat, x):\n",
        "        return F.mse_loss(x_hat, x)\n",
        "\n",
        "\n",
        "class CAE(nn.Module):\n",
        "    def __init__(self):\n",
        "        super(CAE, self).__init__()\n",
        "\n",
        "        # CNN as encoder\n",
        "        self.encoder = nn.Sequential(\n",
        "            nn.Conv2d(1, 16, 3, stride=2, padding=1),\n",
        "            nn.ReLU(),\n",
        "            nn.Conv2d(16, 32, 3, stride=2, padding=1),\n",
        "            nn.ReLU(),\n",
        "            nn.Conv2d(32, 64, 6)\n",
        "        )\n",
        "\n",
        "        # CNN as decoder\n",
        "        self.decoder = nn.Sequential(\n",
        "            nn.ConvTranspose2d(64, 32, 6),\n",
        "            nn.ReLU(),\n",
        "            nn.ConvTranspose2d(32, 16, 3, stride=2, padding=1, output_padding=1),\n",
        "            nn.ReLU(),\n",
        "            nn.ConvTranspose2d(16, 1, 3, stride=2, padding=1, output_padding=1),\n",
        "            nn.Sigmoid()\n",
        "        )\n",
        "\n",
        "    def forward(self, x):\n",
        "        z = self.encoder(x)\n",
        "        return self.decoder(z)\n",
        "\n",
        "    @staticmethod\n",
        "    def loss_func(x_hat, x):\n",
        "        return F.mse_loss(x_hat, x)\n",
        "\n",
        "\n",
        "class VAE(nn.Module):\n",
        "    def __init__(self, input_dim=28 * 28, latent_dim=8):\n",
        "        super(VAE, self).__init__()\n",
        "        self.input_dim = input_dim\n",
        "\n",
        "        # DNN as encoder\n",
        "        self.encoder1 = nn.Linear(input_dim, 128)\n",
        "        self.mu = nn.Linear(128, 8)\n",
        "        self.log_sigma2 = nn.Linear(128, 8)\n",
        "\n",
        "        # DNN as decoder\n",
        "        self.decoder1 = nn.Linear(8, 128)\n",
        "        self.decoder2 = nn.Linear(128, input_dim)\n",
        "\n",
        "    def encoder(self, x):\n",
        "        h = torch.tanh(self.encoder1(x))\n",
        "        return self.mu(h), torch.sqrt(torch.exp(self.log_sigma2(h)))\n",
        "\n",
        "    @staticmethod\n",
        "    def sampling(mu, std):  # Re parameterization trick\n",
        "        eps = torch.randn_like(std)\n",
        "        return mu + eps * std\n",
        "\n",
        "    def decoder(self, z):\n",
        "        h = torch.tanh(self.decoder1(z))\n",
        "        return torch.sigmoid(self.decoder2(h))\n",
        "\n",
        "    def forward(self, x):\n",
        "        mu, std = self.encoder(x.view(-1, self.input_dim))\n",
        "        z = self.sampling(mu, std)\n",
        "        return self.decoder(z), mu, std\n",
        "\n",
        "    @staticmethod\n",
        "    def loss_func(x_hat, x, mu, std):\n",
        "        # define the reconstruction loss\n",
        "        ERR = F.binary_cross_entropy(x_hat, x, reduction='sum')\n",
        "        # define the KL divergence loss\n",
        "        KLD = -0.5 * torch.sum(1 + torch.log(std ** 2) - mu ** 2 - std ** 2)\n",
        "        return ERR + KLD, ERR, KLD\n"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### Classification Models"
      ],
      "metadata": {
        "id": "f349y3iuE-Cz"
      }
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {
        "id": "s2hbJUcCjiEY"
      },
      "outputs": [],
      "source": [
        "from sklearn.neural_network import MLPClassifier\n",
        "\n",
        "def get_metrics(clf, X_test_encoded, y_test):\n",
        "    y_true, y_pred = y_test, clf.predict(X_test_encoded)\n",
        "    y_prob = clf.predict_proba(X_test_encoded)[:, 1]\n",
        "    # Performance Metrics: AUC, ACC, Recall, Precision, F1_score\n",
        "    metrics = {\n",
        "        'AUC': roc_auc_score(y_true, y_prob),\n",
        "        'ACC': accuracy_score(y_true, y_pred),\n",
        "        'Recall': recall_score(y_true, y_pred),\n",
        "        'Precision': precision_score(y_true, y_pred),\n",
        "        'F1_score': f1_score(y_true, y_pred),\n",
        "    }\n",
        "    return metrics\n",
        "\n",
        "def classification_with_pytorch_model(model, X_train, y_train, X_test, y_test, device, is_vae=False):\n",
        "    model = model.to(device)\n",
        "    # check if X_train is tensor already\n",
        "    if not torch.is_tensor(X_train):\n",
        "        X_train = torch.tensor(X_train).to(device)\n",
        "        X_test = torch.tensor(X_test).to(device)\n",
        "\n",
        "    if is_vae:\n",
        "        X_train_encoded = model.encoder(X_train)[0].cpu().detach().numpy()\n",
        "        X_test_encoded = model.encoder(X_test)[0].cpu().detach().numpy()\n",
        "    else:\n",
        "        X_train_encoded = model.encoder(X_train).cpu().detach().numpy()\n",
        "        X_test_encoded = model.encoder(X_test).cpu().detach().numpy()\n",
        "    X_train_encoded = np.reshape(X_train_encoded, (X_train_encoded.shape[0], -1))\n",
        "    X_test_encoded = np.reshape(X_test_encoded, (X_test_encoded.shape[0], -1))\n",
        "    return classification(X_train_encoded, y_train, X_test_encoded, y_test)\n",
        "\n",
        "\n",
        "def run_sub_classification(method, X_train_encoded, y_train, X_test_encoded, y_test):\n",
        "    clf = \"\"\n",
        "    hyper_parameters = {}\n",
        "    # hyper-parameter grids for classifiers\n",
        "    rf_hyper_parameters = [{'n_estimators': [s for s in range(100, 1001, 200)],\n",
        "                            'max_features': ['sqrt', 'log2'],\n",
        "                            'min_samples_leaf': [1, 2, 3, 4, 5],\n",
        "                            'criterion': ['gini', 'entropy']\n",
        "                            }, ]\n",
        "    svm_hyper_parameters = [{'C': [2 ** s for s in range(-5, 6, 2)], 'kernel': ['linear']},\n",
        "                            {'C': [2 ** s for s in range(-5, 6, 2)], 'gamma': [2 ** s for s in range(3, -15, -2)],\n",
        "                             'kernel': ['rbf']}]\n",
        "    mlp_hyper_parameters = [{'hidden_layer_sizes': [(64, 64), (128, 128), (256, 256), (512, 512)],\n",
        "                             'alpha': [2 ** s for s in range(-15, 1, 2)]}]\n",
        "\n",
        "    if method == 'SVM':\n",
        "        clf = SVC(kernel='linear', probability=True)\n",
        "        hyper_parameters = svm_hyper_parameters\n",
        "    elif method == 'Random Forest':\n",
        "        clf = RandomForestClassifier(n_estimators=100)\n",
        "        hyper_parameters = rf_hyper_parameters\n",
        "    elif method == 'MLP':\n",
        "        clf = MLPClassifier(hidden_layer_sizes=(64, 64), max_iter=1000)\n",
        "        hyper_parameters = mlp_hyper_parameters\n",
        "    clf = GridSearchCV(clf, param_grid=hyper_parameters,\n",
        "                                cv=StratifiedKFold(n_splits=5, shuffle=True), scoring='roc_auc',\n",
        "                                verbose=1, n_jobs=-1)\n",
        "    clf.fit(X_train_encoded, y_train)\n",
        "    metrics = get_metrics(clf, X_test_encoded, y_test)\n",
        "    metrics['classification_method'] = method\n",
        "    print(metrics)\n",
        "    print(\"Best hyper-parameters: \", clf.best_params_)\n",
        "    metrics['hyper_parameters'] = clf.best_params_\n",
        "    return metrics\n",
        "\n",
        "\n",
        "# Classification by using the transformed data\n",
        "def classification(X_train_encoded, y_train, X_test_encoded, y_test):\n",
        "    methods = ['MLP', 'SVM', 'Random Forest']\n",
        "    metrics_list = []\n",
        "    for method in methods:\n",
        "        classification_start_time = time.time()\n",
        "        metrics = run_sub_classification(method, X_train_encoded.copy(), y_train.copy(), X_test_encoded.copy(),\n",
        "                                         y_test.copy())\n",
        "        classification_end_time = time.time()\n",
        "        print(method, \"Classification time: \", classification_end_time - classification_start_time)\n",
        "        metrics['classification_time'] = classification_end_time - classification_start_time\n",
        "        metrics_list.append(metrics)\n",
        "    return metrics_list"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "LS3jIzt9uj9B"
      },
      "source": [
        "## 4.4. Training\n",
        "\n",
        "### 4.4.1. Training Objectives\n",
        "- Split each dataset into a training set, validation set, and test set (64% training, 16% validation, and 20% test).\n",
        "- Exclude the test set from model training.\n",
        "- Implement early-stopping strategy: train models on the training set, compute reconstruction loss for the validation set after each epoch, stop training if no improvement in validation loss is observed for 20 epochs.\n",
        "- Select the model with the lowest validation loss as the best model.\n",
        "Utilize mean squared error as the reconstruction loss metric.\n",
        "- Apply adaptive moment estimation (Adam) optimizer with default parameters (learning rate: 0.001, epsilon: 1e-07) as specified in the original paper.\n",
        "- Utilize the encoder part of the best model to generate low-dimensional representations of microbiome data for subsequent disease prediction tasks.\n",
        "\n",
        "### 4.4.2. Hyperparams\n",
        "\n",
        "In the base case of svm classification, we have hyperparameter number of folds equal to 5. The epoch is set at 200 and batch set is 32 for each autoencoder algorithm.\n",
        "\n",
        "The original implementation use command line arguments to facilitate the execution of experiments with varying configurations. In order to enhance usability within a Colab notebook environment, we modified this approach by setting the hyperparameters in our model directly.\n",
        "\n",
        "\n",
        "### 4.4.3. Computation Requirements\n",
        "\n",
        "According to the original paper, an equipment with an Intel Core i9-9820X CPU (10 cores), 64 GB Memory, and a GPU of NVIDIA GTX 1080 Ti, running on Ubuntu 18.04 have the latency in time as in Table 3 below.\n",
        "\n",
        "![Table 3](assets/Computation_Table3.png)\n",
        "\n",
        "In our experiments, we utilized the following computational resources\n",
        "\n",
        "1. Hardware: Mac mini with Apple M1 chip.\n",
        "2. Average runtime for each epoch:\n",
        "  - PCA:  10s\n",
        "  - Random Forest: 10s\n",
        "  - AE:  970s\n",
        "  - CAE: 1940s\n",
        "  - VAE: 248s\n",
        "3. GPU hours used: Since the Apple M1 chip integrates both the CPU and GPU, the usage is described in terms of CPU hours rather than specifically GPU hours.\n"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### Training Code"
      ],
      "metadata": {
        "id": "qGdp9R-IFwAZ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "from torch.utils.data import DataLoader, TensorDataset\n",
        "\n",
        "epoch_num = 200\n",
        "seed = 42\n",
        "\n",
        "def train_vae(model, train_loader, test_loader, device):\n",
        "    model = model.to(device)\n",
        "    optimizer = torch.optim.AdamW(model.parameters(), lr=0.001)\n",
        "    rec_loss, kl_loss = [], []\n",
        "    for epoch in range(1, epoch_num + 1):\n",
        "\n",
        "        \"\"\" model training \"\"\"\n",
        "        model.train()\n",
        "        cur_rec_loss, cur_kl_loss = [], []\n",
        "        for batch_idx, (data, _) in enumerate(train_loader):\n",
        "            data = data.to(device)\n",
        "            rec, mu, std = model(data)\n",
        "            loss, err, kl = model.loss_func(rec, data.reshape(-1, model.input_dim), mu, std)\n",
        "\n",
        "            optimizer.zero_grad()\n",
        "            loss.backward()\n",
        "            optimizer.step()\n",
        "\n",
        "            cur_rec_loss.append(err.item())\n",
        "            cur_kl_loss.append(kl.item())\n",
        "\n",
        "        rec_loss.append(np.mean(cur_rec_loss))\n",
        "        kl_loss.append(np.mean(cur_kl_loss))\n",
        "\n",
        "        \"\"\" model evaluation \"\"\"\n",
        "        with torch.no_grad():\n",
        "            test_loss = []\n",
        "            for batch_idx, (data, _) in enumerate(test_loader):\n",
        "                data = data.to(device)\n",
        "                rec, mu, std = model(data)\n",
        "                _, mse, _ = model.loss_func(rec, data.reshape(data.shape[0], -1), mu, std)\n",
        "                test_loss.append(mse.item())\n",
        "\n",
        "        if epoch == 1 or epoch % 10 == 0:\n",
        "            print(\n",
        "                f\"-- epoch {epoch} --, train MSE: {np.mean(cur_rec_loss)}, train KL: {np.mean(cur_kl_loss)}, test MSE: {np.mean(test_loss)}\")\n",
        "    return model\n",
        "\n",
        "\n",
        "def train_ae(model, train_loader, test_loader, device):\n",
        "    # TODO: cache the model\n",
        "    optimizer = torch.optim.AdamW(model.parameters(), lr=0.002)\n",
        "\n",
        "    losses = []\n",
        "    for epoch in range(1, epoch_num + 1):\n",
        "\n",
        "        \"\"\" model training \"\"\"\n",
        "        model.train()\n",
        "        cur_rec_loss = []\n",
        "        for batch_idx, (data, _) in enumerate(train_loader):\n",
        "            data = data.to(device)\n",
        "            rec = model(data)\n",
        "            loss = model.loss_func(rec, data)\n",
        "\n",
        "            optimizer.zero_grad()\n",
        "            loss.backward()\n",
        "            optimizer.step()\n",
        "\n",
        "            cur_rec_loss.append(loss.item())\n",
        "        losses.append(np.mean(cur_rec_loss))\n",
        "\n",
        "        \"\"\" model evaluation \"\"\"\n",
        "        with torch.no_grad():\n",
        "            test_loss = []\n",
        "            for batch_idx, (data, _) in enumerate(test_loader):\n",
        "                data = data.to(device)\n",
        "                rec = model(data)\n",
        "                loss = model.loss_func(rec, data)\n",
        "                test_loss.append(loss.item())\n",
        "\n",
        "        if epoch == 1 or epoch % 10 == 0:\n",
        "            print(f\"-- epoch {epoch} --, train MSE: {np.mean(cur_rec_loss)}, test MSE: {np.mean(test_loss)}\")\n",
        "\n",
        "    return model\n",
        "\n",
        "\n",
        "def run_shallow_ae_experiment(X_train, X_test, y_train, y_test, train_loader, test_loader, device):\n",
        "    model = ShallowAutoEncoder(input_dim=X_train.shape[1]).to(device)\n",
        "    model = train_ae(model, train_loader, test_loader, device)\n",
        "    return classification_with_pytorch_model(model, X_train, y_train, X_test, y_test, device)\n",
        "\n",
        "\n",
        "def run_ae_experiment(X_train, X_test, y_train, y_test, train_loader, test_loader, device):\n",
        "    model = AutoEncoder(input_dim=X_train.shape[1]).to(device)\n",
        "    model = train_ae(model, train_loader, test_loader, device)\n",
        "    return classification_with_pytorch_model(model, X_train, y_train, X_test, y_test, device)\n",
        "\n",
        "\n",
        "def run_cae_experiment(X_train, X_test, y_train, y_test, device):\n",
        "    # one_side_dim = int(math.sqrt(X_train.shape[1])) + 1\n",
        "    one_side_dim = 348\n",
        "    enlarged_dim = one_side_dim ** 2\n",
        "    X_train = np.column_stack((X_train, np.zeros((X_train.shape[0], enlarged_dim - X_train.shape[1]))))\n",
        "    X_test = np.column_stack((X_test, np.zeros((X_test.shape[0], enlarged_dim - X_test.shape[1]))))\n",
        "    # reshape\n",
        "    X_train = np.reshape(X_train, (len(X_train), one_side_dim, one_side_dim, 1))\n",
        "    X_test = np.reshape(X_test, (len(X_test), one_side_dim, one_side_dim, 1))\n",
        "    # to float32\n",
        "    X_train = X_train.astype('float32')\n",
        "    X_test = X_test.astype('float32')\n",
        "    X_train = torch.tensor(X_train).to(device)\n",
        "    X_test = torch.tensor(X_test).to(device)\n",
        "    X_train = X_train.transpose(1, 3)\n",
        "    X_test = X_test.transpose(1, 3)\n",
        "    train_loader = DataLoader(TensorDataset(torch.tensor(X_train), torch.tensor(y_train)), batch_size=32, shuffle=True)\n",
        "    test_loader = DataLoader(TensorDataset(torch.tensor(X_test), torch.tensor(y_test)), batch_size=32, shuffle=False)\n",
        "    model = CAE().to(device)\n",
        "    model = train_ae(model, train_loader, test_loader, device)\n",
        "    return classification_with_pytorch_model(model, X_train, y_train, X_test, y_test, device)\n",
        "\n",
        "\n",
        "def run_vae_experiment(X_train, X_test, y_train, y_test, train_loader, test_loader, device):\n",
        "    model = VAE(input_dim=X_train.shape[1]).to(device)\n",
        "    model = train_vae(model, train_loader, test_loader, device)\n",
        "    return classification_with_pytorch_model(model, X_train, y_train, X_test, y_test, device, is_vae=True)\n",
        "\n",
        "\n",
        "def run_pca_experiment(X_train, X_test, y_train, y_test):\n",
        "    pca = PCA(n_components=10)\n",
        "    pca.fit(X_train)\n",
        "    X_train_pca = pca.transform(X_train)\n",
        "    X_test_pca = pca.transform(X_test)\n",
        "    return classification(X_train_pca, y_train, X_test_pca, y_test)\n",
        "\n",
        "\n",
        "def run_rp_experiment(X_train, X_test, y_train, y_test):\n",
        "    rp = GaussianRandomProjection(eps=0.5)\n",
        "    rp.fit(X_train)\n",
        "    X_train_rp = rp.transform(X_train)\n",
        "    X_test_rp = rp.transform(X_test)\n",
        "    return classification(X_train_rp, y_train, X_test_rp, y_test)\n",
        "\n",
        "\n",
        "def run_plain_experiment(X_train, X_test, y_train, y_test):\n",
        "    return classification(X_train, y_train, X_test, y_test)\n",
        "\n",
        "\n",
        "def run_sub_experiment_on_data(model_type, X_train, X_test, y_train, y_test, train_loader, test_loader, device):\n",
        "    metrics_list = []\n",
        "    if model_type == 'shallow_ae':\n",
        "        metrics_list = run_shallow_ae_experiment(X_train, X_test, y_train, y_test, train_loader, test_loader, device)\n",
        "    elif model_type == 'ae':\n",
        "        metrics_list = run_ae_experiment(X_train, X_test, y_train, y_test, train_loader, test_loader, device)\n",
        "    elif model_type == 'cae':\n",
        "        metrics_list = run_cae_experiment(X_train, X_test, y_train, y_test, device)\n",
        "    elif model_type == 'vae':\n",
        "        metrics_list = run_vae_experiment(X_train, X_test, y_train, y_test, train_loader, test_loader, device)\n",
        "    elif model_type == 'pca':\n",
        "        metrics_list = run_pca_experiment(X_train, X_test, y_train, y_test)\n",
        "    elif model_type == 'rp':\n",
        "        metrics_list = run_rp_experiment(X_train, X_test, y_train, y_test)\n",
        "    elif model_type == 'plain':\n",
        "        metrics_list = run_plain_experiment(X_train, X_test, y_train, y_test)\n",
        "    return metrics_list\n",
        "\n"
      ],
      "metadata": {
        "id": "xhLeFpd1Fobs"
      },
      "execution_count": 11,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## 4.5. Evaluation\n",
        "\n",
        "### 4.5.1. Metrics Descriptions\n",
        "\n",
        "- Conduct 5-fold cross-validation on the reduced training set. This involves dividing the training set into five subsets, using four subsets for training and one for validation in each fold. This is to vary hyper-parameters and explore different combinations to find the best performing configuration.\n",
        "- Evaluate the performance of the models using the area under the receiver operating characteristics curve (AUC). This metric assesses the model's ability to distinguish between different classes and is commonly used for classification tasks.\n",
        "- Train a final classification model using the entire training set and the best hyper-parameter combination identified during cross-validation. This model aims to achieve optimal performance based on the selected configuration. Then test the final classification model on the separate test set, which was not used during training. This evaluation provides an unbiased assessment of the model's performance on unseen data.\n",
        "- Repeat the entire procedure five times, each time using a different random partition seed to create new training, validation, and test sets. This helps account for potential variations in performance due to the specific data splits. Average the resulting AUC scores obtained from the five repetitions. This average serves as a summary metric to compare the performance of different models or approaches. It provides a more robust assessment by considering multiple iterations of the evaluation process.\n",
        "\n",
        "### 4.5.2. Evaluation Code\n",
        "\n",
        "The detailed cross-validation codes are in the `run_sub_classification` function of the [Classification Models](#Classification-Models) section."
      ],
      "metadata": {
        "id": "Nbdu_gV5FetZ"
      }
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {
        "id": "5WDva6WosGE6"
      },
      "outputs": [],
      "source": [
        "\n",
        "def run_experiment_on_data(data_dir, data_string):\n",
        "    X_train, X_test, y_train, y_test = load_data(data_dir, data_string, dtype='float32')\n",
        "    train_loader = DataLoader(\n",
        "        TensorDataset(torch.tensor(X_train), torch.tensor(y_train)),\n",
        "        batch_size=32, shuffle=True)\n",
        "    test_loader = DataLoader(TensorDataset(torch.tensor(X_test), torch.tensor(y_test)),\n",
        "                             batch_size=32, shuffle=False)\n",
        "    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')\n",
        "    if torch.backends.mps.is_available():\n",
        "        device = torch.device(\"mps\")\n",
        "\n",
        "    all_metrics_df_on_current_data = pd.DataFrame()\n",
        "    model_types = ['shallow_ae', 'ae', 'cae', 'vae', 'pca', 'rp', 'plain']\n",
        "    for model_type in model_types:\n",
        "        print(\"Running experiment on model: \", model_type)\n",
        "        experiment_start_time = time.time()\n",
        "        metrics_list = run_sub_experiment_on_data(model_type, X_train, X_test, y_train, y_test, train_loader,\n",
        "                                                  test_loader, device)\n",
        "        experiment_end_time = time.time()\n",
        "        print(\"Experiment time: \", experiment_end_time - experiment_start_time)\n",
        "\n",
        "        for metrics in metrics_list:\n",
        "            metrics['experiment_time'] = experiment_end_time - experiment_start_time\n",
        "            metrics['data'] = data_string\n",
        "            metrics['model'] = model_type\n",
        "            metrics['seed'] = seed\n",
        "            metrics['epoch_num'] = epoch_num\n",
        "            metrics['device'] = device\n",
        "            print(metrics)\n",
        "            # save metrics\n",
        "        metrics_df = pd.DataFrame(metrics_list)\n",
        "        all_metrics_df_on_current_data = pd.concat([all_metrics_df_on_current_data, metrics_df], ignore_index=True)\n",
        "    return all_metrics_df_on_current_data\n",
        "\n",
        "\n",
        "def run_experiment():\n",
        "    data_dir = './data/marker/'\n",
        "    # main_metrics_df = run_experiment_on_data(data_dir, \"marker_Cirrhosis.txt\")\n",
        "    data_list = os.listdir(data_dir)\n",
        "    for data_string in data_list:\n",
        "        start = datetime.datetime.now()\n",
        "        if data_string == \"marker_Cirrhosis.txt\":\n",
        "            continue\n",
        "        main_metrics_df = pd.DataFrame()\n",
        "        print(\"Running experiment on data: \", data_string)\n",
        "        all_metrics_df_on_current_data = run_experiment_on_data(data_dir, data_string)\n",
        "        main_metrics_df = pd.concat([main_metrics_df, all_metrics_df_on_current_data], ignore_index=True)\n",
        "        end = datetime.datetime.now()\n",
        "        csv_file_name = end.strftime(\"%Y%m%d-%H%M%S\") + f'_metrics_{int((end - start).seconds)}sec.csv'\n",
        "        main_metrics_df.to_csv(csv_file_name, index=False)\n",
        "        print(\"Metrics saved to file: \", csv_file_name)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "hY7oRbSRldrn"
      },
      "source": [
        "#### Run Experiments"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "jY5P3723sFjE",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "edd12254-5b2d-4d7e-cff5-dca4440907ee"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Running experiment on data:  marker_IBD.txt\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "<ipython-input-8-510a80379f97>:28: DtypeWarning: Columns (1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110) have mixed types. Specify dtype option on import or set low_memory=False.\n",
            "  raw = pd.read_csv(filename, sep='\\t', index_col=0, header=None)\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Train data shape:  (88, 91756)\n",
            "Test data shape:  (22, 91756)\n",
            "Running experiment on model:  shallow_ae\n",
            "-- epoch 1 --, train MSE: 713403.5625, test MSE: 522879.8125\n",
            "-- epoch 10 --, train MSE: 288214.0729166667, test MSE: 217030.890625\n",
            "-- epoch 20 --, train MSE: 231141.49479166666, test MSE: 184107.4375\n",
            "-- epoch 30 --, train MSE: 224634.04166666666, test MSE: 180357.8125\n",
            "-- epoch 40 --, train MSE: 222423.33333333334, test MSE: 179167.65625\n",
            "-- epoch 50 --, train MSE: 221288.28125, test MSE: 178567.546875\n",
            "-- epoch 60 --, train MSE: 220570.76041666666, test MSE: 178338.15625\n",
            "-- epoch 70 --, train MSE: 220101.22395833334, test MSE: 178111.796875\n",
            "-- epoch 80 --, train MSE: 219756.390625, test MSE: 178020.078125\n",
            "-- epoch 90 --, train MSE: 219485.48958333334, test MSE: 177948.6875\n",
            "-- epoch 100 --, train MSE: 219266.55208333334, test MSE: 177860.875\n",
            "-- epoch 110 --, train MSE: 219095.72395833334, test MSE: 177858.375\n",
            "-- epoch 120 --, train MSE: 218916.5625, test MSE: 177906.484375\n",
            "-- epoch 130 --, train MSE: 218777.00520833334, test MSE: 177860.140625\n",
            "-- epoch 140 --, train MSE: 218647.33854166666, test MSE: 177859.78125\n",
            "-- epoch 150 --, train MSE: 218533.49479166666, test MSE: 177821.328125\n",
            "-- epoch 160 --, train MSE: 218432.76041666666, test MSE: 177861.984375\n",
            "-- epoch 170 --, train MSE: 218318.5625, test MSE: 177861.796875\n",
            "-- epoch 180 --, train MSE: 218217.91145833334, test MSE: 177877.90625\n",
            "-- epoch 190 --, train MSE: 218116.79166666666, test MSE: 177839.4375\n",
            "-- epoch 200 --, train MSE: 218020.08854166666, test MSE: 177847.125\n",
            "Fitting 5 folds for each of 32 candidates, totalling 160 fits\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/fork_exec.py:38: RuntimeWarning: os.fork() was called. os.fork() is incompatible with multithreaded code, and JAX is multithreaded, so this will likely lead to a deadlock.\n",
            "  pid = os.fork()\n",
            "/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/fork_exec.py:38: RuntimeWarning: os.fork() was called. os.fork() is incompatible with multithreaded code, and JAX is multithreaded, so this will likely lead to a deadlock.\n",
            "  pid = os.fork()\n",
            "/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.\n",
            "  _warn_prf(average, modifier, msg_start, len(result))\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "{'AUC': 0.5, 'ACC': 0.7727272727272727, 'Recall': 0.0, 'Precision': 0.0, 'F1_score': 0.0, 'classification_method': 'MLP'}\n",
            "Best hyper-parameters:  {'alpha': 0.03125, 'hidden_layer_sizes': (64, 64)}\n",
            "MLP Classification time:  45.14319205284119\n",
            "Fitting 5 folds for each of 60 candidates, totalling 300 fits\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.\n",
            "  _warn_prf(average, modifier, msg_start, len(result))\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "{'AUC': 0.5, 'ACC': 0.7727272727272727, 'Recall': 0.0, 'Precision': 0.0, 'F1_score': 0.0, 'classification_method': 'SVM'}\n",
            "Best hyper-parameters:  {'C': 0.03125, 'kernel': 'linear'}\n",
            "SVM Classification time:  1.2850995063781738\n",
            "Fitting 5 folds for each of 100 candidates, totalling 500 fits\n"
          ]
        }
      ],
      "source": [
        "run_experiment()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "gX6bCcZNuxmz"
      },
      "source": [
        "# Results\n",
        "In this section, you should finish training your model training or loading your trained model. That is a great experiment! You should share the results with others with necessary metrics and figures.\n",
        "\n",
        "Please test and report results for all experiments that you run with:\n",
        "\n",
        "*   specific numbers (accuracy, AUC, RMSE, etc)\n",
        "*   figures (loss shrinkage, outputs from GAN, annotation or label of sample pictures, etc)\n",
        "\n",
        "\n",
        "\n",
        "---\n",
        "\n",
        "\n",
        "**Current Progress:**\n",
        "\n",
        "Successfully run model with ae, vae, pca, cae with auc results.\n",
        "\n",
        "**To-Do:**\n",
        "\n",
        "Show figures for each (By Apr 21)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Y4WgUnso82d1"
      },
      "outputs": [],
      "source": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "8EAWAy_LwHlV"
      },
      "source": [
        "## Model comparison"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "n_30uovDnQxs"
      },
      "source": [
        "**To-Do:**\n",
        "\n",
        "Train model in full epoches and compare results. (By Apr 28)\n",
        "- we can compare the results of the different models in terms of accuracy, AUC, etc. And,\n",
        "- the time taken to train each model\n",
        "- the best hyperparameters found for each model"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "qH75TNU71eRH"
      },
      "source": [
        "# Discussion\n",
        "\n",
        "In this section,you should discuss your work and make future plan. The discussion should address the following questions:\n",
        "  * Make assessment that the paper is reproducible or not.\n",
        "  * Explain why it is not reproducible if your results are kind negative.\n",
        "  * Describe “What was easy” and “What was difficult” during the reproduction.\n",
        "  * Make suggestions to the author or other reproducers on how to improve the reproducibility.\n",
        "  * What will you do in next phase.\n",
        "\n",
        "---\n",
        "\n",
        "**To-Do:**\n",
        "\n",
        "Add discussion. (By Apr 28)\n",
        "\n",
        "- discuss the reproducibility of the paper and the ease of reproducing the results.\n",
        "- discuss what was easy and what was difficult during the reproduction.\n",
        "- make suggestions to the author or other reproducers on how to improve the reproducibility.\n",
        "- discuss the results of the model comparison and the implications of the results.\n",
        "- discuss the limitations of the study and the potential future work.\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "SHMI2chl9omn"
      },
      "source": [
        "# References\n",
        "\n",
        "1. Cho, I., & Blaser, M. J. (2012). The human microbiome: at the interface of health and disease. Nature Reviews Genetics, 13(4), 260-270.\n",
        "2. Huttenhower, C. et al. Structure, function and diversity of the healthy human microbiome. nature 486, 207 (2012).\n",
        "3. McQuade, J. L., Daniel, C. R., Helmink, B. A. & Wargo, J. A. Modulating the microbiome to improve therapeutic response in cancer. The Lancet Oncology 20, e77–e91 (2019).\n",
        "4. Eloe-Fadrosh, E. A. & Rasko, D. A. The human microbiome: from symbiosis to pathogenesis. Annual review of medicine 64, 145–163 (2013).\n",
        "5. Scholz, M. et al. Strain-level microbial epidemiology and population genomics from shotgun metagenomics. Nature methods 13, 435 (2016).\n",
        "6. Truong, D. T. et al. MetaPhlAn2 for enhanced metagenomic taxonomic profiling. Nature methods 12, 902 (2015).\n",
        "7. Oh, M., & Zhang, L. (2020). DeepMicro: deep representation learning for disease prediction based on microbiome data. Scientific reports, 10(1), 6026.\n",
        "8. Pasolli, E., Truong, D. T., Malik, F., Waldron, L. & Segata, N. Machine learning meta-analysis of large metagenomic datasets: tools and biological insights. PLoS computational biology 12, e1004977 (2016).\n",
        "9. Cawley, G. C. & Talbot, N. L. On over-fitting in model selection and subsequent selection bias in performance evaluation. Journal of Machine Learning Research 11, 2079–2107 (2010).\n",
        "10. Varma, S. & Simon, R. Bias in error estimation when using cross-validation for model selection. BMC bioinformatics 7, 91 (2006).\n"
      ]
    }
  ],
  "metadata": {
    "accelerator": "GPU",
    "colab": {
      "gpuType": "T4",
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3 (ipykernel)",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}