{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/ZKingQ/CS598-DLH-SP24/blob/main/DLH_Team_71_PyTorch.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "nfyi6FdlRDyg"
      },
      "source": [
        "# CS598 Deep Learning for Healthcare - Final Project - DeepMicro: Deep Representation Learning for Disease Prediction based on Microbiome Data\n",
        "\n",
        "### Lotte Zhu, Kaiqing Zhang, Matthew Trueblood (Team ID: 71)\n",
        "#### GitHub Link: https://github.com/ZKingQ/CS598-DLH-SP24"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "MQ0sNuMePBXx"
      },
      "source": [
        "# 1. Introduction\n",
        "<!-- This is an introduction to your report, you should edit this text/mardown section to compose. In this text/markdown, you should introduce:\n",
        "\n",
        "*   Background of the problem\n",
        "  * what type of problem: disease/readmission/mortality prediction,  feature engineeing, data processing, etc\n",
        "  * what is the importance/meaning of solving the problem\n",
        "  * what is the difficulty of the problem\n",
        "  * the state of the art methods and effectiveness.\n",
        "*   Paper explanation\n",
        "  * what did the paper propose\n",
        "  * what is the innovations of the method\n",
        "  * how well the proposed method work (in its own metrics)\n",
        "  * what is the contribution to the reasearch regime (referring the Background above, how important the paper is to the problem). -->\n",
        "\n",
        "\n",
        "The expanding knowledge of microbiota uncovers its crucial role in human health [1]. It plays an important role in immune system, metabolism functions and even carcinognesis of certain cancers, hence microbiota can be used to predict various disease with emerging sequencing technologies [1,2,3,4,5,6]. However, there are three major challenges to realize the predictions in practice [7]. First, the low number of samples, together with the large number of features, leads to the curse of dimensionality. Second, there is a research gap in using strain-level profiles to classify samples into patient and healthy control groups across different diseases. Third, a rigorous validation framework is essential. Prior research has shown that tuning hyperparameters on the test set without a separate validation set may lead to an overestimation of model performance [8,9,10].\n",
        "\n",
        "The DeepMicro paper proposes the deployment of autoencoders to learn low-dimensional representations from microbiota data and then predict disease by another classification model based on the learned representations, with both trainings having thorough validation schemes [7]. The authors hypothesize that these innovations can contribute to the followings:\n",
        "\n",
        "1. The appropriate autoencoders can effectively solve the curse of dimensionality.\n",
        "\n",
        "2. They also reduce latency compared with alternative models without representation learing, in the mean time maintaining favorable training metrics.\n",
        "\n",
        "In this project, our primary objective revolves around the implementation and evaluation of various autoencoder architectures. We aim to validate the disease prediction capabilities as outlined in the paper. Furthermore, we discuss the ablations inherent in the innovation of autoencoder. The rationale for choosing this paper stems from our pursuit of knowledge in deep representation learning.\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "2bwDxCqqbwhb"
      },
      "source": [
        "# 2. Github Repository\n",
        "\n",
        "Our project is published on [CS598-DLH-SP24](https://github.com/ZKingQ/CS598-DLH-SP24/tree/main)."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "uygL9tTPSVHB"
      },
      "source": [
        "# 3. Scope of Reproducibility\n",
        "\n",
        "<!-- List hypotheses from the paper you will test and the corresponding experiments you will run.\n",
        "\n",
        "\n",
        "1.   Hypothesis 1: xxxxxxx\n",
        "2.   Hypothesis 2: xxxxxxx\n",
        "\n",
        "You can insert images in this notebook text, [see this link](https://stackoverflow.com/questions/50670920/how-to-insert-an-inline-image-in-google-colaboratory-from-google-drive) and example below:\n",
        "\n",
        "![sample_image.png](https://drive.google.com/uc?export=view&id=1g2efvsRJDxTxKz-OY3loMhihrEUdBxbc)\n",
        "\n",
        "\n",
        "You can also use code to display images, see the code below.\n",
        "\n",
        "The images must be saved in Google Drive first.\n",
        "\n",
        "-->\n",
        "\n",
        "Our team achieved successful replication of the model by leveraging its open source codebase, resulting in metrics that are comparable to those showcased in the DeepMicro paper. In our endeavor, we thoroughly examined and addressed the claims below put forth in the original paper.\n",
        "\n",
        "1. **Dimensionality reduction engineering with traditional statistical techniques including Principal Component Analysis (PCA) and Gaussian Random Projection (GRP)**\n",
        "\n",
        "- Principal Component Analysis (PCA) aims to capture the most significant patterns and variations in a dataset by identifying orthogonal axes, known as principal components, that maximize the data variance.\n",
        "- Gaussian Random Projection (GRP) seeks to preserve pairwise distances between data points by projecting them onto a lower-dimensional space using random projections drawn from a Gaussian distribution.\n",
        "\n",
        "2. **Innovated representation learning employing with four different autoencoders including Shallow Autoencoder (SAE), Deep Autoencoder (DAE), Variational Autoencoder (VAE), and Convolutional Autoencoder (CAE)**\n",
        "\n",
        "- Shallow Autoencoder (SAE): This is the simplest form of an autoencoder, consisting of a fully\n",
        "connected encoder layer and a decoder layer. The latent representation is obtained from the encoder\n",
        "layer, which is a lower-dimensional space compared to the original input.\n",
        "- Deep Autoencoder (DAE): In addition to the encoder and decoder layers, DAE introduces hidden\n",
        "layers between the input and latent layers and between the latent and output layers. Rectified Linear\n",
        "Unit (ReLU) activation functions are used in the hidden layers.\n",
        "- Variational Autoencoder (VAE): VAE learns probabilistic representations by approximating the\n",
        "true posterior distribution of latent embeddings. It assumes that the posterior distribution follows\n",
        "a Gaussian distribution. VAE uses an encoder network to encode the means and variances of the\n",
        "Gaussian distribution and samples the latent representation from this distribution. The decoder\n",
        "network then reconstructs the input based on the sampled latent representation.\n",
        "- Convolutional Autoencoder (CAE): Instead of fully connected layers, CAE incorporates convo-\n",
        "lutional layers, where each unit is connected to local regions of the previous layer. Convolutional layers use filters (kernels) to perform convolution operations. CAE employs convolutional transpose layers (deconvolutional layers) to make the decoder symmetric to the encoder. No pooling layers\n",
        "are used in CAE.\n",
        "\n",
        "3. **Classification learning including including Support Vector Machine (SVM), Random Forest (RF), and Multi-Layer Perceptron (MLP)**\n",
        "- Support Vector Machine (SVM) is a supervised learning algorithm that aims to find an optimal hyperplane to classify data by maximizing the margin between different classes.\n",
        "- Random Forest (RF) is an ensemble learning method that constructs a multitude of decision trees and combines their predictions to make classifications.\n",
        "- Multi-Layer Perceptron (MLP) is a type of neural network that consists of multiple layers of interconnected nodes, enabling it to learn non-linear relationships and perform classification tasks.\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "xWAHJ_1CdtaA"
      },
      "source": [
        "# 4. Methodology\n",
        "\n",
        "<!-- This methodology is the core of your project. It consists of run-able codes with necessary annotations to show the expeiment you executed for testing the hypotheses.\n",
        "\n",
        "The methodology at least contains two subsections **data** and **model** in your experiment. -->"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "zPov0WCKWbPh"
      },
      "source": [
        "## 4.1. Environment"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "0dxJ28Zaccme"
      },
      "source": [
        "### 4.1.1. Python Version\n",
        "\n",
        "We use colab default python environment which is `python==3.10.12`."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "oTXyVLh8YFl7",
        "outputId": "9157c559-1b7c-488f-ba1c-81abe32682e2"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Python 3.11.6\n"
          ]
        }
      ],
      "source": [
        "# Python version\n",
        "!python --version"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "-IMSOXGXcYM9"
      },
      "source": [
        "### 4.1.2. Packages\n",
        "\n",
        "We utilize PyTorch to build our models so this project primarily depends on `sklearn` and `pytorch`."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "ExecuteTime": {
          "end_time": "2024-04-29T11:48:38.502933Z",
          "start_time": "2024-04-29T11:48:33.190976Z"
        },
        "id": "Lyf5Jfp7sGWw"
      },
      "outputs": [],
      "source": [
        "# import packages needed in this project\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "import time\n",
        "import datetime\n",
        "import matplotlib\n",
        "matplotlib.use('agg')\n",
        "import matplotlib.pyplot as plt\n",
        "\n",
        "from sklearn.neural_network import MLPClassifier\n",
        "from sklearn.model_selection import train_test_split\n",
        "from sklearn.decomposition import PCA\n",
        "from sklearn.random_projection import GaussianRandomProjection\n",
        "from sklearn.model_selection import GridSearchCV\n",
        "from sklearn.svm import SVC\n",
        "from sklearn.ensemble import RandomForestClassifier\n",
        "from sklearn.model_selection import StratifiedKFold\n",
        "from sklearn.metrics import roc_auc_score\n",
        "from sklearn.metrics import accuracy_score\n",
        "from sklearn.metrics import precision_score\n",
        "from sklearn.metrics import recall_score\n",
        "from sklearn.metrics import f1_score\n",
        "\n",
        "# importing PyTorch\n",
        "import torch\n",
        "import torch.nn as nn\n",
        "import torch.nn.functional as F\n",
        "from torch.utils.data import DataLoader, TensorDataset"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "2NbPHUTMbkD3"
      },
      "source": [
        "##  4.2. Data\n",
        "<!-- Data includes raw data (MIMIC III tables), descriptive statistics (our homework questions), and data processing (feature engineering).\n",
        "  * Source of the data: where the data is collected from; if data is synthetic or self-generated, explain how. If possible, please provide a link to the raw datasets.\n",
        "  * Statistics: include basic descriptive statistics of the dataset like size, cross validation split, label distribution, etc.\n",
        "  * Data process: how do you munipulate the data, e.g., change the class labels, split the dataset to train/valid/test, refining the dataset.\n",
        "  * Illustration: printing results, plotting figures for illustration.\n",
        "  * You can upload your raw dataset to Google Drive and mount this Colab to the same directory. If your raw dataset is too large, you can upload the processed dataset and have a code to load the processed dataset. -->"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "SRRY_dsgYL2Q"
      },
      "source": [
        "### 4.2.1 Data Download Instruction\n",
        "\n",
        "We store two datasets, abundance and marker from the [DeepMicro codebase](https://github.com/minoh0201/DeepMicro/tree/master/data) in the path `raw_data_dir` of the [colab project](https://drive.google.com/file/d/1c6ZPVp346e1wQH6opg4YwlCRjRgfTXCk/view?usp=drive_link).\n",
        "\n",
        "When we run this project locally, we use the function `fetch_data_from_remote` to download the datasets from the google drive to current working directory."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {
        "ExecuteTime": {
          "end_time": "2024-04-29T11:48:47.979130Z",
          "start_time": "2024-04-29T11:48:47.975250Z"
        },
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "yGCbNWqfcRcs",
        "outputId": "8195b56b-26b9-4196-edea-66d9f93c71b4"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Data already exists locally\n",
            "abundance_Cirrhosis.csv  abundance_IBD.txt        abundance_WT2D.txt\n",
            "abundance_Cirrhosis.txt  abundance_Obesity.txt\n",
            "abundance_Colorectal.txt abundance_T2D.txt\n",
            "marker_Cirrhosis.txt  marker_IBD.txt        marker_T2D.txt\n",
            "marker_Colorectal.txt marker_Obesity.txt    marker_WT2D.txt\n"
          ]
        }
      ],
      "source": [
        "# data dir\n",
        "raw_data_dir = 'data/'\n",
        "\n",
        "# Load data\n",
        "import os\n",
        "import requests\n",
        "from zipfile import ZipFile\n",
        "from io import BytesIO\n",
        "\n",
        "def fetch_data_from_remote():\n",
        "    # cloud location where the .zip file of the data used by this project is stored\n",
        "    url = 'https://drive.google.com/file/d/1c6ZPVp346e1wQH6opg4YwlCRjRgfTXCk/view?usp=drive_link'\n",
        "\n",
        "    file_id = url.split('/')[-2]\n",
        "    dwn_url = 'https://drive.google.com/uc?export=download&id=' + file_id\n",
        "\n",
        "    # unzip the data into the local working directory\n",
        "    response = requests.get(dwn_url)\n",
        "    zip_content = BytesIO(response.content)\n",
        "    with ZipFile(zip_content, 'r') as main_zip:\n",
        "        main_zip.extractall()\n",
        "        with main_zip.open('data/abundance.zip') as nested_zip_abundance:\n",
        "            with ZipFile(nested_zip_abundance) as zip_ref_abundance:\n",
        "                zip_ref_abundance.extractall('data/abundance')\n",
        "\n",
        "        with main_zip.open('data/marker.zip') as nested_zip_marker:\n",
        "            with ZipFile(nested_zip_marker) as zip_ref_marker:\n",
        "                zip_ref_marker.extractall('data/marker')\n",
        "\n",
        "if os.path.exists('data'):\n",
        "    print(\"Data already exists locally\")\n",
        "else:\n",
        "    fetch_data_from_remote()\n",
        "    print(\"Data fetched and unzipped successfully\")\n",
        "\n",
        "# show that the data is available locally now\n",
        "!ls data/abundance/\n",
        "!ls data/marker/"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "FJNbhZZ7cTjS"
      },
      "source": [
        "### 4.2.2. Data Description\n",
        "\n",
        "Our reproductivity utilizes the same datasets as the original paper, which include six disease (Table 1). They are inflammatory bowel disease (IBD), type 2 diabetes in European women (EW-T2D), type 2 diabetes in Chinese (C-T2D), obesity (Obesity), liver cirrhosis (Cirrhosis), and colorectal cancer (Colorectal).\n",
        "\n",
        "![](assets/Data_Table1.png)\n",
        "\n",
        "In each dataset, marker profile and abundance profile of microbiome are used to train our models (Table 2).\n",
        "\n",
        "![](assets/Data_Table2.png)\n",
        "\n",
        "All the data are stored in txt format and the data path structure is as following.\n",
        "\n",
        "```\n",
        ".\n",
        "├── ...\n",
        "├── data                              # Data folder\n",
        "│   ├── marker                        # Marker profile data\n",
        "│       ├── marker_IBD.txt            # Inflammatory bowel disease (IBD)\n",
        "│       ├── marker_WT2D.txt           # Type 2 diabetes in European women (EW-T2D)\n",
        "│       ├── marker_T2D.txt            # Type 2 diabetes in Chinese (C-T2D)\n",
        "│       ├── marker_Obesity.txt        # Obesity (Obesity)\n",
        "│       ├── marker_Cirrhosis.txt      # Liver cirrhosis (Cirrhosis)\n",
        "│       ├── marker_Colorectal.txt     # Colorectal cancer (Colorectal)\n",
        "│   ├── abundance                     # Abundance profile data\n",
        "│       ├── abundance_IBD.txt         # Inflammatory bowel disease (IBD)\n",
        "│       ├── abundance_WT2D.txt        # Type 2 diabetes in European women (EW-T2D)\n",
        "│       ├── abundance_T2D.txt         # Type 2 diabetes in Chinese (C-T2D)\n",
        "│       ├── abundance_Obesity.txt     # Obesity (Obesity)\n",
        "│       ├── abundance_Cirrhosis.txt   # Liver cirrhosis (Cirrhosis)\n",
        "│       ├── abundance_Colorectal.txt  # Colorectal cancer (Colorectal)\n",
        "└── ...\n",
        "```\n",
        "In each txt file, it has different number of features and data points as shown in Table 1 and Table 2 above. In the sample demonstration and visualizations below. The colorectal cancer of microbiome abundance profile has 503 features (exluding those dummy labels) and 121 entries.\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {
        "ExecuteTime": {
          "end_time": "2024-04-29T11:51:20.979319Z",
          "start_time": "2024-04-29T11:51:20.828998Z"
        },
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 586
        },
        "id": "BYGSgWXzjdeX",
        "outputId": "15c00de3-bd3f-4875-9eff-74e85e55b89d"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>dataset_name</th>\n",
              "      <th>sampleID</th>\n",
              "      <th>subjectID</th>\n",
              "      <th>bodysite</th>\n",
              "      <th>disease</th>\n",
              "      <th>age</th>\n",
              "      <th>gender</th>\n",
              "      <th>country</th>\n",
              "      <th>sequencing_technology</th>\n",
              "      <th>pubmedid</th>\n",
              "      <th>...</th>\n",
              "      <th>k__Eukaryota|p__Ascomycota|c__Saccharomycetes|o__Saccharomycetales|f__Saccharomycetaceae|g__Eremothecium|s__Eremothecium_unclassified</th>\n",
              "      <th>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Lactobacillaceae|g__Lactobacillus|s__Lactobacillus_antri</th>\n",
              "      <th>k__Bacteria|p__Firmicutes|c__Bacilli|o__Bacillales|f__Bacillaceae|g__Lysinibacillus|s__Lysinibacillus_fusiformis</th>\n",
              "      <th>k__Archaea|p__Euryarchaeota|c__Methanobacteria|o__Methanobacteriales|f__Methanobacteriaceae|g__Methanobacterium|s__Methanobacterium_unclassified</th>\n",
              "      <th>k__Bacteria|p__Firmicutes|c__Bacilli|o__Bacillales|f__Bacillaceae|g__Lysinibacillus|s__Lysinibacillus_boronitolerans</th>\n",
              "      <th>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Enterococcaceae|g__Bavariicoccus|s__Bavariicoccus_seileri</th>\n",
              "      <th>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Enterococcaceae|g__Enterococcus|s__Enterococcus_gilvus</th>\n",
              "      <th>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Lactobacillaceae|g__Lactobacillus|s__Lactobacillus_otakiensis</th>\n",
              "      <th>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Peptococcaceae|g__Desulfotomaculum|s__Desulfotomaculum_ruminis</th>\n",
              "      <th>k__Bacteria|p__Firmicutes|c__Negativicutes|o__Selenomonadales|f__Veillonellaceae|g__Megasphaera|s__Megasphaera_sp_BV3C16_1</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Zeller_fecal_colorectal_cancer</td>\n",
              "      <td>CCIS00146684ST-4-0</td>\n",
              "      <td>fr-726</td>\n",
              "      <td>stool</td>\n",
              "      <td>n</td>\n",
              "      <td>72</td>\n",
              "      <td>female</td>\n",
              "      <td>france</td>\n",
              "      <td>Illumina</td>\n",
              "      <td>25432777</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>Zeller_fecal_colorectal_cancer</td>\n",
              "      <td>CCIS00281083ST-3-0</td>\n",
              "      <td>fr-060</td>\n",
              "      <td>stool</td>\n",
              "      <td>n</td>\n",
              "      <td>53</td>\n",
              "      <td>male</td>\n",
              "      <td>france</td>\n",
              "      <td>Illumina</td>\n",
              "      <td>25432777</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>Zeller_fecal_colorectal_cancer</td>\n",
              "      <td>CCIS02124300ST-4-0</td>\n",
              "      <td>fr-568</td>\n",
              "      <td>stool</td>\n",
              "      <td>n</td>\n",
              "      <td>35</td>\n",
              "      <td>male</td>\n",
              "      <td>france</td>\n",
              "      <td>Illumina</td>\n",
              "      <td>25432777</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>Zeller_fecal_colorectal_cancer</td>\n",
              "      <td>CCIS02379307ST-4-0</td>\n",
              "      <td>fr-828</td>\n",
              "      <td>stool</td>\n",
              "      <td>cancer</td>\n",
              "      <td>67</td>\n",
              "      <td>male</td>\n",
              "      <td>france</td>\n",
              "      <td>Illumina</td>\n",
              "      <td>25432777</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>Zeller_fecal_colorectal_cancer</td>\n",
              "      <td>CCIS03473770ST-4-0</td>\n",
              "      <td>fr-192</td>\n",
              "      <td>stool</td>\n",
              "      <td>n</td>\n",
              "      <td>29</td>\n",
              "      <td>male</td>\n",
              "      <td>france</td>\n",
              "      <td>Illumina</td>\n",
              "      <td>25432777</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0.03756</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>Zeller_fecal_colorectal_cancer</td>\n",
              "      <td>CCIS06260551ST-3-0</td>\n",
              "      <td>fr-200</td>\n",
              "      <td>stool</td>\n",
              "      <td>cancer</td>\n",
              "      <td>58</td>\n",
              "      <td>male</td>\n",
              "      <td>france</td>\n",
              "      <td>Illumina</td>\n",
              "      <td>25432777</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0.31121</td>\n",
              "      <td>0</td>\n",
              "      <td>0.03562</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>Zeller_fecal_colorectal_cancer</td>\n",
              "      <td>CCIS07539127ST-4-0</td>\n",
              "      <td>fr-460</td>\n",
              "      <td>stool</td>\n",
              "      <td>n</td>\n",
              "      <td>77</td>\n",
              "      <td>female</td>\n",
              "      <td>france</td>\n",
              "      <td>Illumina</td>\n",
              "      <td>25432777</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>Zeller_fecal_colorectal_cancer</td>\n",
              "      <td>CCIS07648107ST-4-0</td>\n",
              "      <td>fr-053</td>\n",
              "      <td>stool</td>\n",
              "      <td>n</td>\n",
              "      <td>62</td>\n",
              "      <td>female</td>\n",
              "      <td>france</td>\n",
              "      <td>Illumina</td>\n",
              "      <td>25432777</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>Zeller_fecal_colorectal_cancer</td>\n",
              "      <td>CCIS08668806ST-3-0</td>\n",
              "      <td>fr-214</td>\n",
              "      <td>stool</td>\n",
              "      <td>small_adenoma</td>\n",
              "      <td>63</td>\n",
              "      <td>male</td>\n",
              "      <td>france</td>\n",
              "      <td>Illumina</td>\n",
              "      <td>25432777</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>Zeller_fecal_colorectal_cancer</td>\n",
              "      <td>CCIS09568613ST-4-0</td>\n",
              "      <td>fr-400</td>\n",
              "      <td>stool</td>\n",
              "      <td>n</td>\n",
              "      <td>67</td>\n",
              "      <td>male</td>\n",
              "      <td>france</td>\n",
              "      <td>Illumina</td>\n",
              "      <td>25432777</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>10 rows × 714 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "0                     dataset_name            sampleID subjectID bodysite  \\\n",
              "1   Zeller_fecal_colorectal_cancer  CCIS00146684ST-4-0    fr-726    stool   \n",
              "2   Zeller_fecal_colorectal_cancer  CCIS00281083ST-3-0    fr-060    stool   \n",
              "3   Zeller_fecal_colorectal_cancer  CCIS02124300ST-4-0    fr-568    stool   \n",
              "4   Zeller_fecal_colorectal_cancer  CCIS02379307ST-4-0    fr-828    stool   \n",
              "5   Zeller_fecal_colorectal_cancer  CCIS03473770ST-4-0    fr-192    stool   \n",
              "6   Zeller_fecal_colorectal_cancer  CCIS06260551ST-3-0    fr-200    stool   \n",
              "7   Zeller_fecal_colorectal_cancer  CCIS07539127ST-4-0    fr-460    stool   \n",
              "8   Zeller_fecal_colorectal_cancer  CCIS07648107ST-4-0    fr-053    stool   \n",
              "9   Zeller_fecal_colorectal_cancer  CCIS08668806ST-3-0    fr-214    stool   \n",
              "10  Zeller_fecal_colorectal_cancer  CCIS09568613ST-4-0    fr-400    stool   \n",
              "\n",
              "0         disease age  gender country sequencing_technology  pubmedid  ...  \\\n",
              "1               n  72  female  france              Illumina  25432777  ...   \n",
              "2               n  53    male  france              Illumina  25432777  ...   \n",
              "3               n  35    male  france              Illumina  25432777  ...   \n",
              "4          cancer  67    male  france              Illumina  25432777  ...   \n",
              "5               n  29    male  france              Illumina  25432777  ...   \n",
              "6          cancer  58    male  france              Illumina  25432777  ...   \n",
              "7               n  77  female  france              Illumina  25432777  ...   \n",
              "8               n  62  female  france              Illumina  25432777  ...   \n",
              "9   small_adenoma  63    male  france              Illumina  25432777  ...   \n",
              "10              n  67    male  france              Illumina  25432777  ...   \n",
              "\n",
              "0  k__Eukaryota|p__Ascomycota|c__Saccharomycetes|o__Saccharomycetales|f__Saccharomycetaceae|g__Eremothecium|s__Eremothecium_unclassified  \\\n",
              "1                                                   0                                                                                      \n",
              "2                                                   0                                                                                      \n",
              "3                                                   0                                                                                      \n",
              "4                                                   0                                                                                      \n",
              "5                                                   0                                                                                      \n",
              "6                                                   0                                                                                      \n",
              "7                                                   0                                                                                      \n",
              "8                                                   0                                                                                      \n",
              "9                                                   0                                                                                      \n",
              "10                                                  0                                                                                      \n",
              "\n",
              "0  k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Lactobacillaceae|g__Lactobacillus|s__Lactobacillus_antri  \\\n",
              "1                                                   0                                                                    \n",
              "2                                                   0                                                                    \n",
              "3                                                   0                                                                    \n",
              "4                                                   0                                                                    \n",
              "5                                                   0                                                                    \n",
              "6                                                   0                                                                    \n",
              "7                                                   0                                                                    \n",
              "8                                                   0                                                                    \n",
              "9                                                   0                                                                    \n",
              "10                                                  0                                                                    \n",
              "\n",
              "0  k__Bacteria|p__Firmicutes|c__Bacilli|o__Bacillales|f__Bacillaceae|g__Lysinibacillus|s__Lysinibacillus_fusiformis  \\\n",
              "1                                                   0                                                                 \n",
              "2                                                   0                                                                 \n",
              "3                                                   0                                                                 \n",
              "4                                                   0                                                                 \n",
              "5                                                   0                                                                 \n",
              "6                                             0.31121                                                                 \n",
              "7                                                   0                                                                 \n",
              "8                                                   0                                                                 \n",
              "9                                                   0                                                                 \n",
              "10                                                  0                                                                 \n",
              "\n",
              "0  k__Archaea|p__Euryarchaeota|c__Methanobacteria|o__Methanobacteriales|f__Methanobacteriaceae|g__Methanobacterium|s__Methanobacterium_unclassified  \\\n",
              "1                                                   0                                                                                                 \n",
              "2                                                   0                                                                                                 \n",
              "3                                                   0                                                                                                 \n",
              "4                                                   0                                                                                                 \n",
              "5                                                   0                                                                                                 \n",
              "6                                                   0                                                                                                 \n",
              "7                                                   0                                                                                                 \n",
              "8                                                   0                                                                                                 \n",
              "9                                                   0                                                                                                 \n",
              "10                                                  0                                                                                                 \n",
              "\n",
              "0  k__Bacteria|p__Firmicutes|c__Bacilli|o__Bacillales|f__Bacillaceae|g__Lysinibacillus|s__Lysinibacillus_boronitolerans  \\\n",
              "1                                                   0                                                                     \n",
              "2                                                   0                                                                     \n",
              "3                                                   0                                                                     \n",
              "4                                                   0                                                                     \n",
              "5                                                   0                                                                     \n",
              "6                                             0.03562                                                                     \n",
              "7                                                   0                                                                     \n",
              "8                                                   0                                                                     \n",
              "9                                                   0                                                                     \n",
              "10                                                  0                                                                     \n",
              "\n",
              "0  k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Enterococcaceae|g__Bavariicoccus|s__Bavariicoccus_seileri  \\\n",
              "1                                                   0                                                                     \n",
              "2                                                   0                                                                     \n",
              "3                                                   0                                                                     \n",
              "4                                                   0                                                                     \n",
              "5                                                   0                                                                     \n",
              "6                                                   0                                                                     \n",
              "7                                                   0                                                                     \n",
              "8                                                   0                                                                     \n",
              "9                                                   0                                                                     \n",
              "10                                                  0                                                                     \n",
              "\n",
              "0  k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Enterococcaceae|g__Enterococcus|s__Enterococcus_gilvus  \\\n",
              "1                                                   0                                                                  \n",
              "2                                                   0                                                                  \n",
              "3                                                   0                                                                  \n",
              "4                                                   0                                                                  \n",
              "5                                             0.03756                                                                  \n",
              "6                                                   0                                                                  \n",
              "7                                                   0                                                                  \n",
              "8                                                   0                                                                  \n",
              "9                                                   0                                                                  \n",
              "10                                                  0                                                                  \n",
              "\n",
              "0  k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Lactobacillaceae|g__Lactobacillus|s__Lactobacillus_otakiensis  \\\n",
              "1                                                   0                                                                         \n",
              "2                                                   0                                                                         \n",
              "3                                                   0                                                                         \n",
              "4                                                   0                                                                         \n",
              "5                                                   0                                                                         \n",
              "6                                                   0                                                                         \n",
              "7                                                   0                                                                         \n",
              "8                                                   0                                                                         \n",
              "9                                                   0                                                                         \n",
              "10                                                  0                                                                         \n",
              "\n",
              "0  k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Peptococcaceae|g__Desulfotomaculum|s__Desulfotomaculum_ruminis  \\\n",
              "1                                                   0                                                                           \n",
              "2                                                   0                                                                           \n",
              "3                                                   0                                                                           \n",
              "4                                                   0                                                                           \n",
              "5                                                   0                                                                           \n",
              "6                                                   0                                                                           \n",
              "7                                                   0                                                                           \n",
              "8                                                   0                                                                           \n",
              "9                                                   0                                                                           \n",
              "10                                                  0                                                                           \n",
              "\n",
              "0  k__Bacteria|p__Firmicutes|c__Negativicutes|o__Selenomonadales|f__Veillonellaceae|g__Megasphaera|s__Megasphaera_sp_BV3C16_1  \n",
              "1                                                   0                                                                          \n",
              "2                                                   0                                                                          \n",
              "3                                                   0                                                                          \n",
              "4                                                   0                                                                          \n",
              "5                                                   0                                                                          \n",
              "6                                                   0                                                                          \n",
              "7                                                   0                                                                          \n",
              "8                                                   0                                                                          \n",
              "9                                                   0                                                                          \n",
              "10                                                  0                                                                          \n",
              "\n",
              "[10 rows x 714 columns]"
            ]
          },
          "execution_count": 4,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Extract a sample data of colorectal cancer of microbiome abundance profile\n",
        "sample = pd.read_csv(os.path.join(raw_data_dir, \"abundance\", \"abundance_Colorectal.txt\"), sep='\\t', index_col=0, header=None)\n",
        "sample = sample.T\n",
        "sample.head(10)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "5EjqDkrWm0lc",
        "outputId": "15340101-fc13-4b6e-a015-d39312ebdff0"
      },
      "outputs": [
        {
          "data": {
            "text/plain": [
              "503"
            ]
          },
          "execution_count": 5,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Count the features excluding patients information\n",
        "len([ _ for _ in sample.columns if '|' in _ ])"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "i93VorpFlPnl",
        "outputId": "ed802f7e-6598-4294-95c9-5e25a3da9437"
      },
      "outputs": [
        {
          "data": {
            "text/plain": [
              "(121, 714)"
            ]
          },
          "execution_count": 6,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Calculate the dimensions of the sameple dataset\n",
        "sample.shape"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 475
        },
        "id": "PPFi9QXolIoa",
        "outputId": "de7698a4-45fe-4468-cf36-d6ed72ed1b02"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>count</th>\n",
              "      <th>unique</th>\n",
              "      <th>top</th>\n",
              "      <th>freq</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>dataset_name</th>\n",
              "      <td>121</td>\n",
              "      <td>1</td>\n",
              "      <td>Zeller_fecal_colorectal_cancer</td>\n",
              "      <td>121</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>sampleID</th>\n",
              "      <td>121</td>\n",
              "      <td>121</td>\n",
              "      <td>CCIS00146684ST-4-0</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>subjectID</th>\n",
              "      <td>121</td>\n",
              "      <td>121</td>\n",
              "      <td>fr-726</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>bodysite</th>\n",
              "      <td>121</td>\n",
              "      <td>1</td>\n",
              "      <td>stool</td>\n",
              "      <td>121</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>disease</th>\n",
              "      <td>121</td>\n",
              "      <td>3</td>\n",
              "      <td>cancer</td>\n",
              "      <td>48</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Enterococcaceae|g__Bavariicoccus|s__Bavariicoccus_seileri</th>\n",
              "      <td>121</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>120</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Enterococcaceae|g__Enterococcus|s__Enterococcus_gilvus</th>\n",
              "      <td>121</td>\n",
              "      <td>5</td>\n",
              "      <td>0</td>\n",
              "      <td>117</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Lactobacillaceae|g__Lactobacillus|s__Lactobacillus_otakiensis</th>\n",
              "      <td>121</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>120</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Peptococcaceae|g__Desulfotomaculum|s__Desulfotomaculum_ruminis</th>\n",
              "      <td>121</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>120</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>k__Bacteria|p__Firmicutes|c__Negativicutes|o__Selenomonadales|f__Veillonellaceae|g__Megasphaera|s__Megasphaera_sp_BV3C16_1</th>\n",
              "      <td>121</td>\n",
              "      <td>2</td>\n",
              "      <td>0</td>\n",
              "      <td>120</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>714 rows × 4 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "                                                   count unique  \\\n",
              "0                                                                 \n",
              "dataset_name                                         121      1   \n",
              "sampleID                                             121    121   \n",
              "subjectID                                            121    121   \n",
              "bodysite                                             121      1   \n",
              "disease                                              121      3   \n",
              "...                                                  ...    ...   \n",
              "k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactoba...   121      2   \n",
              "k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactoba...   121      5   \n",
              "k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactoba...   121      2   \n",
              "k__Bacteria|p__Firmicutes|c__Clostridia|o__Clos...   121      2   \n",
              "k__Bacteria|p__Firmicutes|c__Negativicutes|o__S...   121      2   \n",
              "\n",
              "                                                                               top  \\\n",
              "0                                                                                    \n",
              "dataset_name                                        Zeller_fecal_colorectal_cancer   \n",
              "sampleID                                                        CCIS00146684ST-4-0   \n",
              "subjectID                                                                   fr-726   \n",
              "bodysite                                                                     stool   \n",
              "disease                                                                     cancer   \n",
              "...                                                                            ...   \n",
              "k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactoba...                               0   \n",
              "k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactoba...                               0   \n",
              "k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactoba...                               0   \n",
              "k__Bacteria|p__Firmicutes|c__Clostridia|o__Clos...                               0   \n",
              "k__Bacteria|p__Firmicutes|c__Negativicutes|o__S...                               0   \n",
              "\n",
              "                                                   freq  \n",
              "0                                                        \n",
              "dataset_name                                        121  \n",
              "sampleID                                              1  \n",
              "subjectID                                             1  \n",
              "bodysite                                            121  \n",
              "disease                                              48  \n",
              "...                                                 ...  \n",
              "k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactoba...  120  \n",
              "k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactoba...  117  \n",
              "k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactoba...  120  \n",
              "k__Bacteria|p__Firmicutes|c__Clostridia|o__Clos...  120  \n",
              "k__Bacteria|p__Firmicutes|c__Negativicutes|o__S...  120  \n",
              "\n",
              "[714 rows x 4 columns]"
            ]
          },
          "execution_count": 7,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Have a generate descrption of the sample dataset\n",
        "sample.describe().T"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "QTkZz-HrofL1"
      },
      "source": [
        "### 4.2.3. Data Preprocess\n",
        "\n",
        "According to the original paper, the original data provided in the codebase have been preprocessed and cleaned properly. However, we still need to identify the features columns and labels to extract `X` and `Y`, respectively.\n",
        "\n",
        "In abundance data, we extradct features by `k__`. In marker data, the features are chracterized with `gi|`.\n",
        "\n",
        "For labels of different deseases, we created an dictionary `label_dict` to make the labels into binary numbers (i.e. either 1 or 0)."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {
        "id": "acW06e9MbqgM"
      },
      "outputs": [],
      "source": [
        "# define random seed\n",
        "seed = 42\n",
        "\n",
        "label_dict = {\n",
        "    # Controls\n",
        "    'n': 0,\n",
        "    # Chirrhosis\n",
        "    'cirrhosis': 1,\n",
        "    # Colorectal Cancer\n",
        "    'cancer': 1, 'small_adenoma': 0,\n",
        "    # IBD\n",
        "    'ibd_ulcerative_colitis': 1, 'ibd_crohn_disease': 1,\n",
        "    # T2D and WT2D\n",
        "    't2d': 1,\n",
        "    # Obesity\n",
        "    'leaness': 0, 'obesity': 1,\n",
        "}\n",
        "\n",
        "\n",
        "def load_data(data_dir, filename, dtype=None):\n",
        "    feature_string = ''\n",
        "    if filename.split('_')[0] == 'abundance':\n",
        "        feature_string = \"k__\"\n",
        "    if filename.split('_')[0] == 'marker':\n",
        "        feature_string = \"gi|\"\n",
        "    # read file\n",
        "    filename = data_dir + filename\n",
        "    if not os.path.isfile(filename):\n",
        "        print(\"FileNotFoundError: File {} does not exist\".format(filename))\n",
        "        exit()\n",
        "    raw = pd.read_csv(filename, sep='\\t', index_col=0, header=None)\n",
        "\n",
        "    # select rows having feature index identifier string\n",
        "    X = raw.loc[raw.index.str.contains(feature_string, regex=False)].T\n",
        "\n",
        "    # get class labels\n",
        "    Y = raw.loc['disease']\n",
        "    Y = Y.replace(label_dict)\n",
        "\n",
        "    # train and test split\n",
        "    X_train, X_test, y_train, y_test = train_test_split(\n",
        "        X.values.astype(dtype), Y.values.astype('int'), test_size=0.2,\n",
        "        random_state=seed, stratify=Y.values\n",
        "        )\n",
        "    print(\"Train data shape: \", X_train.shape)\n",
        "    print(\"Test data shape: \", X_test.shape)\n",
        "    return X_train, X_test, y_train, y_test"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "GXSpN9sMbqgN"
      },
      "source": [
        "##   4.3. Model\n",
        "\n",
        "### 4.3.1. Citation to the Original Paper\n",
        "\n",
        "Oh, M., & Zhang, L. (2020). DeepMicro: deep representation learning for disease prediction based on microbiome data. Scientific reports, 10(1), 6026.\n",
        "\n",
        "### 4.3.2. Link to the Original Paper's Repo\n",
        "[DeepMicro Github Repository](https://github.com/minoh0201/DeepMicro)\n",
        "\n",
        "### 4.3.3. Model Descriptions\n",
        "\n",
        "An autoencoder represents a type of neural network designed for the purpose of reconstructing its input data, denoted as $x$. In its fundamental structure, it comprises an encoder function, denoted as $f_\\phi(⋅)$, and a decoder function, denoted as $f_\\theta′(⋅)$, with $\\phi$ and $\\theta$ serving as the parameters associated with the encoder and decoder functions, respectively.\n",
        "The training objective of an autoencoder is to minimize the disparity between the original input $x$ and its reconstructed counterpart $x′$. This discrepancy, typically quantified using a reconstruction loss metric such as squared error, can be mathematically expressed as $L(x, x')=||x-x'||^2=||x-f_\\theta'(f_\\phi(x))||^2$.\n",
        "\n",
        "In this project, we focus on the utilization of a trained autoencoder to obtain a lower-dimensional latent representation $z = f_\\phi(x)$ of the input. There are four autoencoders that we incorporate as below.\n",
        "\n",
        "1. Shallow Autoencoder (SAE): a fully connected encoder connecting the input layer to the latent layer, and a decoder producing the reconstructed input $x′$ by combining the outputs of the latent layer using weighted sums, with both the latent and output layers utilizing a linear activation function.\n",
        "\n",
        "2. Deep Autoencoder (DAE): enhanced SAE model by inserting hidden layers with Rectified Linear Unit (ReLu) activation function and Glorot uniform initializer between the input and latent layers, maintaining an equal number of hidden layers (either one or two layers) in both the encoder and decoder sections.\n",
        "\n",
        "3. Variational autoencoder (VAE): it learns probabilistic representations $z$ by approximating the true posterior distribution with $q_\\phi(z|x)$ assuming a Gaussian distribution. The encoder encodes the means and variances of the Gaussian distribution, allowing sampling of latent representation $z$. This sampled representation is then fed into the decoder network to generate the reconstructed input $x′ \\sim g_\\theta(x|z)$.\n",
        "\n",
        "4. convolutional autoencoder (CAE): equipped with convolutional layers where each unit is connected locally to the previous layer. The layers consist of multiple filters with weights for convolution operations. We employed ReLu activation, Glorot uniform initializer, and avoided pooling layers to prevent excessive generalization. The $n$-dimensional input vector was reshaped into a squared image of size $d \\times d \\times 1$, where $d = ⌊ \\sqrt{n} ⌋ + 1$.\n",
        "\n",
        "After getting the low-dimentional representations of profiles, we build classification models.\n",
        "\n",
        "1. Support Vector Machine (SVM): a grid search using SVM is conducted to explore the hyper-parameter space. We considered both radial basis function (RBF) and linear kernels, adjusting penalty parameter C and kernel coefficient gamma for RBF.\n",
        "\n",
        "2. Random Forest (RF): We examine two criteria, Gini impurity and information gain, for selecting features to split a node in a decision tree. The maximum number of features considered for the best split at each node was determined using the square root of the sample size and the logarithm to base 2 of the sample size.\n",
        "\n",
        "3. Multi-Layer Perceptron (MLP): ReLu activations were used for the hidden layers, while the output layer employed sigmoid activation with a single unit. The number of units in the hidden layers was set to half of the preceding layer, excluding the first hidden layer.\n",
        "\n",
        "The overall workflow is demonstrated in the Figure 1 below.\n",
        "\n",
        "![](assets/Model_Figure1.png)\n",
        "\n",
        "### 4.3.4. Implementation Code\n",
        "\n",
        "Please find the implementations of autoencoders in the section [Autoencoders Definitions](#autoencoders-definitions) below."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "c3XEF8AmtwSI"
      },
      "source": [
        "#### Autoencoders Definitions\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {
        "id": "pWkNlD2ORZoG"
      },
      "outputs": [],
      "source": [
        "class EarlyStopper:\n",
        "    def __init__(self, patience=5, delta=0.0):\n",
        "        self.patience = patience\n",
        "        self.delta = delta\n",
        "        self.counter = 0\n",
        "        self.best_loss = None\n",
        "        self.early_stop = False\n",
        "\n",
        "    def __call__(self, loss):\n",
        "        if self.best_loss is None:\n",
        "            self.best_loss = loss\n",
        "        elif loss > self.best_loss + self.delta:\n",
        "            self.counter += 1\n",
        "            if self.counter >= self.patience:\n",
        "                self.early_stop = True\n",
        "        else:\n",
        "            self.best_loss = loss\n",
        "            self.counter = 0\n",
        "\n",
        "        return self.early_stop\n",
        "\n",
        "\n",
        "class ShallowAutoEncoder(nn.Module):\n",
        "    def __init__(self, input_dim=28 * 28):\n",
        "        super(ShallowAutoEncoder, self).__init__()\n",
        "        self.input_dim = input_dim\n",
        "        self.encoder1 = nn.Linear(input_dim, 32)\n",
        "        self.decoder1 = nn.Linear(32, input_dim)\n",
        "\n",
        "    def encoder(self, x):\n",
        "        return torch.tanh(self.encoder1(x))\n",
        "\n",
        "    def decoder(self, z):\n",
        "        return torch.sigmoid(self.decoder1(z))\n",
        "\n",
        "    def forward(self, x):\n",
        "        z = self.encoder(x.view(-1, self.input_dim))\n",
        "        return self.decoder(z)\n",
        "\n",
        "    @staticmethod\n",
        "    def loss_func(x_hat, x):\n",
        "        return F.mse_loss(x_hat, x, reduction='sum')\n",
        "\n",
        "\n",
        "class AutoEncoder(nn.Module):\n",
        "    def __init__(self, input_dim=28*28):\n",
        "        super(AutoEncoder, self).__init__()\n",
        "        self.input_dim = input_dim\n",
        "\n",
        "        # DNN as encoder\n",
        "        # 512-256-128\n",
        "        self.encoder1 = nn.Linear(input_dim, 512)\n",
        "        self.encoder2 = nn.Linear(512, 256)\n",
        "        self.encoder3 = nn.Linear(256, 128)\n",
        "\n",
        "        # DNN as decoder\n",
        "        self.decoder1 = nn.Linear(128, 256)\n",
        "        self.decoder2 = nn.Linear(256, 512)\n",
        "        self.decoder3 = nn.Linear(512, input_dim)\n",
        "\n",
        "    def encoder(self, x):\n",
        "        h = torch.tanh(self.encoder1(x))\n",
        "        h = torch.tanh(self.encoder2(h))\n",
        "        return torch.tanh(self.encoder3(h))\n",
        "\n",
        "    def decoder(self, z):\n",
        "        h = torch.tanh(self.decoder1(z))\n",
        "        h = torch.tanh(self.decoder2(h))\n",
        "        return torch.sigmoid(self.decoder3(h))\n",
        "\n",
        "    def forward(self, x):\n",
        "        z = self.encoder(x.view(-1, self.input_dim))\n",
        "        return self.decoder(z)\n",
        "\n",
        "    @staticmethod\n",
        "    def loss_func(x_hat, x):\n",
        "        return F.mse_loss(x_hat, x)\n",
        "\n",
        "\n",
        "class CAE(nn.Module):\n",
        "    def __init__(self):\n",
        "        super(CAE, self).__init__()\n",
        "\n",
        "        # CNN as encoder\n",
        "        self.encoder = nn.Sequential(\n",
        "            nn.Conv2d(1, 16, 3, stride=2, padding=1),\n",
        "            nn.ReLU(),\n",
        "            nn.Conv2d(16, 32, 3, stride=2, padding=1),\n",
        "            nn.ReLU(),\n",
        "            nn.Conv2d(32, 64, 6)\n",
        "        )\n",
        "\n",
        "        # CNN as decoder\n",
        "        self.decoder = nn.Sequential(\n",
        "            nn.ConvTranspose2d(64, 32, 6),\n",
        "            nn.ReLU(),\n",
        "            nn.ConvTranspose2d(32, 16, 3, stride=2, padding=1, output_padding=1),\n",
        "            nn.ReLU(),\n",
        "            nn.ConvTranspose2d(16, 1, 3, stride=2, padding=1, output_padding=1),\n",
        "            nn.Sigmoid()\n",
        "        )\n",
        "\n",
        "    def forward(self, x):\n",
        "        z = self.encoder(x)\n",
        "        return self.decoder(z)\n",
        "\n",
        "    @staticmethod\n",
        "    def loss_func(x_hat, x):\n",
        "        return F.mse_loss(x_hat, x)\n",
        "\n",
        "\n",
        "class VAE(nn.Module):\n",
        "    def __init__(self, input_dim=28 * 28, latent_dim=8):\n",
        "        super(VAE, self).__init__()\n",
        "        self.input_dim = input_dim\n",
        "\n",
        "        # DNN as encoder\n",
        "        self.encoder1 = nn.Linear(input_dim, 128)\n",
        "        self.mu = nn.Linear(128, 8)\n",
        "        self.log_sigma2 = nn.Linear(128, 8)\n",
        "\n",
        "        # DNN as decoder\n",
        "        self.decoder1 = nn.Linear(8, 128)\n",
        "        self.decoder2 = nn.Linear(128, input_dim)\n",
        "\n",
        "    def encoder(self, x):\n",
        "        h = torch.tanh(self.encoder1(x))\n",
        "        return self.mu(h), torch.sqrt(torch.exp(self.log_sigma2(h)))\n",
        "\n",
        "    @staticmethod\n",
        "    def sampling(mu, std):  # Re parameterization trick\n",
        "        eps = torch.randn_like(std)\n",
        "        return mu + eps * std\n",
        "\n",
        "    def decoder(self, z):\n",
        "        h = torch.tanh(self.decoder1(z))\n",
        "        return torch.sigmoid(self.decoder2(h))\n",
        "\n",
        "    def forward(self, x):\n",
        "        mu, std = self.encoder(x.view(-1, self.input_dim))\n",
        "        z = self.sampling(mu, std)\n",
        "        return self.decoder(z), mu, std\n",
        "\n",
        "    @staticmethod\n",
        "    def loss_func(x_hat, x, mu, std):\n",
        "        # define the reconstruction loss\n",
        "        ERR = F.binary_cross_entropy(x_hat, x, reduction='sum')\n",
        "        # define the KL divergence loss\n",
        "        KLD = -0.5 * torch.sum(1 + torch.log(std ** 2) - mu ** 2 - std ** 2)\n",
        "        return ERR + KLD, ERR, KLD"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "f349y3iuE-Cz"
      },
      "source": [
        "#### Classification Models"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {
        "id": "s2hbJUcCjiEY"
      },
      "outputs": [],
      "source": [
        "global_data_identifier = '' # will be set later\n",
        "cache_trained_model = False # will be set later\n",
        "train_only = False # will be set later\n",
        "enable_grid_search = False # will be set later\n",
        "\n",
        "def get_metrics(clf, X_test_encoded, y_test):\n",
        "    y_true, y_pred = y_test, clf.predict(X_test_encoded)\n",
        "    y_prob = clf.predict_proba(X_test_encoded)[:, 1]\n",
        "    # Performance Metrics: AUC, ACC, Recall, Precision, F1_score\n",
        "    metrics = {\n",
        "        'AUC': roc_auc_score(y_true, y_prob),\n",
        "        'ACC': accuracy_score(y_true, y_pred),\n",
        "        'Recall': recall_score(y_true, y_pred),\n",
        "        'Precision': precision_score(y_true, y_pred),\n",
        "        'F1_score': f1_score(y_true, y_pred),\n",
        "    }\n",
        "    return metrics\n",
        "\n",
        "\n",
        "def classification_with_pytorch_model(model, X_train, y_train, X_test, y_test, device, is_vae=False):\n",
        "    model = model.to(device)\n",
        "    # check if X_train is tensor already\n",
        "    if not torch.is_tensor(X_train):\n",
        "        X_train = torch.tensor(X_train).to(device)\n",
        "        X_test = torch.tensor(X_test).to(device)\n",
        "\n",
        "    if is_vae:\n",
        "        X_train_encoded = model.encoder(X_train)[0].cpu().detach().numpy()\n",
        "        X_test_encoded = model.encoder(X_test)[0].cpu().detach().numpy()\n",
        "    else:\n",
        "        X_train_encoded = model.encoder(X_train).cpu().detach().numpy()\n",
        "        X_test_encoded = model.encoder(X_test).cpu().detach().numpy()\n",
        "    X_train_encoded = np.reshape(X_train_encoded, (X_train_encoded.shape[0], -1))\n",
        "    X_test_encoded = np.reshape(X_test_encoded, (X_test_encoded.shape[0], -1))\n",
        "    return classification(X_train_encoded, y_train, X_test_encoded, y_test)\n",
        "\n",
        "\n",
        "def run_sub_classification(method, X_train_encoded, y_train, X_test_encoded, y_test):\n",
        "    clf = \"\"\n",
        "    hyper_parameters = {}\n",
        "    # hyper-parameter grids for classifiers\n",
        "    rf_hyper_parameters = [{'n_estimators': [s for s in range(100, 1001, 200)],\n",
        "                            'max_features': ['sqrt', 'log2'],\n",
        "                            'min_samples_leaf': [1, 2, 3, 4, 5],\n",
        "                            'criterion': ['gini', 'entropy']\n",
        "                            }, ]\n",
        "    svm_hyper_parameters = [{'C': [2 ** s for s in range(-5, 6, 2)], 'kernel': ['linear']},\n",
        "                            {'C': [2 ** s for s in range(-5, 6, 2)], 'gamma': [2 ** s for s in range(3, -15, -2)],\n",
        "                             'kernel': ['rbf']}]\n",
        "    mlp_hyper_parameters = [{'hidden_layer_sizes': [(64, 64), (128, 128), (256, 256), (512, 512)],\n",
        "                             'alpha': [2 ** s for s in range(-15, 1, 2)]}]\n",
        "\n",
        "    if method == 'SVM':\n",
        "        clf = SVC(kernel='linear', probability=True)\n",
        "        hyper_parameters = svm_hyper_parameters\n",
        "    elif method == 'Random Forest':\n",
        "        clf = RandomForestClassifier(n_estimators=100)\n",
        "        hyper_parameters = rf_hyper_parameters\n",
        "    elif method == 'MLP':\n",
        "        clf = MLPClassifier(hidden_layer_sizes=(64, 64), max_iter=1000)\n",
        "        hyper_parameters = mlp_hyper_parameters\n",
        "    if enable_grid_search:\n",
        "        clf = GridSearchCV(clf, param_grid=hyper_parameters,\n",
        "                                cv=StratifiedKFold(n_splits=5, shuffle=True), scoring='roc_auc',\n",
        "                                verbose=1, n_jobs=2)\n",
        "    clf.fit(X_train_encoded, y_train)\n",
        "    metrics = get_metrics(clf, X_test_encoded, y_test)\n",
        "    metrics['classification_method'] = method\n",
        "    print(metrics)\n",
        "    if enable_grid_search:\n",
        "        print(\"Best hyper-parameters: \", clf.best_params_)\n",
        "        metrics['hyper_parameters'] = clf.best_params_\n",
        "    return metrics\n",
        "\n",
        "\n",
        "# Classification by using the transformed data\n",
        "def classification(X_train_encoded, y_train, X_test_encoded, y_test):\n",
        "    methods = ['MLP', 'SVM', 'Random Forest']\n",
        "    metrics_list = []\n",
        "    for method in methods:\n",
        "        classification_start_time = time.time()\n",
        "        metrics = run_sub_classification(method, X_train_encoded.copy(), y_train.copy(), X_test_encoded.copy(),\n",
        "                                         y_test.copy())\n",
        "        classification_end_time = time.time()\n",
        "        print(method, \"Classification time: \", classification_end_time - classification_start_time)\n",
        "        metrics['classification_time'] = classification_end_time - classification_start_time\n",
        "        metrics_list.append(metrics)\n",
        "    return metrics_list"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "LS3jIzt9uj9B"
      },
      "source": [
        "## 4.4. Training\n",
        "\n",
        "### 4.4.1. Training Objectives\n",
        "- Split each dataset into a training set, validation set, and test set (64% training, 16% validation, and 20% test).\n",
        "- Exclude the test set from model training.\n",
        "- Implement early-stopping strategy: train models on the training set, compute reconstruction loss for the validation set after each epoch, stop training if no improvement in validation loss is observed for 20 epochs.\n",
        "- Select the model with the lowest validation loss as the best model.\n",
        "Utilize mean squared error as the reconstruction loss metric.\n",
        "- Apply adaptive moment estimation (Adam) optimizer with default parameters (learning rate: 0.001, epsilon: 1e-07) as specified in the original paper.\n",
        "- Utilize the encoder part of the best model to generate low-dimensional representations of microbiome data for subsequent disease prediction tasks.\n",
        "\n",
        "### 4.4.2. Hyperparams\n",
        "\n",
        "In the base case of svm classification, we have hyperparameter number of folds equal to 5. The epoch is set at max 200-2000, batch set is 32 for each autoencoder algorithm.\n",
        "\n",
        "The original implementation use command line arguments to facilitate the execution of experiments with varying configurations. In order to enhance usability within a Colab notebook environment, we modified this approach by setting the hyperparameters in our model directly.\n",
        "\n",
        "\n",
        "### 4.4.3. Computation Requirements\n",
        "\n",
        "According to the original paper, an equipment with an Intel Core i9-9820X CPU (10 cores), 64 GB Memory, and a GPU of NVIDIA GTX 1080 Ti, running on Ubuntu 18.04 have the latency in time as in Table 3 below.\n",
        "\n",
        "![Table 3](assets/Computation_Table3.png)\n",
        "\n",
        "In our experiments, we utilized the following computational resources\n",
        "\n",
        "1. Hardware:\n",
        "  - iMac with Apple M1 chip / 16GB RAM / 8-core CPU / 8-core GPU (for initial code verification and testing)\n",
        "  - Google Colab Pro with GPU and High-RAM (for AE models training)\n",
        "  - Google Colab Pro with CPU (for sklearn classification models)\n",
        "2. Average runtime for each epoch:\n",
        "  - PCA: 0.6s\n",
        "  - RP: 0.6s\n",
        "  - Shallow AE: 0.3s\n",
        "  - AE: 2.6s\n",
        "  - CAE: 1.5s\n",
        "  - VAE: 1.0s\n",
        "3. GPU hours used: around 1 hour for all AE models training, excluding the time for data download, preprocessing and classification models\n",
        "4. CPU hours used: since the sklearn classification models are run on CPU, the time is around 6 hour for all classification models training for each dataset.\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "qGdp9R-IFwAZ"
      },
      "source": [
        "#### Training Code"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {
        "id": "xhLeFpd1Fobs"
      },
      "outputs": [],
      "source": [
        "max_epoch_num = 2000\n",
        "\n",
        "def train_vae(model, train_loader, test_loader, device):\n",
        "    model = model.to(device)\n",
        "    optimizer = torch.optim.AdamW(model.parameters(), lr=0.001)\n",
        "    rec_loss, kl_loss = [], []\n",
        "    early_stopper = EarlyStopper(patience=5, delta=0.0)\n",
        "    for epoch in range(1, max_epoch_num + 1):\n",
        "\n",
        "        \"\"\" model training \"\"\"\n",
        "        model.train()\n",
        "        cur_rec_loss, cur_kl_loss = [], []\n",
        "        for batch_idx, (data, _) in enumerate(train_loader):\n",
        "            data = data.to(device)\n",
        "            rec, mu, std = model(data)\n",
        "            loss, err, kl = model.loss_func(rec, data.reshape(-1, model.input_dim), mu, std)\n",
        "\n",
        "            optimizer.zero_grad()\n",
        "            loss.backward()\n",
        "            optimizer.step()\n",
        "\n",
        "            cur_rec_loss.append(err.item())\n",
        "            cur_kl_loss.append(kl.item())\n",
        "\n",
        "        rec_loss.append(np.mean(cur_rec_loss))\n",
        "        kl_loss.append(np.mean(cur_kl_loss))\n",
        "\n",
        "        \"\"\" model evaluation \"\"\"\n",
        "        with torch.no_grad():\n",
        "            test_loss = []\n",
        "            for batch_idx, (data, _) in enumerate(test_loader):\n",
        "                data = data.to(device)\n",
        "                rec, mu, std = model(data)\n",
        "                _, mse, _ = model.loss_func(rec, data.reshape(data.shape[0], -1), mu, std)\n",
        "                test_loss.append(mse.item())\n",
        "            if early_stopper(np.mean(test_loss)):\n",
        "                print(f\"Early stopping at epoch {epoch}\")\n",
        "                break\n",
        "\n",
        "        if epoch == 1 or epoch % 10 == 0:\n",
        "            print(\n",
        "                f\"-- epoch {epoch} --, train MSE: {np.mean(cur_rec_loss)}, train KL: {np.mean(cur_kl_loss)}, test MSE: {np.mean(test_loss)}\")\n",
        "    return model\n",
        "\n",
        "\n",
        "def train_ae(model, train_loader, test_loader, device):\n",
        "    optimizer = torch.optim.AdamW(model.parameters(), lr=0.002)\n",
        "\n",
        "    losses = []\n",
        "    early_stopper = EarlyStopper(patience=5, delta=0.0)\n",
        "    for epoch in range(1, max_epoch_num + 1):\n",
        "\n",
        "        \"\"\" model training \"\"\"\n",
        "        model.train()\n",
        "        cur_rec_loss = []\n",
        "        for batch_idx, (data, _) in enumerate(train_loader):\n",
        "            data = data.to(device)\n",
        "            rec = model(data)\n",
        "            loss = model.loss_func(rec, data)\n",
        "\n",
        "            optimizer.zero_grad()\n",
        "            loss.backward()\n",
        "            optimizer.step()\n",
        "\n",
        "            cur_rec_loss.append(loss.item())\n",
        "        losses.append(np.mean(cur_rec_loss))\n",
        "\n",
        "        \"\"\" model evaluation \"\"\"\n",
        "        with torch.no_grad():\n",
        "            test_loss = []\n",
        "            for batch_idx, (data, _) in enumerate(test_loader):\n",
        "                data = data.to(device)\n",
        "                rec = model(data)\n",
        "                loss = model.loss_func(rec, data)\n",
        "                test_loss.append(loss.item())\n",
        "            if early_stopper(np.mean(test_loss)):\n",
        "                print(f\"Early stopping at epoch {epoch}\")\n",
        "                break\n",
        "\n",
        "        if epoch == 1 or epoch % 10 == 0:\n",
        "            print(f\"-- epoch {epoch} --, train MSE: {np.mean(cur_rec_loss)}, test MSE: {np.mean(test_loss)}\")\n",
        "\n",
        "    return model\n",
        "\n",
        "def run_model_training_and_classification(model_type, model, X_train, X_test, y_train, y_test, train_loader, test_loader, device):\n",
        "    path = './cached_model/' + global_data_identifier + '_' + model_type + '_model.pth'\n",
        "    print(\"Checking model path: \", path)\n",
        "    if cache_trained_model and os.path.exists(path):\n",
        "        model = torch.load(path, map_location=device)\n",
        "    else:\n",
        "        if model_type == 'vae':\n",
        "            model = train_vae(model, train_loader, test_loader, device)\n",
        "        else:\n",
        "            model = train_ae(model, train_loader, test_loader, device)\n",
        "        if cache_trained_model:\n",
        "            torch.save(model, path)\n",
        "    if train_only:\n",
        "        return []\n",
        "    return classification_with_pytorch_model(model, X_train, y_train, X_test, y_test, device, is_vae=(model_type == 'vae'))\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {},
      "outputs": [],
      "source": [
        "def run_shallow_ae_experiment(X_train, X_test, y_train, y_test, train_loader, test_loader, device):\n",
        "    model = ShallowAutoEncoder(input_dim=X_train.shape[1]).to(device)\n",
        "    return run_model_training_and_classification('shallow_ae', model, X_train, X_test, y_train, y_test, train_loader, test_loader, device)\n",
        "\n",
        "\n",
        "def run_ae_experiment(X_train, X_test, y_train, y_test, train_loader, test_loader, device):\n",
        "    model = AutoEncoder(input_dim=X_train.shape[1]).to(device)\n",
        "    return run_model_training_and_classification('ae', model, X_train, X_test, y_train, y_test, train_loader, test_loader, device)\n",
        "\n",
        "\n",
        "def run_cae_experiment(X_train, X_test, y_train, y_test, device):\n",
        "    # one_side_dim = int(math.sqrt(X_train.shape[1])) + 1\n",
        "    one_side_dim = 348\n",
        "    enlarged_dim = one_side_dim ** 2\n",
        "    X_train = np.column_stack((X_train, np.zeros((X_train.shape[0], enlarged_dim - X_train.shape[1]))))\n",
        "    X_test = np.column_stack((X_test, np.zeros((X_test.shape[0], enlarged_dim - X_test.shape[1]))))\n",
        "    # reshape\n",
        "    X_train = np.reshape(X_train, (len(X_train), one_side_dim, one_side_dim, 1))\n",
        "    X_test = np.reshape(X_test, (len(X_test), one_side_dim, one_side_dim, 1))\n",
        "    # to float32\n",
        "    X_train = X_train.astype('float32')\n",
        "    X_test = X_test.astype('float32')\n",
        "    X_train = torch.tensor(X_train).to(device)\n",
        "    X_test = torch.tensor(X_test).to(device)\n",
        "    X_train = X_train.transpose(1, 3)\n",
        "    X_test = X_test.transpose(1, 3)\n",
        "    train_loader = DataLoader(TensorDataset(torch.tensor(X_train), torch.tensor(y_train)), batch_size=32, shuffle=True)\n",
        "    test_loader = DataLoader(TensorDataset(torch.tensor(X_test), torch.tensor(y_test)), batch_size=32, shuffle=False)\n",
        "    model = CAE().to(device)\n",
        "    return run_model_training_and_classification('cae', model, X_train, X_test, y_train, y_test, train_loader, test_loader, device)\n",
        "\n",
        "\n",
        "def run_vae_experiment(X_train, X_test, y_train, y_test, train_loader, test_loader, device):\n",
        "    model = VAE(input_dim=X_train.shape[1]).to(device)\n",
        "    return run_model_training_and_classification('vae', model, X_train, X_test, y_train, y_test, train_loader, test_loader, device)\n",
        "\n",
        "\n",
        "def run_pca_experiment(X_train, X_test, y_train, y_test):\n",
        "    if train_only:\n",
        "        return []\n",
        "    pca = PCA(n_components=10)\n",
        "    pca.fit(X_train)\n",
        "    X_train_pca = pca.transform(X_train)\n",
        "    X_test_pca = pca.transform(X_test)\n",
        "    return classification(X_train_pca, y_train, X_test_pca, y_test)\n",
        "\n",
        "\n",
        "def run_rp_experiment(X_train, X_test, y_train, y_test):\n",
        "    if train_only:\n",
        "        return []\n",
        "    rp = GaussianRandomProjection(eps=0.5)\n",
        "    rp.fit(X_train)\n",
        "    X_train_rp = rp.transform(X_train)\n",
        "    X_test_rp = rp.transform(X_test)\n",
        "    return classification(X_train_rp, y_train, X_test_rp, y_test)\n",
        "\n",
        "\n",
        "def run_plain_experiment(X_train, X_test, y_train, y_test):\n",
        "    if train_only:\n",
        "        return []\n",
        "    return classification(X_train, y_train, X_test, y_test)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Nbdu_gV5FetZ"
      },
      "source": [
        "## 4.5. Evaluation\n",
        "\n",
        "### 4.5.1. Metrics Descriptions\n",
        "\n",
        "- Conduct 5-fold cross-validation on the reduced training set. This involves dividing the training set into five subsets, using four subsets for training and one for validation in each fold. This is to vary hyper-parameters and explore different combinations to find the best performing configuration.\n",
        "- Evaluate the performance of the models using the area under the receiver operating characteristics curve (AUC). This metric assesses the model's ability to distinguish between different classes and is commonly used for classification tasks.\n",
        "- Train a final classification model using the entire training set and the best hyper-parameter combination identified during cross-validation. This model aims to achieve optimal performance based on the selected configuration. Then test the final classification model on the separate test set, which was not used during training. This evaluation provides an unbiased assessment of the model's performance on unseen data.\n",
        "- Repeat the entire procedure five times, each time using a different random partition seed to create new training, validation, and test sets. This helps account for potential variations in performance due to the specific data splits. Average the resulting AUC scores obtained from the five repetitions. This average serves as a summary metric to compare the performance of different models or approaches. It provides a more robust assessment by considering multiple iterations of the evaluation process.\n",
        "\n",
        "### 4.5.2. Evaluation Code\n",
        "\n",
        "The detailed cross-validation codes are in the `run_sub_classification` function of the [Classification Models](#classification-models) section."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "metadata": {
        "id": "5WDva6WosGE6"
      },
      "outputs": [],
      "source": [
        "def run_sub_experiment_on_data(model_type, X_train, X_test, y_train, y_test, train_loader, test_loader, device):\n",
        "    metrics_list = []\n",
        "    if model_type == 'shallow_ae':\n",
        "        metrics_list = run_shallow_ae_experiment(X_train, X_test, y_train, y_test, train_loader, test_loader, device)\n",
        "    elif model_type == 'ae':\n",
        "        metrics_list = run_ae_experiment(X_train, X_test, y_train, y_test, train_loader, test_loader, device)\n",
        "    elif model_type == 'cae':\n",
        "        metrics_list = run_cae_experiment(X_train, X_test, y_train, y_test, device)\n",
        "    elif model_type == 'vae':\n",
        "        metrics_list = run_vae_experiment(X_train, X_test, y_train, y_test, train_loader, test_loader, device)\n",
        "    elif model_type == 'pca':\n",
        "        metrics_list = run_pca_experiment(X_train, X_test, y_train, y_test)\n",
        "    elif model_type == 'rp':\n",
        "        metrics_list = run_rp_experiment(X_train, X_test, y_train, y_test)\n",
        "    elif model_type == 'plain':\n",
        "        metrics_list = run_plain_experiment(X_train, X_test, y_train, y_test)\n",
        "    return metrics_list\n",
        "\n",
        "\n",
        "def run_experiment_on_data(data_dir, data_string):\n",
        "    X_train, X_test, y_train, y_test = load_data(data_dir, data_string, dtype='float32')\n",
        "    train_loader = DataLoader(\n",
        "        TensorDataset(torch.tensor(X_train), torch.tensor(y_train)),\n",
        "        batch_size=32, shuffle=True)\n",
        "    test_loader = DataLoader(TensorDataset(torch.tensor(X_test), torch.tensor(y_test)),\n",
        "                             batch_size=32, shuffle=False)\n",
        "    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')\n",
        "    if torch.backends.mps.is_available():\n",
        "        device = torch.device(\"mps\")\n",
        "\n",
        "    all_metrics_df_on_current_data = pd.DataFrame()\n",
        "    model_types = ['cae', 'shallow_ae', 'ae', 'vae', 'pca', 'rp', 'plain']\n",
        "    for model_type in model_types:\n",
        "        print(\"Running experiment on model: \", model_type)\n",
        "        experiment_start_time = time.time()\n",
        "        metrics_list = run_sub_experiment_on_data(model_type, X_train, X_test, y_train, y_test, train_loader,\n",
        "                                                  test_loader, device)\n",
        "        experiment_end_time = time.time()\n",
        "        print(\"Experiment time: \", experiment_end_time - experiment_start_time)\n",
        "        if train_only:\n",
        "            continue\n",
        "\n",
        "        for metrics in metrics_list:\n",
        "            metrics['experiment_time'] = experiment_end_time - experiment_start_time\n",
        "            metrics['data'] = data_string\n",
        "            metrics['model'] = model_type\n",
        "            metrics['seed'] = seed\n",
        "            metrics['max_epoch_num'] = max_epoch_num\n",
        "            metrics['device'] = device\n",
        "            print(metrics)\n",
        "            # save metrics\n",
        "        metrics_df = pd.DataFrame(metrics_list)\n",
        "        all_metrics_df_on_current_data = pd.concat([all_metrics_df_on_current_data, metrics_df], ignore_index=True)\n",
        "    return all_metrics_df_on_current_data\n",
        "\n",
        "\n",
        "def run_experiment():\n",
        "    data_dir = './data/marker/'\n",
        "    data_list = os.listdir(data_dir)\n",
        "    for data_string in data_list:\n",
        "        global global_data_identifier\n",
        "        global_data_identifier = data_string.split('.')[0]\n",
        "        main_metrics_df = pd.DataFrame()\n",
        "        print(\"Running experiment on data: \", data_string)\n",
        "        all_metrics_df_on_current_data = run_experiment_on_data(data_dir, data_string)\n",
        "        if train_only:\n",
        "            continue\n",
        "        main_metrics_df = pd.concat([main_metrics_df, all_metrics_df_on_current_data], ignore_index=True)\n",
        "        csv_file_name = datetime.datetime.now().strftime(\"%Y%m%d-%H%M%S\") + '_metrics.csv'\n",
        "        main_metrics_df.to_csv(csv_file_name, index=False)\n",
        "        print(\"Metrics saved to file: \", csv_file_name)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "hY7oRbSRldrn"
      },
      "source": [
        "#### Run Experiments"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 14,
      "metadata": {},
      "outputs": [],
      "source": [
        "# run_experiment()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "Due to run_experiment() takes a long time to run, we will run the following code to test the functions"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Train data shape:  (185, 542)\n",
            "Test data shape:  (47, 542)\n",
            "Running experiment on model:  cae\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "/var/folders/w2/bzjy3qrs24365t_51844lvjh0000gn/T/ipykernel_27053/2425531048.py:27: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).\n",
            "  train_loader = DataLoader(TensorDataset(torch.tensor(X_train), torch.tensor(y_train)), batch_size=32, shuffle=True)\n",
            "/var/folders/w2/bzjy3qrs24365t_51844lvjh0000gn/T/ipykernel_27053/2425531048.py:28: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).\n",
            "  test_loader = DataLoader(TensorDataset(torch.tensor(X_test), torch.tensor(y_test)), batch_size=32, shuffle=False)\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Checking model path:  ./cached_model/_cae_model.pth\n",
            "-- epoch 1 --, train MSE: 0.14357408694922924, test MSE: 0.030834849923849106\n",
            "{'AUC': 0.5, 'ACC': 0.5106382978723404, 'Recall': 1.0, 'Precision': 0.5106382978723404, 'F1_score': 0.676056338028169, 'classification_method': 'MLP'}\n",
            "MLP Classification time:  5.686659812927246\n",
            "{'AUC': 0.7807971014492754, 'ACC': 0.7021276595744681, 'Recall': 0.5833333333333334, 'Precision': 0.7777777777777778, 'F1_score': 0.6666666666666666, 'classification_method': 'SVM'}\n",
            "SVM Classification time:  10.78099012374878\n",
            "{'AUC': 0.9456521739130436, 'ACC': 0.8723404255319149, 'Recall': 0.8333333333333334, 'Precision': 0.9090909090909091, 'F1_score': 0.8695652173913043, 'classification_method': 'Random Forest'}\n",
            "Random Forest Classification time:  0.5832839012145996\n",
            "Experiment time:  19.57994318008423\n",
            "{'AUC': 0.5, 'ACC': 0.5106382978723404, 'Recall': 1.0, 'Precision': 0.5106382978723404, 'F1_score': 0.676056338028169, 'classification_method': 'MLP', 'classification_time': 5.686659812927246, 'experiment_time': 19.57994318008423, 'data': 'abundance_Cirrhosis.txt', 'model': 'cae', 'seed': 42, 'max_epoch_num': 1, 'device': device(type='mps')}\n",
            "{'AUC': 0.7807971014492754, 'ACC': 0.7021276595744681, 'Recall': 0.5833333333333334, 'Precision': 0.7777777777777778, 'F1_score': 0.6666666666666666, 'classification_method': 'SVM', 'classification_time': 10.78099012374878, 'experiment_time': 19.57994318008423, 'data': 'abundance_Cirrhosis.txt', 'model': 'cae', 'seed': 42, 'max_epoch_num': 1, 'device': device(type='mps')}\n",
            "{'AUC': 0.9456521739130436, 'ACC': 0.8723404255319149, 'Recall': 0.8333333333333334, 'Precision': 0.9090909090909091, 'F1_score': 0.8695652173913043, 'classification_method': 'Random Forest', 'classification_time': 0.5832839012145996, 'experiment_time': 19.57994318008423, 'data': 'abundance_Cirrhosis.txt', 'model': 'cae', 'seed': 42, 'max_epoch_num': 1, 'device': device(type='mps')}\n",
            "Running experiment on model:  shallow_ae\n",
            "Checking model path:  ./cached_model/_shallow_ae_model.pth\n",
            "-- epoch 1 --, train MSE: 44341.095052083336, test MSE: 29365.63671875\n",
            "{'AUC': 0.7028985507246377, 'ACC': 0.6595744680851063, 'Recall': 0.5416666666666666, 'Precision': 0.7222222222222222, 'F1_score': 0.619047619047619, 'classification_method': 'MLP'}\n",
            "MLP Classification time:  0.4948279857635498\n",
            "{'AUC': 0.5996376811594203, 'ACC': 0.5957446808510638, 'Recall': 0.4583333333333333, 'Precision': 0.6470588235294118, 'F1_score': 0.5365853658536585, 'classification_method': 'SVM'}\n",
            "SVM Classification time:  0.006590843200683594\n",
            "{'AUC': 0.6956521739130435, 'ACC': 0.6595744680851063, 'Recall': 0.625, 'Precision': 0.6818181818181818, 'F1_score': 0.6521739130434783, 'classification_method': 'Random Forest'}\n",
            "Random Forest Classification time:  0.06829094886779785\n",
            "Experiment time:  0.7566220760345459\n",
            "{'AUC': 0.7028985507246377, 'ACC': 0.6595744680851063, 'Recall': 0.5416666666666666, 'Precision': 0.7222222222222222, 'F1_score': 0.619047619047619, 'classification_method': 'MLP', 'classification_time': 0.4948279857635498, 'experiment_time': 0.7566220760345459, 'data': 'abundance_Cirrhosis.txt', 'model': 'shallow_ae', 'seed': 42, 'max_epoch_num': 1, 'device': device(type='mps')}\n",
            "{'AUC': 0.5996376811594203, 'ACC': 0.5957446808510638, 'Recall': 0.4583333333333333, 'Precision': 0.6470588235294118, 'F1_score': 0.5365853658536585, 'classification_method': 'SVM', 'classification_time': 0.006590843200683594, 'experiment_time': 0.7566220760345459, 'data': 'abundance_Cirrhosis.txt', 'model': 'shallow_ae', 'seed': 42, 'max_epoch_num': 1, 'device': device(type='mps')}\n",
            "{'AUC': 0.6956521739130435, 'ACC': 0.6595744680851063, 'Recall': 0.625, 'Precision': 0.6818181818181818, 'F1_score': 0.6521739130434783, 'classification_method': 'Random Forest', 'classification_time': 0.06829094886779785, 'experiment_time': 0.7566220760345459, 'data': 'abundance_Cirrhosis.txt', 'model': 'shallow_ae', 'seed': 42, 'max_epoch_num': 1, 'device': device(type='mps')}\n",
            "Running experiment on model:  ae\n",
            "Checking model path:  ./cached_model/_ae_model.pth\n",
            "-- epoch 1 --, train MSE: 2.509113331635793, test MSE: 2.0104936361312866\n",
            "{'AUC': 0.7119565217391305, 'ACC': 0.6170212765957447, 'Recall': 0.6666666666666666, 'Precision': 0.6153846153846154, 'F1_score': 0.64, 'classification_method': 'MLP'}\n",
            "MLP Classification time:  0.7365589141845703\n",
            "{'AUC': 0.6612318840579711, 'ACC': 0.6170212765957447, 'Recall': 0.4166666666666667, 'Precision': 0.7142857142857143, 'F1_score': 0.5263157894736842, 'classification_method': 'SVM'}\n",
            "SVM Classification time:  0.009451150894165039\n",
            "{'AUC': 0.8070652173913043, 'ACC': 0.723404255319149, 'Recall': 0.6666666666666666, 'Precision': 0.7619047619047619, 'F1_score': 0.7111111111111111, 'classification_method': 'Random Forest'}\n",
            "Random Forest Classification time:  0.09013581275939941\n",
            "Experiment time:  1.3188390731811523\n",
            "{'AUC': 0.7119565217391305, 'ACC': 0.6170212765957447, 'Recall': 0.6666666666666666, 'Precision': 0.6153846153846154, 'F1_score': 0.64, 'classification_method': 'MLP', 'classification_time': 0.7365589141845703, 'experiment_time': 1.3188390731811523, 'data': 'abundance_Cirrhosis.txt', 'model': 'ae', 'seed': 42, 'max_epoch_num': 1, 'device': device(type='mps')}\n",
            "{'AUC': 0.6612318840579711, 'ACC': 0.6170212765957447, 'Recall': 0.4166666666666667, 'Precision': 0.7142857142857143, 'F1_score': 0.5263157894736842, 'classification_method': 'SVM', 'classification_time': 0.009451150894165039, 'experiment_time': 1.3188390731811523, 'data': 'abundance_Cirrhosis.txt', 'model': 'ae', 'seed': 42, 'max_epoch_num': 1, 'device': device(type='mps')}\n",
            "{'AUC': 0.8070652173913043, 'ACC': 0.723404255319149, 'Recall': 0.6666666666666666, 'Precision': 0.7619047619047619, 'F1_score': 0.7111111111111111, 'classification_method': 'Random Forest', 'classification_time': 0.09013581275939941, 'experiment_time': 1.3188390731811523, 'data': 'abundance_Cirrhosis.txt', 'model': 'ae', 'seed': 42, 'max_epoch_num': 1, 'device': device(type='mps')}\n",
            "Running experiment on model:  vae\n",
            "Checking model path:  ./cached_model/_vae_model.pth\n",
            "-- epoch 1 --, train MSE: 11332.626139322916, train KL: 31.046048482259113, test MSE: 7890.856689453125\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "/opt/homebrew/Caskroom/miniforge/base/envs/torch-mps/lib/python3.11/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.\n",
            "  warnings.warn(\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "{'AUC': 0.5996376811594203, 'ACC': 0.5319148936170213, 'Recall': 0.4166666666666667, 'Precision': 0.5555555555555556, 'F1_score': 0.4761904761904762, 'classification_method': 'MLP'}\n",
            "MLP Classification time:  1.1176891326904297\n",
            "{'AUC': 0.7336956521739131, 'ACC': 0.7659574468085106, 'Recall': 0.75, 'Precision': 0.782608695652174, 'F1_score': 0.7659574468085107, 'classification_method': 'SVM'}\n",
            "SVM Classification time:  0.004920244216918945\n",
            "{'AUC': 0.6458333333333333, 'ACC': 0.6808510638297872, 'Recall': 0.5833333333333334, 'Precision': 0.7368421052631579, 'F1_score': 0.6511627906976745, 'classification_method': 'Random Forest'}\n",
            "Random Forest Classification time:  0.058403968811035156\n",
            "Experiment time:  1.6211912631988525\n",
            "{'AUC': 0.5996376811594203, 'ACC': 0.5319148936170213, 'Recall': 0.4166666666666667, 'Precision': 0.5555555555555556, 'F1_score': 0.4761904761904762, 'classification_method': 'MLP', 'classification_time': 1.1176891326904297, 'experiment_time': 1.6211912631988525, 'data': 'abundance_Cirrhosis.txt', 'model': 'vae', 'seed': 42, 'max_epoch_num': 1, 'device': device(type='mps')}\n",
            "{'AUC': 0.7336956521739131, 'ACC': 0.7659574468085106, 'Recall': 0.75, 'Precision': 0.782608695652174, 'F1_score': 0.7659574468085107, 'classification_method': 'SVM', 'classification_time': 0.004920244216918945, 'experiment_time': 1.6211912631988525, 'data': 'abundance_Cirrhosis.txt', 'model': 'vae', 'seed': 42, 'max_epoch_num': 1, 'device': device(type='mps')}\n",
            "{'AUC': 0.6458333333333333, 'ACC': 0.6808510638297872, 'Recall': 0.5833333333333334, 'Precision': 0.7368421052631579, 'F1_score': 0.6511627906976745, 'classification_method': 'Random Forest', 'classification_time': 0.058403968811035156, 'experiment_time': 1.6211912631988525, 'data': 'abundance_Cirrhosis.txt', 'model': 'vae', 'seed': 42, 'max_epoch_num': 1, 'device': device(type='mps')}\n",
            "Running experiment on model:  pca\n",
            "{'AUC': 0.8043478260869564, 'ACC': 0.723404255319149, 'Recall': 0.7083333333333334, 'Precision': 0.7391304347826086, 'F1_score': 0.723404255319149, 'classification_method': 'MLP'}\n",
            "MLP Classification time:  0.5269961357116699\n",
            "{'AUC': 0.7916666666666666, 'ACC': 0.7659574468085106, 'Recall': 0.7083333333333334, 'Precision': 0.8095238095238095, 'F1_score': 0.7555555555555556, 'classification_method': 'SVM'}\n",
            "SVM Classification time:  0.09855413436889648\n",
            "{'AUC': 0.8043478260869565, 'ACC': 0.7659574468085106, 'Recall': 0.7083333333333334, 'Precision': 0.8095238095238095, 'F1_score': 0.7555555555555556, 'classification_method': 'Random Forest'}\n",
            "Random Forest Classification time:  0.05981111526489258\n",
            "Experiment time:  0.6932432651519775\n",
            "{'AUC': 0.8043478260869564, 'ACC': 0.723404255319149, 'Recall': 0.7083333333333334, 'Precision': 0.7391304347826086, 'F1_score': 0.723404255319149, 'classification_method': 'MLP', 'classification_time': 0.5269961357116699, 'experiment_time': 0.6932432651519775, 'data': 'abundance_Cirrhosis.txt', 'model': 'pca', 'seed': 42, 'max_epoch_num': 1, 'device': device(type='mps')}\n",
            "{'AUC': 0.7916666666666666, 'ACC': 0.7659574468085106, 'Recall': 0.7083333333333334, 'Precision': 0.8095238095238095, 'F1_score': 0.7555555555555556, 'classification_method': 'SVM', 'classification_time': 0.09855413436889648, 'experiment_time': 0.6932432651519775, 'data': 'abundance_Cirrhosis.txt', 'model': 'pca', 'seed': 42, 'max_epoch_num': 1, 'device': device(type='mps')}\n",
            "{'AUC': 0.8043478260869565, 'ACC': 0.7659574468085106, 'Recall': 0.7083333333333334, 'Precision': 0.8095238095238095, 'F1_score': 0.7555555555555556, 'classification_method': 'Random Forest', 'classification_time': 0.05981111526489258, 'experiment_time': 0.6932432651519775, 'data': 'abundance_Cirrhosis.txt', 'model': 'pca', 'seed': 42, 'max_epoch_num': 1, 'device': device(type='mps')}\n",
            "Running experiment on model:  rp\n",
            "{'AUC': 0.7445652173913043, 'ACC': 0.6808510638297872, 'Recall': 0.625, 'Precision': 0.7142857142857143, 'F1_score': 0.6666666666666666, 'classification_method': 'MLP'}\n",
            "MLP Classification time:  0.38222217559814453\n",
            "{'AUC': 0.5697463768115942, 'ACC': 0.6382978723404256, 'Recall': 0.6666666666666666, 'Precision': 0.64, 'F1_score': 0.6530612244897959, 'classification_method': 'SVM'}\n",
            "SVM Classification time:  0.06549406051635742\n",
            "{'AUC': 0.7663043478260869, 'ACC': 0.6808510638297872, 'Recall': 0.7083333333333334, 'Precision': 0.68, 'F1_score': 0.6938775510204083, 'classification_method': 'Random Forest'}\n",
            "Random Forest Classification time:  0.10489797592163086\n",
            "Experiment time:  0.5581879615783691\n",
            "{'AUC': 0.7445652173913043, 'ACC': 0.6808510638297872, 'Recall': 0.625, 'Precision': 0.7142857142857143, 'F1_score': 0.6666666666666666, 'classification_method': 'MLP', 'classification_time': 0.38222217559814453, 'experiment_time': 0.5581879615783691, 'data': 'abundance_Cirrhosis.txt', 'model': 'rp', 'seed': 42, 'max_epoch_num': 1, 'device': device(type='mps')}\n",
            "{'AUC': 0.5697463768115942, 'ACC': 0.6382978723404256, 'Recall': 0.6666666666666666, 'Precision': 0.64, 'F1_score': 0.6530612244897959, 'classification_method': 'SVM', 'classification_time': 0.06549406051635742, 'experiment_time': 0.5581879615783691, 'data': 'abundance_Cirrhosis.txt', 'model': 'rp', 'seed': 42, 'max_epoch_num': 1, 'device': device(type='mps')}\n",
            "{'AUC': 0.7663043478260869, 'ACC': 0.6808510638297872, 'Recall': 0.7083333333333334, 'Precision': 0.68, 'F1_score': 0.6938775510204083, 'classification_method': 'Random Forest', 'classification_time': 0.10489797592163086, 'experiment_time': 0.5581879615783691, 'data': 'abundance_Cirrhosis.txt', 'model': 'rp', 'seed': 42, 'max_epoch_num': 1, 'device': device(type='mps')}\n",
            "Running experiment on model:  plain\n",
            "{'AUC': 0.7844202898550724, 'ACC': 0.6808510638297872, 'Recall': 0.5833333333333334, 'Precision': 0.7368421052631579, 'F1_score': 0.6511627906976745, 'classification_method': 'MLP'}\n",
            "MLP Classification time:  0.312791109085083\n",
            "{'AUC': 0.6657608695652174, 'ACC': 0.6382978723404256, 'Recall': 0.6666666666666666, 'Precision': 0.64, 'F1_score': 0.6530612244897959, 'classification_method': 'SVM'}\n",
            "SVM Classification time:  0.05573010444641113\n",
            "{'AUC': 0.980072463768116, 'ACC': 0.8936170212765957, 'Recall': 0.8333333333333334, 'Precision': 0.9523809523809523, 'F1_score': 0.888888888888889, 'classification_method': 'Random Forest'}\n",
            "Random Forest Classification time:  0.07001090049743652\n",
            "Experiment time:  0.438554048538208\n",
            "{'AUC': 0.7844202898550724, 'ACC': 0.6808510638297872, 'Recall': 0.5833333333333334, 'Precision': 0.7368421052631579, 'F1_score': 0.6511627906976745, 'classification_method': 'MLP', 'classification_time': 0.312791109085083, 'experiment_time': 0.438554048538208, 'data': 'abundance_Cirrhosis.txt', 'model': 'plain', 'seed': 42, 'max_epoch_num': 1, 'device': device(type='mps')}\n",
            "{'AUC': 0.6657608695652174, 'ACC': 0.6382978723404256, 'Recall': 0.6666666666666666, 'Precision': 0.64, 'F1_score': 0.6530612244897959, 'classification_method': 'SVM', 'classification_time': 0.05573010444641113, 'experiment_time': 0.438554048538208, 'data': 'abundance_Cirrhosis.txt', 'model': 'plain', 'seed': 42, 'max_epoch_num': 1, 'device': device(type='mps')}\n",
            "{'AUC': 0.980072463768116, 'ACC': 0.8936170212765957, 'Recall': 0.8333333333333334, 'Precision': 0.9523809523809523, 'F1_score': 0.888888888888889, 'classification_method': 'Random Forest', 'classification_time': 0.07001090049743652, 'experiment_time': 0.438554048538208, 'data': 'abundance_Cirrhosis.txt', 'model': 'plain', 'seed': 42, 'max_epoch_num': 1, 'device': device(type='mps')}\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>AUC</th>\n",
              "      <th>ACC</th>\n",
              "      <th>Recall</th>\n",
              "      <th>Precision</th>\n",
              "      <th>F1_score</th>\n",
              "      <th>classification_method</th>\n",
              "      <th>classification_time</th>\n",
              "      <th>experiment_time</th>\n",
              "      <th>data</th>\n",
              "      <th>model</th>\n",
              "      <th>seed</th>\n",
              "      <th>max_epoch_num</th>\n",
              "      <th>device</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>0.500000</td>\n",
              "      <td>0.510638</td>\n",
              "      <td>1.000000</td>\n",
              "      <td>0.510638</td>\n",
              "      <td>0.676056</td>\n",
              "      <td>MLP</td>\n",
              "      <td>5.686660</td>\n",
              "      <td>19.579943</td>\n",
              "      <td>abundance_Cirrhosis.txt</td>\n",
              "      <td>cae</td>\n",
              "      <td>42</td>\n",
              "      <td>1</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>0.780797</td>\n",
              "      <td>0.702128</td>\n",
              "      <td>0.583333</td>\n",
              "      <td>0.777778</td>\n",
              "      <td>0.666667</td>\n",
              "      <td>SVM</td>\n",
              "      <td>10.780990</td>\n",
              "      <td>19.579943</td>\n",
              "      <td>abundance_Cirrhosis.txt</td>\n",
              "      <td>cae</td>\n",
              "      <td>42</td>\n",
              "      <td>1</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>0.945652</td>\n",
              "      <td>0.872340</td>\n",
              "      <td>0.833333</td>\n",
              "      <td>0.909091</td>\n",
              "      <td>0.869565</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>0.583284</td>\n",
              "      <td>19.579943</td>\n",
              "      <td>abundance_Cirrhosis.txt</td>\n",
              "      <td>cae</td>\n",
              "      <td>42</td>\n",
              "      <td>1</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>0.702899</td>\n",
              "      <td>0.659574</td>\n",
              "      <td>0.541667</td>\n",
              "      <td>0.722222</td>\n",
              "      <td>0.619048</td>\n",
              "      <td>MLP</td>\n",
              "      <td>0.494828</td>\n",
              "      <td>0.756622</td>\n",
              "      <td>abundance_Cirrhosis.txt</td>\n",
              "      <td>shallow_ae</td>\n",
              "      <td>42</td>\n",
              "      <td>1</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>0.599638</td>\n",
              "      <td>0.595745</td>\n",
              "      <td>0.458333</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>0.536585</td>\n",
              "      <td>SVM</td>\n",
              "      <td>0.006591</td>\n",
              "      <td>0.756622</td>\n",
              "      <td>abundance_Cirrhosis.txt</td>\n",
              "      <td>shallow_ae</td>\n",
              "      <td>42</td>\n",
              "      <td>1</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>0.695652</td>\n",
              "      <td>0.659574</td>\n",
              "      <td>0.625000</td>\n",
              "      <td>0.681818</td>\n",
              "      <td>0.652174</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>0.068291</td>\n",
              "      <td>0.756622</td>\n",
              "      <td>abundance_Cirrhosis.txt</td>\n",
              "      <td>shallow_ae</td>\n",
              "      <td>42</td>\n",
              "      <td>1</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>0.711957</td>\n",
              "      <td>0.617021</td>\n",
              "      <td>0.666667</td>\n",
              "      <td>0.615385</td>\n",
              "      <td>0.640000</td>\n",
              "      <td>MLP</td>\n",
              "      <td>0.736559</td>\n",
              "      <td>1.318839</td>\n",
              "      <td>abundance_Cirrhosis.txt</td>\n",
              "      <td>ae</td>\n",
              "      <td>42</td>\n",
              "      <td>1</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>0.661232</td>\n",
              "      <td>0.617021</td>\n",
              "      <td>0.416667</td>\n",
              "      <td>0.714286</td>\n",
              "      <td>0.526316</td>\n",
              "      <td>SVM</td>\n",
              "      <td>0.009451</td>\n",
              "      <td>1.318839</td>\n",
              "      <td>abundance_Cirrhosis.txt</td>\n",
              "      <td>ae</td>\n",
              "      <td>42</td>\n",
              "      <td>1</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>0.807065</td>\n",
              "      <td>0.723404</td>\n",
              "      <td>0.666667</td>\n",
              "      <td>0.761905</td>\n",
              "      <td>0.711111</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>0.090136</td>\n",
              "      <td>1.318839</td>\n",
              "      <td>abundance_Cirrhosis.txt</td>\n",
              "      <td>ae</td>\n",
              "      <td>42</td>\n",
              "      <td>1</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>0.599638</td>\n",
              "      <td>0.531915</td>\n",
              "      <td>0.416667</td>\n",
              "      <td>0.555556</td>\n",
              "      <td>0.476190</td>\n",
              "      <td>MLP</td>\n",
              "      <td>1.117689</td>\n",
              "      <td>1.621191</td>\n",
              "      <td>abundance_Cirrhosis.txt</td>\n",
              "      <td>vae</td>\n",
              "      <td>42</td>\n",
              "      <td>1</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>0.733696</td>\n",
              "      <td>0.765957</td>\n",
              "      <td>0.750000</td>\n",
              "      <td>0.782609</td>\n",
              "      <td>0.765957</td>\n",
              "      <td>SVM</td>\n",
              "      <td>0.004920</td>\n",
              "      <td>1.621191</td>\n",
              "      <td>abundance_Cirrhosis.txt</td>\n",
              "      <td>vae</td>\n",
              "      <td>42</td>\n",
              "      <td>1</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>0.645833</td>\n",
              "      <td>0.680851</td>\n",
              "      <td>0.583333</td>\n",
              "      <td>0.736842</td>\n",
              "      <td>0.651163</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>0.058404</td>\n",
              "      <td>1.621191</td>\n",
              "      <td>abundance_Cirrhosis.txt</td>\n",
              "      <td>vae</td>\n",
              "      <td>42</td>\n",
              "      <td>1</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>0.804348</td>\n",
              "      <td>0.723404</td>\n",
              "      <td>0.708333</td>\n",
              "      <td>0.739130</td>\n",
              "      <td>0.723404</td>\n",
              "      <td>MLP</td>\n",
              "      <td>0.526996</td>\n",
              "      <td>0.693243</td>\n",
              "      <td>abundance_Cirrhosis.txt</td>\n",
              "      <td>pca</td>\n",
              "      <td>42</td>\n",
              "      <td>1</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>0.791667</td>\n",
              "      <td>0.765957</td>\n",
              "      <td>0.708333</td>\n",
              "      <td>0.809524</td>\n",
              "      <td>0.755556</td>\n",
              "      <td>SVM</td>\n",
              "      <td>0.098554</td>\n",
              "      <td>0.693243</td>\n",
              "      <td>abundance_Cirrhosis.txt</td>\n",
              "      <td>pca</td>\n",
              "      <td>42</td>\n",
              "      <td>1</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>0.804348</td>\n",
              "      <td>0.765957</td>\n",
              "      <td>0.708333</td>\n",
              "      <td>0.809524</td>\n",
              "      <td>0.755556</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>0.059811</td>\n",
              "      <td>0.693243</td>\n",
              "      <td>abundance_Cirrhosis.txt</td>\n",
              "      <td>pca</td>\n",
              "      <td>42</td>\n",
              "      <td>1</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>0.744565</td>\n",
              "      <td>0.680851</td>\n",
              "      <td>0.625000</td>\n",
              "      <td>0.714286</td>\n",
              "      <td>0.666667</td>\n",
              "      <td>MLP</td>\n",
              "      <td>0.382222</td>\n",
              "      <td>0.558188</td>\n",
              "      <td>abundance_Cirrhosis.txt</td>\n",
              "      <td>rp</td>\n",
              "      <td>42</td>\n",
              "      <td>1</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>0.569746</td>\n",
              "      <td>0.638298</td>\n",
              "      <td>0.666667</td>\n",
              "      <td>0.640000</td>\n",
              "      <td>0.653061</td>\n",
              "      <td>SVM</td>\n",
              "      <td>0.065494</td>\n",
              "      <td>0.558188</td>\n",
              "      <td>abundance_Cirrhosis.txt</td>\n",
              "      <td>rp</td>\n",
              "      <td>42</td>\n",
              "      <td>1</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>0.766304</td>\n",
              "      <td>0.680851</td>\n",
              "      <td>0.708333</td>\n",
              "      <td>0.680000</td>\n",
              "      <td>0.693878</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>0.104898</td>\n",
              "      <td>0.558188</td>\n",
              "      <td>abundance_Cirrhosis.txt</td>\n",
              "      <td>rp</td>\n",
              "      <td>42</td>\n",
              "      <td>1</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>0.784420</td>\n",
              "      <td>0.680851</td>\n",
              "      <td>0.583333</td>\n",
              "      <td>0.736842</td>\n",
              "      <td>0.651163</td>\n",
              "      <td>MLP</td>\n",
              "      <td>0.312791</td>\n",
              "      <td>0.438554</td>\n",
              "      <td>abundance_Cirrhosis.txt</td>\n",
              "      <td>plain</td>\n",
              "      <td>42</td>\n",
              "      <td>1</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>0.665761</td>\n",
              "      <td>0.638298</td>\n",
              "      <td>0.666667</td>\n",
              "      <td>0.640000</td>\n",
              "      <td>0.653061</td>\n",
              "      <td>SVM</td>\n",
              "      <td>0.055730</td>\n",
              "      <td>0.438554</td>\n",
              "      <td>abundance_Cirrhosis.txt</td>\n",
              "      <td>plain</td>\n",
              "      <td>42</td>\n",
              "      <td>1</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20</th>\n",
              "      <td>0.980072</td>\n",
              "      <td>0.893617</td>\n",
              "      <td>0.833333</td>\n",
              "      <td>0.952381</td>\n",
              "      <td>0.888889</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>0.070011</td>\n",
              "      <td>0.438554</td>\n",
              "      <td>abundance_Cirrhosis.txt</td>\n",
              "      <td>plain</td>\n",
              "      <td>42</td>\n",
              "      <td>1</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "         AUC       ACC    Recall  Precision  F1_score classification_method  \\\n",
              "0   0.500000  0.510638  1.000000   0.510638  0.676056                   MLP   \n",
              "1   0.780797  0.702128  0.583333   0.777778  0.666667                   SVM   \n",
              "2   0.945652  0.872340  0.833333   0.909091  0.869565         Random Forest   \n",
              "3   0.702899  0.659574  0.541667   0.722222  0.619048                   MLP   \n",
              "4   0.599638  0.595745  0.458333   0.647059  0.536585                   SVM   \n",
              "5   0.695652  0.659574  0.625000   0.681818  0.652174         Random Forest   \n",
              "6   0.711957  0.617021  0.666667   0.615385  0.640000                   MLP   \n",
              "7   0.661232  0.617021  0.416667   0.714286  0.526316                   SVM   \n",
              "8   0.807065  0.723404  0.666667   0.761905  0.711111         Random Forest   \n",
              "9   0.599638  0.531915  0.416667   0.555556  0.476190                   MLP   \n",
              "10  0.733696  0.765957  0.750000   0.782609  0.765957                   SVM   \n",
              "11  0.645833  0.680851  0.583333   0.736842  0.651163         Random Forest   \n",
              "12  0.804348  0.723404  0.708333   0.739130  0.723404                   MLP   \n",
              "13  0.791667  0.765957  0.708333   0.809524  0.755556                   SVM   \n",
              "14  0.804348  0.765957  0.708333   0.809524  0.755556         Random Forest   \n",
              "15  0.744565  0.680851  0.625000   0.714286  0.666667                   MLP   \n",
              "16  0.569746  0.638298  0.666667   0.640000  0.653061                   SVM   \n",
              "17  0.766304  0.680851  0.708333   0.680000  0.693878         Random Forest   \n",
              "18  0.784420  0.680851  0.583333   0.736842  0.651163                   MLP   \n",
              "19  0.665761  0.638298  0.666667   0.640000  0.653061                   SVM   \n",
              "20  0.980072  0.893617  0.833333   0.952381  0.888889         Random Forest   \n",
              "\n",
              "    classification_time  experiment_time                     data       model  \\\n",
              "0              5.686660        19.579943  abundance_Cirrhosis.txt         cae   \n",
              "1             10.780990        19.579943  abundance_Cirrhosis.txt         cae   \n",
              "2              0.583284        19.579943  abundance_Cirrhosis.txt         cae   \n",
              "3              0.494828         0.756622  abundance_Cirrhosis.txt  shallow_ae   \n",
              "4              0.006591         0.756622  abundance_Cirrhosis.txt  shallow_ae   \n",
              "5              0.068291         0.756622  abundance_Cirrhosis.txt  shallow_ae   \n",
              "6              0.736559         1.318839  abundance_Cirrhosis.txt          ae   \n",
              "7              0.009451         1.318839  abundance_Cirrhosis.txt          ae   \n",
              "8              0.090136         1.318839  abundance_Cirrhosis.txt          ae   \n",
              "9              1.117689         1.621191  abundance_Cirrhosis.txt         vae   \n",
              "10             0.004920         1.621191  abundance_Cirrhosis.txt         vae   \n",
              "11             0.058404         1.621191  abundance_Cirrhosis.txt         vae   \n",
              "12             0.526996         0.693243  abundance_Cirrhosis.txt         pca   \n",
              "13             0.098554         0.693243  abundance_Cirrhosis.txt         pca   \n",
              "14             0.059811         0.693243  abundance_Cirrhosis.txt         pca   \n",
              "15             0.382222         0.558188  abundance_Cirrhosis.txt          rp   \n",
              "16             0.065494         0.558188  abundance_Cirrhosis.txt          rp   \n",
              "17             0.104898         0.558188  abundance_Cirrhosis.txt          rp   \n",
              "18             0.312791         0.438554  abundance_Cirrhosis.txt       plain   \n",
              "19             0.055730         0.438554  abundance_Cirrhosis.txt       plain   \n",
              "20             0.070011         0.438554  abundance_Cirrhosis.txt       plain   \n",
              "\n",
              "    seed  max_epoch_num device  \n",
              "0     42              1    mps  \n",
              "1     42              1    mps  \n",
              "2     42              1    mps  \n",
              "3     42              1    mps  \n",
              "4     42              1    mps  \n",
              "5     42              1    mps  \n",
              "6     42              1    mps  \n",
              "7     42              1    mps  \n",
              "8     42              1    mps  \n",
              "9     42              1    mps  \n",
              "10    42              1    mps  \n",
              "11    42              1    mps  \n",
              "12    42              1    mps  \n",
              "13    42              1    mps  \n",
              "14    42              1    mps  \n",
              "15    42              1    mps  \n",
              "16    42              1    mps  \n",
              "17    42              1    mps  \n",
              "18    42              1    mps  \n",
              "19    42              1    mps  \n",
              "20    42              1    mps  "
            ]
          },
          "execution_count": 15,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "data_dir = './data/abundance/'\n",
        "data_string = 'abundance_Cirrhosis.txt'\n",
        "max_epoch_num = 1\n",
        "enable_grid_search = False\n",
        "run_experiment_on_data(data_dir, data_string)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "gX6bCcZNuxmz"
      },
      "source": [
        "# Results\n",
        "We recorded the best AUC scores for each model, classification algorithm, and dataset. The results are stored in the following csv files."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 16,
      "metadata": {
        "id": "Y4WgUnso82d1"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>AUC</th>\n",
              "      <th>ACC</th>\n",
              "      <th>Recall</th>\n",
              "      <th>Precision</th>\n",
              "      <th>F1_score</th>\n",
              "      <th>classification_method</th>\n",
              "      <th>hyper_parameters</th>\n",
              "      <th>classification_time</th>\n",
              "      <th>experiment_time</th>\n",
              "      <th>data</th>\n",
              "      <th>model</th>\n",
              "      <th>seed</th>\n",
              "      <th>max_epoch_num</th>\n",
              "      <th>device</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>0.947464</td>\n",
              "      <td>0.829787</td>\n",
              "      <td>0.791667</td>\n",
              "      <td>0.863636</td>\n",
              "      <td>0.826087</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 0.0001220703125, 'hidden_layer_sizes...</td>\n",
              "      <td>66.259109</td>\n",
              "      <td>238.745647</td>\n",
              "      <td>marker_Cirrhosis.txt</td>\n",
              "      <td>shallow_ae</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>0.916667</td>\n",
              "      <td>0.872340</td>\n",
              "      <td>0.875000</td>\n",
              "      <td>0.875000</td>\n",
              "      <td>0.875000</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 32, 'gamma': 0.0078125, 'kernel': 'rbf'}</td>\n",
              "      <td>1.120851</td>\n",
              "      <td>238.745647</td>\n",
              "      <td>marker_Cirrhosis.txt</td>\n",
              "      <td>shallow_ae</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>0.943841</td>\n",
              "      <td>0.829787</td>\n",
              "      <td>0.791667</td>\n",
              "      <td>0.863636</td>\n",
              "      <td>0.826087</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'entropy', 'max_features': 'sqrt...</td>\n",
              "      <td>111.178497</td>\n",
              "      <td>238.745647</td>\n",
              "      <td>marker_Cirrhosis.txt</td>\n",
              "      <td>shallow_ae</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>0.926630</td>\n",
              "      <td>0.829787</td>\n",
              "      <td>0.750000</td>\n",
              "      <td>0.900000</td>\n",
              "      <td>0.818182</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 3.0517578125e-05, 'hidden_layer_size...</td>\n",
              "      <td>64.426869</td>\n",
              "      <td>730.488681</td>\n",
              "      <td>marker_Cirrhosis.txt</td>\n",
              "      <td>ae</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>0.926630</td>\n",
              "      <td>0.851064</td>\n",
              "      <td>0.750000</td>\n",
              "      <td>0.947368</td>\n",
              "      <td>0.837209</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 2, 'kernel': 'linear'}</td>\n",
              "      <td>1.863737</td>\n",
              "      <td>730.488681</td>\n",
              "      <td>marker_Cirrhosis.txt</td>\n",
              "      <td>ae</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>0.902174</td>\n",
              "      <td>0.829787</td>\n",
              "      <td>0.916667</td>\n",
              "      <td>0.785714</td>\n",
              "      <td>0.846154</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'gini', 'max_features': 'log2', ...</td>\n",
              "      <td>119.985000</td>\n",
              "      <td>730.488681</td>\n",
              "      <td>marker_Cirrhosis.txt</td>\n",
              "      <td>ae</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>0.927536</td>\n",
              "      <td>0.829787</td>\n",
              "      <td>0.666667</td>\n",
              "      <td>1.000000</td>\n",
              "      <td>0.800000</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 0.0001220703125, 'hidden_layer_sizes...</td>\n",
              "      <td>2337.573239</td>\n",
              "      <td>13683.493345</td>\n",
              "      <td>marker_Cirrhosis.txt</td>\n",
              "      <td>cae</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>0.981884</td>\n",
              "      <td>0.893617</td>\n",
              "      <td>0.875000</td>\n",
              "      <td>0.913043</td>\n",
              "      <td>0.893617</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 0.03125, 'kernel': 'linear'}</td>\n",
              "      <td>8910.858763</td>\n",
              "      <td>13683.493345</td>\n",
              "      <td>marker_Cirrhosis.txt</td>\n",
              "      <td>cae</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>0.956522</td>\n",
              "      <td>0.851064</td>\n",
              "      <td>0.750000</td>\n",
              "      <td>0.947368</td>\n",
              "      <td>0.837209</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'entropy', 'max_features': 'sqrt...</td>\n",
              "      <td>2144.996647</td>\n",
              "      <td>13683.493345</td>\n",
              "      <td>marker_Cirrhosis.txt</td>\n",
              "      <td>cae</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>0.934783</td>\n",
              "      <td>0.808511</td>\n",
              "      <td>0.750000</td>\n",
              "      <td>0.857143</td>\n",
              "      <td>0.800000</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 0.0078125, 'hidden_layer_sizes': (25...</td>\n",
              "      <td>111.703098</td>\n",
              "      <td>433.025449</td>\n",
              "      <td>marker_Cirrhosis.txt</td>\n",
              "      <td>vae</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>0.951087</td>\n",
              "      <td>0.829787</td>\n",
              "      <td>0.708333</td>\n",
              "      <td>0.944444</td>\n",
              "      <td>0.809524</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 2, 'gamma': 0.0001220703125, 'kernel': '...</td>\n",
              "      <td>1.306651</td>\n",
              "      <td>433.025449</td>\n",
              "      <td>marker_Cirrhosis.txt</td>\n",
              "      <td>vae</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>0.936594</td>\n",
              "      <td>0.851064</td>\n",
              "      <td>0.791667</td>\n",
              "      <td>0.904762</td>\n",
              "      <td>0.844444</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'entropy', 'max_features': 'sqrt...</td>\n",
              "      <td>124.495656</td>\n",
              "      <td>433.025449</td>\n",
              "      <td>marker_Cirrhosis.txt</td>\n",
              "      <td>vae</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>0.922101</td>\n",
              "      <td>0.829787</td>\n",
              "      <td>0.708333</td>\n",
              "      <td>0.944444</td>\n",
              "      <td>0.809524</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 3.0517578125e-05, 'hidden_layer_size...</td>\n",
              "      <td>81.842762</td>\n",
              "      <td>243.937610</td>\n",
              "      <td>marker_Cirrhosis.txt</td>\n",
              "      <td>pca</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>0.971014</td>\n",
              "      <td>0.936170</td>\n",
              "      <td>0.875000</td>\n",
              "      <td>1.000000</td>\n",
              "      <td>0.933333</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 8, 'gamma': 0.0001220703125, 'kernel': '...</td>\n",
              "      <td>30.035022</td>\n",
              "      <td>243.937610</td>\n",
              "      <td>marker_Cirrhosis.txt</td>\n",
              "      <td>pca</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>0.934783</td>\n",
              "      <td>0.872340</td>\n",
              "      <td>0.791667</td>\n",
              "      <td>0.950000</td>\n",
              "      <td>0.863636</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'gini', 'max_features': 'log2', ...</td>\n",
              "      <td>131.445379</td>\n",
              "      <td>243.937610</td>\n",
              "      <td>marker_Cirrhosis.txt</td>\n",
              "      <td>pca</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>0.980072</td>\n",
              "      <td>0.872340</td>\n",
              "      <td>0.875000</td>\n",
              "      <td>0.875000</td>\n",
              "      <td>0.875000</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 0.5, 'hidden_layer_sizes': (256, 256)}</td>\n",
              "      <td>123.187775</td>\n",
              "      <td>315.226107</td>\n",
              "      <td>marker_Cirrhosis.txt</td>\n",
              "      <td>rp</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>0.974638</td>\n",
              "      <td>0.914894</td>\n",
              "      <td>0.833333</td>\n",
              "      <td>1.000000</td>\n",
              "      <td>0.909091</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 0.5, 'gamma': 0.0001220703125, 'kernel':...</td>\n",
              "      <td>4.674169</td>\n",
              "      <td>315.226107</td>\n",
              "      <td>marker_Cirrhosis.txt</td>\n",
              "      <td>rp</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>0.965580</td>\n",
              "      <td>0.914894</td>\n",
              "      <td>0.875000</td>\n",
              "      <td>0.954545</td>\n",
              "      <td>0.913043</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'gini', 'max_features': 'log2', ...</td>\n",
              "      <td>186.715222</td>\n",
              "      <td>315.226107</td>\n",
              "      <td>marker_Cirrhosis.txt</td>\n",
              "      <td>rp</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>0.989130</td>\n",
              "      <td>0.829787</td>\n",
              "      <td>1.000000</td>\n",
              "      <td>0.750000</td>\n",
              "      <td>0.857143</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 0.0001220703125, 'hidden_layer_sizes...</td>\n",
              "      <td>5058.035164</td>\n",
              "      <td>7645.472544</td>\n",
              "      <td>marker_Cirrhosis.txt</td>\n",
              "      <td>plain</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>0.983696</td>\n",
              "      <td>0.914894</td>\n",
              "      <td>0.875000</td>\n",
              "      <td>0.954545</td>\n",
              "      <td>0.913043</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 0.03125, 'kernel': 'linear'}</td>\n",
              "      <td>2277.429844</td>\n",
              "      <td>7645.472544</td>\n",
              "      <td>marker_Cirrhosis.txt</td>\n",
              "      <td>plain</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20</th>\n",
              "      <td>0.992754</td>\n",
              "      <td>0.914894</td>\n",
              "      <td>0.833333</td>\n",
              "      <td>1.000000</td>\n",
              "      <td>0.909091</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'gini', 'max_features': 'log2', ...</td>\n",
              "      <td>310.007509</td>\n",
              "      <td>7645.472544</td>\n",
              "      <td>marker_Cirrhosis.txt</td>\n",
              "      <td>plain</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>21</th>\n",
              "      <td>0.500000</td>\n",
              "      <td>0.352941</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 3.0517578125e-05, 'hidden_layer_size...</td>\n",
              "      <td>1486.891025</td>\n",
              "      <td>9414.390030</td>\n",
              "      <td>marker_Obesity.txt</td>\n",
              "      <td>cae</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>22</th>\n",
              "      <td>0.592593</td>\n",
              "      <td>0.607843</td>\n",
              "      <td>0.757576</td>\n",
              "      <td>0.675676</td>\n",
              "      <td>0.714286</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 0.03125, 'kernel': 'linear'}</td>\n",
              "      <td>7005.718553</td>\n",
              "      <td>9414.390030</td>\n",
              "      <td>marker_Obesity.txt</td>\n",
              "      <td>cae</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23</th>\n",
              "      <td>0.640572</td>\n",
              "      <td>0.627451</td>\n",
              "      <td>0.969697</td>\n",
              "      <td>0.640000</td>\n",
              "      <td>0.771084</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'entropy', 'max_features': 'sqrt...</td>\n",
              "      <td>918.833811</td>\n",
              "      <td>9414.390030</td>\n",
              "      <td>marker_Obesity.txt</td>\n",
              "      <td>cae</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>24</th>\n",
              "      <td>0.560606</td>\n",
              "      <td>0.450980</td>\n",
              "      <td>0.424242</td>\n",
              "      <td>0.608696</td>\n",
              "      <td>0.500000</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 0.03125, 'hidden_layer_sizes': (128,...</td>\n",
              "      <td>2762.416396</td>\n",
              "      <td>4146.935735</td>\n",
              "      <td>marker_Obesity.txt</td>\n",
              "      <td>plain</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>25</th>\n",
              "      <td>0.681818</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>1.000000</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>0.785714</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 0.125, 'gamma': 0.0001220703125, 'kernel...</td>\n",
              "      <td>1190.073292</td>\n",
              "      <td>4146.935735</td>\n",
              "      <td>marker_Obesity.txt</td>\n",
              "      <td>plain</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>26</th>\n",
              "      <td>0.739057</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>1.000000</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>0.785714</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'gini', 'max_features': 'log2', ...</td>\n",
              "      <td>194.442154</td>\n",
              "      <td>4146.935735</td>\n",
              "      <td>marker_Obesity.txt</td>\n",
              "      <td>plain</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>27</th>\n",
              "      <td>0.500000</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>1.000000</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>0.785714</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 3.0517578125e-05, 'hidden_layer_size...</td>\n",
              "      <td>16.176070</td>\n",
              "      <td>100.520304</td>\n",
              "      <td>marker_Obesity.txt</td>\n",
              "      <td>shallow_ae</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>28</th>\n",
              "      <td>0.500000</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>1.000000</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>0.785714</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 0.125, 'gamma': 0.125, 'kernel': 'rbf'}</td>\n",
              "      <td>0.634451</td>\n",
              "      <td>100.520304</td>\n",
              "      <td>marker_Obesity.txt</td>\n",
              "      <td>shallow_ae</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>29</th>\n",
              "      <td>0.500000</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>1.000000</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>0.785714</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'gini', 'max_features': 'sqrt', ...</td>\n",
              "      <td>83.237592</td>\n",
              "      <td>100.520304</td>\n",
              "      <td>marker_Obesity.txt</td>\n",
              "      <td>shallow_ae</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>30</th>\n",
              "      <td>0.439394</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>1.000000</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>0.785714</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 0.00048828125, 'hidden_layer_sizes':...</td>\n",
              "      <td>23.258092</td>\n",
              "      <td>118.358175</td>\n",
              "      <td>marker_Obesity.txt</td>\n",
              "      <td>ae</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>31</th>\n",
              "      <td>0.524411</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>1.000000</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>0.785714</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 0.03125, 'kernel': 'linear'}</td>\n",
              "      <td>0.865882</td>\n",
              "      <td>118.358175</td>\n",
              "      <td>marker_Obesity.txt</td>\n",
              "      <td>ae</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>32</th>\n",
              "      <td>0.524411</td>\n",
              "      <td>0.627451</td>\n",
              "      <td>0.969697</td>\n",
              "      <td>0.640000</td>\n",
              "      <td>0.771084</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'gini', 'max_features': 'sqrt', ...</td>\n",
              "      <td>89.596979</td>\n",
              "      <td>118.358175</td>\n",
              "      <td>marker_Obesity.txt</td>\n",
              "      <td>ae</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>33</th>\n",
              "      <td>0.370370</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>1.000000</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>0.785714</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 0.001953125, 'hidden_layer_sizes': (...</td>\n",
              "      <td>25.970036</td>\n",
              "      <td>90.790034</td>\n",
              "      <td>marker_Obesity.txt</td>\n",
              "      <td>vae</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>34</th>\n",
              "      <td>0.526936</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>1.000000</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>0.785714</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 0.125, 'kernel': 'linear'}</td>\n",
              "      <td>3.912843</td>\n",
              "      <td>90.790034</td>\n",
              "      <td>marker_Obesity.txt</td>\n",
              "      <td>vae</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35</th>\n",
              "      <td>0.501684</td>\n",
              "      <td>0.607843</td>\n",
              "      <td>0.878788</td>\n",
              "      <td>0.644444</td>\n",
              "      <td>0.743590</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'entropy', 'max_features': 'sqrt...</td>\n",
              "      <td>60.502643</td>\n",
              "      <td>90.790034</td>\n",
              "      <td>marker_Obesity.txt</td>\n",
              "      <td>vae</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>36</th>\n",
              "      <td>0.673401</td>\n",
              "      <td>0.627451</td>\n",
              "      <td>0.727273</td>\n",
              "      <td>0.705882</td>\n",
              "      <td>0.716418</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 0.03125, 'hidden_layer_sizes': (64, ...</td>\n",
              "      <td>45.460609</td>\n",
              "      <td>412.816045</td>\n",
              "      <td>marker_Obesity.txt</td>\n",
              "      <td>pca</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>37</th>\n",
              "      <td>0.654882</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>0.969697</td>\n",
              "      <td>0.653061</td>\n",
              "      <td>0.780488</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 32, 'kernel': 'linear'}</td>\n",
              "      <td>292.242113</td>\n",
              "      <td>412.816045</td>\n",
              "      <td>marker_Obesity.txt</td>\n",
              "      <td>pca</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>38</th>\n",
              "      <td>0.675084</td>\n",
              "      <td>0.686275</td>\n",
              "      <td>0.969697</td>\n",
              "      <td>0.680851</td>\n",
              "      <td>0.800000</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'gini', 'max_features': 'log2', ...</td>\n",
              "      <td>74.022311</td>\n",
              "      <td>412.816045</td>\n",
              "      <td>marker_Obesity.txt</td>\n",
              "      <td>pca</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>39</th>\n",
              "      <td>0.646465</td>\n",
              "      <td>0.686275</td>\n",
              "      <td>0.909091</td>\n",
              "      <td>0.697674</td>\n",
              "      <td>0.789474</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 0.5, 'hidden_layer_sizes': (64, 64)}</td>\n",
              "      <td>62.771949</td>\n",
              "      <td>191.006667</td>\n",
              "      <td>marker_Obesity.txt</td>\n",
              "      <td>rp</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>40</th>\n",
              "      <td>0.326599</td>\n",
              "      <td>0.607843</td>\n",
              "      <td>0.666667</td>\n",
              "      <td>0.709677</td>\n",
              "      <td>0.687500</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 0.03125, 'kernel': 'linear'}</td>\n",
              "      <td>5.681792</td>\n",
              "      <td>191.006667</td>\n",
              "      <td>marker_Obesity.txt</td>\n",
              "      <td>rp</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>41</th>\n",
              "      <td>0.685185</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>1.000000</td>\n",
              "      <td>0.647059</td>\n",
              "      <td>0.785714</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'gini', 'max_features': 'sqrt', ...</td>\n",
              "      <td>121.447971</td>\n",
              "      <td>191.006667</td>\n",
              "      <td>marker_Obesity.txt</td>\n",
              "      <td>rp</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>42</th>\n",
              "      <td>0.500000</td>\n",
              "      <td>0.507246</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 0.0001220703125, 'hidden_layer_sizes...</td>\n",
              "      <td>2492.601276</td>\n",
              "      <td>14115.473664</td>\n",
              "      <td>marker_T2D.txt</td>\n",
              "      <td>cae</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>43</th>\n",
              "      <td>0.801681</td>\n",
              "      <td>0.753623</td>\n",
              "      <td>0.794118</td>\n",
              "      <td>0.729730</td>\n",
              "      <td>0.760563</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 0.03125, 'kernel': 'linear'}</td>\n",
              "      <td>10209.280280</td>\n",
              "      <td>14115.473664</td>\n",
              "      <td>marker_T2D.txt</td>\n",
              "      <td>cae</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>44</th>\n",
              "      <td>0.755462</td>\n",
              "      <td>0.710145</td>\n",
              "      <td>0.794118</td>\n",
              "      <td>0.675000</td>\n",
              "      <td>0.729730</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'entropy', 'max_features': 'sqrt...</td>\n",
              "      <td>1409.363473</td>\n",
              "      <td>14115.473664</td>\n",
              "      <td>marker_T2D.txt</td>\n",
              "      <td>cae</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>45</th>\n",
              "      <td>0.807563</td>\n",
              "      <td>0.739130</td>\n",
              "      <td>0.764706</td>\n",
              "      <td>0.722222</td>\n",
              "      <td>0.742857</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 0.00048828125, 'hidden_layer_sizes':...</td>\n",
              "      <td>5410.766285</td>\n",
              "      <td>8130.763001</td>\n",
              "      <td>marker_T2D.txt</td>\n",
              "      <td>plain</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>46</th>\n",
              "      <td>0.771429</td>\n",
              "      <td>0.753623</td>\n",
              "      <td>0.823529</td>\n",
              "      <td>0.717949</td>\n",
              "      <td>0.767123</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 0.03125, 'kernel': 'linear'}</td>\n",
              "      <td>2465.565559</td>\n",
              "      <td>8130.763001</td>\n",
              "      <td>marker_T2D.txt</td>\n",
              "      <td>plain</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>47</th>\n",
              "      <td>0.727731</td>\n",
              "      <td>0.623188</td>\n",
              "      <td>0.617647</td>\n",
              "      <td>0.617647</td>\n",
              "      <td>0.617647</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'gini', 'max_features': 'log2', ...</td>\n",
              "      <td>254.427690</td>\n",
              "      <td>8130.763001</td>\n",
              "      <td>marker_T2D.txt</td>\n",
              "      <td>plain</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>48</th>\n",
              "      <td>0.546639</td>\n",
              "      <td>0.507246</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 0.0078125, 'hidden_layer_sizes': (64...</td>\n",
              "      <td>18.924324</td>\n",
              "      <td>100.779192</td>\n",
              "      <td>marker_T2D.txt</td>\n",
              "      <td>shallow_ae</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>49</th>\n",
              "      <td>0.485294</td>\n",
              "      <td>0.507246</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 0.03125, 'gamma': 0.001953125, 'kernel':...</td>\n",
              "      <td>1.156454</td>\n",
              "      <td>100.779192</td>\n",
              "      <td>marker_T2D.txt</td>\n",
              "      <td>shallow_ae</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>50</th>\n",
              "      <td>0.500000</td>\n",
              "      <td>0.507246</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'gini', 'max_features': 'sqrt', ...</td>\n",
              "      <td>80.019774</td>\n",
              "      <td>100.779192</td>\n",
              "      <td>marker_T2D.txt</td>\n",
              "      <td>shallow_ae</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>51</th>\n",
              "      <td>0.452521</td>\n",
              "      <td>0.492754</td>\n",
              "      <td>1.000000</td>\n",
              "      <td>0.492754</td>\n",
              "      <td>0.660194</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 3.0517578125e-05, 'hidden_layer_size...</td>\n",
              "      <td>28.255768</td>\n",
              "      <td>132.101581</td>\n",
              "      <td>marker_T2D.txt</td>\n",
              "      <td>ae</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>52</th>\n",
              "      <td>0.397479</td>\n",
              "      <td>0.507246</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 0.03125, 'kernel': 'linear'}</td>\n",
              "      <td>1.663867</td>\n",
              "      <td>132.101581</td>\n",
              "      <td>marker_T2D.txt</td>\n",
              "      <td>ae</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>53</th>\n",
              "      <td>0.697899</td>\n",
              "      <td>0.608696</td>\n",
              "      <td>0.264706</td>\n",
              "      <td>0.818182</td>\n",
              "      <td>0.400000</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'gini', 'max_features': 'sqrt', ...</td>\n",
              "      <td>96.230592</td>\n",
              "      <td>132.101581</td>\n",
              "      <td>marker_T2D.txt</td>\n",
              "      <td>ae</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>54</th>\n",
              "      <td>0.502521</td>\n",
              "      <td>0.507246</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 3.0517578125e-05, 'hidden_layer_size...</td>\n",
              "      <td>17.820294</td>\n",
              "      <td>100.809295</td>\n",
              "      <td>marker_T2D.txt</td>\n",
              "      <td>vae</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>55</th>\n",
              "      <td>0.500840</td>\n",
              "      <td>0.507246</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 8, 'gamma': 8, 'kernel': 'rbf'}</td>\n",
              "      <td>1.023353</td>\n",
              "      <td>100.809295</td>\n",
              "      <td>marker_T2D.txt</td>\n",
              "      <td>vae</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>56</th>\n",
              "      <td>0.500840</td>\n",
              "      <td>0.507246</td>\n",
              "      <td>0.029412</td>\n",
              "      <td>0.500000</td>\n",
              "      <td>0.055556</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'gini', 'max_features': 'sqrt', ...</td>\n",
              "      <td>80.574382</td>\n",
              "      <td>100.809295</td>\n",
              "      <td>marker_T2D.txt</td>\n",
              "      <td>vae</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>57</th>\n",
              "      <td>0.608403</td>\n",
              "      <td>0.579710</td>\n",
              "      <td>0.617647</td>\n",
              "      <td>0.567568</td>\n",
              "      <td>0.591549</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 0.00048828125, 'hidden_layer_sizes':...</td>\n",
              "      <td>61.699997</td>\n",
              "      <td>229.635357</td>\n",
              "      <td>marker_T2D.txt</td>\n",
              "      <td>pca</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>58</th>\n",
              "      <td>0.622689</td>\n",
              "      <td>0.579710</td>\n",
              "      <td>0.588235</td>\n",
              "      <td>0.571429</td>\n",
              "      <td>0.579710</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 0.5, 'gamma': 0.001953125, 'kernel': 'rbf'}</td>\n",
              "      <td>71.348738</td>\n",
              "      <td>229.635357</td>\n",
              "      <td>marker_T2D.txt</td>\n",
              "      <td>pca</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>59</th>\n",
              "      <td>0.625630</td>\n",
              "      <td>0.594203</td>\n",
              "      <td>0.558824</td>\n",
              "      <td>0.593750</td>\n",
              "      <td>0.575758</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'entropy', 'max_features': 'log2...</td>\n",
              "      <td>94.713534</td>\n",
              "      <td>229.635357</td>\n",
              "      <td>marker_T2D.txt</td>\n",
              "      <td>pca</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>60</th>\n",
              "      <td>0.567227</td>\n",
              "      <td>0.536232</td>\n",
              "      <td>0.470588</td>\n",
              "      <td>0.533333</td>\n",
              "      <td>0.500000</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 0.0001220703125, 'hidden_layer_sizes...</td>\n",
              "      <td>80.663704</td>\n",
              "      <td>219.272567</td>\n",
              "      <td>marker_T2D.txt</td>\n",
              "      <td>rp</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>61</th>\n",
              "      <td>0.610084</td>\n",
              "      <td>0.579710</td>\n",
              "      <td>0.529412</td>\n",
              "      <td>0.580645</td>\n",
              "      <td>0.553846</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 2, 'gamma': 0.0001220703125, 'kernel': '...</td>\n",
              "      <td>3.265329</td>\n",
              "      <td>219.272567</td>\n",
              "      <td>marker_T2D.txt</td>\n",
              "      <td>rp</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>62</th>\n",
              "      <td>0.640336</td>\n",
              "      <td>0.623188</td>\n",
              "      <td>0.500000</td>\n",
              "      <td>0.653846</td>\n",
              "      <td>0.566667</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'gini', 'max_features': 'log2', ...</td>\n",
              "      <td>134.022436</td>\n",
              "      <td>219.272567</td>\n",
              "      <td>marker_T2D.txt</td>\n",
              "      <td>rp</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>63</th>\n",
              "      <td>0.516667</td>\n",
              "      <td>0.600000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 0.0078125, 'hidden_layer_sizes': (25...</td>\n",
              "      <td>22.348105</td>\n",
              "      <td>102.883997</td>\n",
              "      <td>marker_Colorectal.txt</td>\n",
              "      <td>shallow_ae</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>64</th>\n",
              "      <td>0.550000</td>\n",
              "      <td>0.600000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 32, 'gamma': 0.0001220703125, 'kernel': ...</td>\n",
              "      <td>0.305911</td>\n",
              "      <td>102.883997</td>\n",
              "      <td>marker_Colorectal.txt</td>\n",
              "      <td>shallow_ae</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>65</th>\n",
              "      <td>0.583333</td>\n",
              "      <td>0.600000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'gini', 'max_features': 'sqrt', ...</td>\n",
              "      <td>54.457502</td>\n",
              "      <td>102.883997</td>\n",
              "      <td>marker_Colorectal.txt</td>\n",
              "      <td>shallow_ae</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>66</th>\n",
              "      <td>0.500000</td>\n",
              "      <td>0.600000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 0.001953125, 'hidden_layer_sizes': (...</td>\n",
              "      <td>21.643693</td>\n",
              "      <td>316.989440</td>\n",
              "      <td>marker_Colorectal.txt</td>\n",
              "      <td>ae</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>67</th>\n",
              "      <td>0.500000</td>\n",
              "      <td>0.600000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 0.03125, 'gamma': 0.00048828125, 'kernel...</td>\n",
              "      <td>0.401798</td>\n",
              "      <td>316.989440</td>\n",
              "      <td>marker_Colorectal.txt</td>\n",
              "      <td>ae</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>68</th>\n",
              "      <td>0.500000</td>\n",
              "      <td>0.600000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'gini', 'max_features': 'sqrt', ...</td>\n",
              "      <td>55.509926</td>\n",
              "      <td>316.989440</td>\n",
              "      <td>marker_Colorectal.txt</td>\n",
              "      <td>ae</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>69</th>\n",
              "      <td>0.740000</td>\n",
              "      <td>0.800000</td>\n",
              "      <td>0.500000</td>\n",
              "      <td>1.000000</td>\n",
              "      <td>0.666667</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 0.0001220703125, 'hidden_layer_sizes...</td>\n",
              "      <td>1359.514105</td>\n",
              "      <td>3378.567152</td>\n",
              "      <td>marker_Colorectal.txt</td>\n",
              "      <td>cae</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>70</th>\n",
              "      <td>0.773333</td>\n",
              "      <td>0.720000</td>\n",
              "      <td>0.500000</td>\n",
              "      <td>0.714286</td>\n",
              "      <td>0.588235</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 0.03125, 'kernel': 'linear'}</td>\n",
              "      <td>1354.391862</td>\n",
              "      <td>3378.567152</td>\n",
              "      <td>marker_Colorectal.txt</td>\n",
              "      <td>cae</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>71</th>\n",
              "      <td>0.753333</td>\n",
              "      <td>0.680000</td>\n",
              "      <td>0.400000</td>\n",
              "      <td>0.666667</td>\n",
              "      <td>0.500000</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'gini', 'max_features': 'sqrt', ...</td>\n",
              "      <td>521.247247</td>\n",
              "      <td>3378.567152</td>\n",
              "      <td>marker_Colorectal.txt</td>\n",
              "      <td>cae</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>72</th>\n",
              "      <td>0.640000</td>\n",
              "      <td>0.640000</td>\n",
              "      <td>0.500000</td>\n",
              "      <td>0.555556</td>\n",
              "      <td>0.526316</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 0.5, 'hidden_layer_sizes': (256, 256)}</td>\n",
              "      <td>40.422624</td>\n",
              "      <td>189.599260</td>\n",
              "      <td>marker_Colorectal.txt</td>\n",
              "      <td>vae</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>73</th>\n",
              "      <td>0.653333</td>\n",
              "      <td>0.640000</td>\n",
              "      <td>0.100000</td>\n",
              "      <td>1.000000</td>\n",
              "      <td>0.181818</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 0.5, 'gamma': 0.03125, 'kernel': 'rbf'}</td>\n",
              "      <td>0.399136</td>\n",
              "      <td>189.599260</td>\n",
              "      <td>marker_Colorectal.txt</td>\n",
              "      <td>vae</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>74</th>\n",
              "      <td>0.540000</td>\n",
              "      <td>0.640000</td>\n",
              "      <td>0.100000</td>\n",
              "      <td>1.000000</td>\n",
              "      <td>0.181818</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'entropy', 'max_features': 'sqrt...</td>\n",
              "      <td>59.198460</td>\n",
              "      <td>189.599260</td>\n",
              "      <td>marker_Colorectal.txt</td>\n",
              "      <td>vae</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>75</th>\n",
              "      <td>0.846667</td>\n",
              "      <td>0.760000</td>\n",
              "      <td>0.800000</td>\n",
              "      <td>0.666667</td>\n",
              "      <td>0.727273</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 3.0517578125e-05, 'hidden_layer_size...</td>\n",
              "      <td>30.598888</td>\n",
              "      <td>102.141932</td>\n",
              "      <td>marker_Colorectal.txt</td>\n",
              "      <td>pca</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>76</th>\n",
              "      <td>0.793333</td>\n",
              "      <td>0.760000</td>\n",
              "      <td>0.500000</td>\n",
              "      <td>0.833333</td>\n",
              "      <td>0.625000</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 32, 'gamma': 0.00048828125, 'kernel': 'r...</td>\n",
              "      <td>11.931073</td>\n",
              "      <td>102.141932</td>\n",
              "      <td>marker_Colorectal.txt</td>\n",
              "      <td>pca</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>77</th>\n",
              "      <td>0.766667</td>\n",
              "      <td>0.600000</td>\n",
              "      <td>0.300000</td>\n",
              "      <td>0.500000</td>\n",
              "      <td>0.375000</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'entropy', 'max_features': 'sqrt...</td>\n",
              "      <td>59.363168</td>\n",
              "      <td>102.141932</td>\n",
              "      <td>marker_Colorectal.txt</td>\n",
              "      <td>pca</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>78</th>\n",
              "      <td>0.793333</td>\n",
              "      <td>0.720000</td>\n",
              "      <td>0.400000</td>\n",
              "      <td>0.800000</td>\n",
              "      <td>0.533333</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 0.125, 'hidden_layer_sizes': (512, 5...</td>\n",
              "      <td>49.870998</td>\n",
              "      <td>120.945065</td>\n",
              "      <td>marker_Colorectal.txt</td>\n",
              "      <td>rp</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>79</th>\n",
              "      <td>0.793333</td>\n",
              "      <td>0.600000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 0.125, 'gamma': 0.0001220703125, 'kernel...</td>\n",
              "      <td>0.744758</td>\n",
              "      <td>120.945065</td>\n",
              "      <td>marker_Colorectal.txt</td>\n",
              "      <td>rp</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>80</th>\n",
              "      <td>0.733333</td>\n",
              "      <td>0.640000</td>\n",
              "      <td>0.200000</td>\n",
              "      <td>0.666667</td>\n",
              "      <td>0.307692</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'entropy', 'max_features': 'log2...</td>\n",
              "      <td>69.876427</td>\n",
              "      <td>120.945065</td>\n",
              "      <td>marker_Colorectal.txt</td>\n",
              "      <td>rp</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>81</th>\n",
              "      <td>0.806667</td>\n",
              "      <td>0.720000</td>\n",
              "      <td>0.400000</td>\n",
              "      <td>0.800000</td>\n",
              "      <td>0.533333</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 3.0517578125e-05, 'hidden_layer_size...</td>\n",
              "      <td>1673.659658</td>\n",
              "      <td>2080.805042</td>\n",
              "      <td>marker_Colorectal.txt</td>\n",
              "      <td>plain</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>82</th>\n",
              "      <td>0.780000</td>\n",
              "      <td>0.680000</td>\n",
              "      <td>0.400000</td>\n",
              "      <td>0.666667</td>\n",
              "      <td>0.500000</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 0.03125, 'kernel': 'linear'}</td>\n",
              "      <td>288.527356</td>\n",
              "      <td>2080.805042</td>\n",
              "      <td>marker_Colorectal.txt</td>\n",
              "      <td>plain</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>83</th>\n",
              "      <td>0.820000</td>\n",
              "      <td>0.840000</td>\n",
              "      <td>0.600000</td>\n",
              "      <td>1.000000</td>\n",
              "      <td>0.750000</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'gini', 'max_features': 'sqrt', ...</td>\n",
              "      <td>118.617998</td>\n",
              "      <td>2080.805042</td>\n",
              "      <td>marker_Colorectal.txt</td>\n",
              "      <td>plain</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>84</th>\n",
              "      <td>0.535294</td>\n",
              "      <td>0.772727</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 3.0517578125e-05, 'hidden_layer_size...</td>\n",
              "      <td>25.468925</td>\n",
              "      <td>103.051002</td>\n",
              "      <td>marker_IBD.txt</td>\n",
              "      <td>shallow_ae</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>85</th>\n",
              "      <td>0.523529</td>\n",
              "      <td>0.772727</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 2, 'gamma': 0.5, 'kernel': 'rbf'}</td>\n",
              "      <td>0.228910</td>\n",
              "      <td>103.051002</td>\n",
              "      <td>marker_IBD.txt</td>\n",
              "      <td>shallow_ae</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>86</th>\n",
              "      <td>0.488235</td>\n",
              "      <td>0.772727</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'gini', 'max_features': 'log2', ...</td>\n",
              "      <td>55.099191</td>\n",
              "      <td>103.051002</td>\n",
              "      <td>marker_IBD.txt</td>\n",
              "      <td>shallow_ae</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>87</th>\n",
              "      <td>0.558824</td>\n",
              "      <td>0.772727</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 0.0078125, 'hidden_layer_sizes': (25...</td>\n",
              "      <td>24.641263</td>\n",
              "      <td>285.467308</td>\n",
              "      <td>marker_IBD.txt</td>\n",
              "      <td>ae</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>88</th>\n",
              "      <td>0.588235</td>\n",
              "      <td>0.772727</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 32, 'gamma': 2, 'kernel': 'rbf'}</td>\n",
              "      <td>0.297684</td>\n",
              "      <td>285.467308</td>\n",
              "      <td>marker_IBD.txt</td>\n",
              "      <td>ae</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>89</th>\n",
              "      <td>0.511765</td>\n",
              "      <td>0.727273</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'entropy', 'max_features': 'sqrt...</td>\n",
              "      <td>58.754943</td>\n",
              "      <td>285.467308</td>\n",
              "      <td>marker_IBD.txt</td>\n",
              "      <td>ae</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>90</th>\n",
              "      <td>0.547059</td>\n",
              "      <td>0.727273</td>\n",
              "      <td>0.200000</td>\n",
              "      <td>0.333333</td>\n",
              "      <td>0.250000</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 3.0517578125e-05, 'hidden_layer_size...</td>\n",
              "      <td>1456.672904</td>\n",
              "      <td>3058.399564</td>\n",
              "      <td>marker_IBD.txt</td>\n",
              "      <td>cae</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>91</th>\n",
              "      <td>0.800000</td>\n",
              "      <td>0.727273</td>\n",
              "      <td>0.200000</td>\n",
              "      <td>0.333333</td>\n",
              "      <td>0.250000</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 0.03125, 'kernel': 'linear'}</td>\n",
              "      <td>1137.848792</td>\n",
              "      <td>3058.399564</td>\n",
              "      <td>marker_IBD.txt</td>\n",
              "      <td>cae</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>92</th>\n",
              "      <td>0.929412</td>\n",
              "      <td>0.727273</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'entropy', 'max_features': 'log2...</td>\n",
              "      <td>331.307379</td>\n",
              "      <td>3058.399564</td>\n",
              "      <td>marker_IBD.txt</td>\n",
              "      <td>cae</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>93</th>\n",
              "      <td>0.858824</td>\n",
              "      <td>0.818182</td>\n",
              "      <td>0.400000</td>\n",
              "      <td>0.666667</td>\n",
              "      <td>0.500000</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 0.00048828125, 'hidden_layer_sizes':...</td>\n",
              "      <td>40.435831</td>\n",
              "      <td>183.918113</td>\n",
              "      <td>marker_IBD.txt</td>\n",
              "      <td>vae</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>94</th>\n",
              "      <td>0.764706</td>\n",
              "      <td>0.772727</td>\n",
              "      <td>0.200000</td>\n",
              "      <td>0.500000</td>\n",
              "      <td>0.285714</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 32, 'gamma': 0.03125, 'kernel': 'rbf'}</td>\n",
              "      <td>0.346800</td>\n",
              "      <td>183.918113</td>\n",
              "      <td>marker_IBD.txt</td>\n",
              "      <td>vae</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>95</th>\n",
              "      <td>0.847059</td>\n",
              "      <td>0.772727</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'gini', 'max_features': 'sqrt', ...</td>\n",
              "      <td>58.136487</td>\n",
              "      <td>183.918113</td>\n",
              "      <td>marker_IBD.txt</td>\n",
              "      <td>vae</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>96</th>\n",
              "      <td>0.800000</td>\n",
              "      <td>0.772727</td>\n",
              "      <td>0.600000</td>\n",
              "      <td>0.500000</td>\n",
              "      <td>0.545455</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 0.5, 'hidden_layer_sizes': (128, 128)}</td>\n",
              "      <td>27.264308</td>\n",
              "      <td>87.200318</td>\n",
              "      <td>marker_IBD.txt</td>\n",
              "      <td>pca</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>97</th>\n",
              "      <td>0.764706</td>\n",
              "      <td>0.772727</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 0.03125, 'gamma': 0.001953125, 'kernel':...</td>\n",
              "      <td>1.467895</td>\n",
              "      <td>87.200318</td>\n",
              "      <td>marker_IBD.txt</td>\n",
              "      <td>pca</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>98</th>\n",
              "      <td>0.811765</td>\n",
              "      <td>0.772727</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'entropy', 'max_features': 'sqrt...</td>\n",
              "      <td>58.264915</td>\n",
              "      <td>87.200318</td>\n",
              "      <td>marker_IBD.txt</td>\n",
              "      <td>pca</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>99</th>\n",
              "      <td>0.705882</td>\n",
              "      <td>0.636364</td>\n",
              "      <td>0.200000</td>\n",
              "      <td>0.200000</td>\n",
              "      <td>0.200000</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 0.00048828125, 'hidden_layer_sizes':...</td>\n",
              "      <td>45.109154</td>\n",
              "      <td>112.673155</td>\n",
              "      <td>marker_IBD.txt</td>\n",
              "      <td>rp</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>100</th>\n",
              "      <td>0.788235</td>\n",
              "      <td>0.727273</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 2, 'gamma': 0.0001220703125, 'kernel': '...</td>\n",
              "      <td>0.618138</td>\n",
              "      <td>112.673155</td>\n",
              "      <td>marker_IBD.txt</td>\n",
              "      <td>rp</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>101</th>\n",
              "      <td>0.823529</td>\n",
              "      <td>0.772727</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'entropy', 'max_features': 'log2...</td>\n",
              "      <td>66.569057</td>\n",
              "      <td>112.673155</td>\n",
              "      <td>marker_IBD.txt</td>\n",
              "      <td>rp</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>102</th>\n",
              "      <td>0.776471</td>\n",
              "      <td>0.727273</td>\n",
              "      <td>0.600000</td>\n",
              "      <td>0.428571</td>\n",
              "      <td>0.500000</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 0.5, 'hidden_layer_sizes': (64, 64)}</td>\n",
              "      <td>2195.381789</td>\n",
              "      <td>2497.345249</td>\n",
              "      <td>marker_IBD.txt</td>\n",
              "      <td>plain</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>103</th>\n",
              "      <td>0.811765</td>\n",
              "      <td>0.727273</td>\n",
              "      <td>0.200000</td>\n",
              "      <td>0.333333</td>\n",
              "      <td>0.250000</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 0.03125, 'kernel': 'linear'}</td>\n",
              "      <td>207.017738</td>\n",
              "      <td>2497.345249</td>\n",
              "      <td>marker_IBD.txt</td>\n",
              "      <td>plain</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>104</th>\n",
              "      <td>0.823529</td>\n",
              "      <td>0.772727</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'gini', 'max_features': 'log2', ...</td>\n",
              "      <td>94.945693</td>\n",
              "      <td>2497.345249</td>\n",
              "      <td>marker_IBD.txt</td>\n",
              "      <td>plain</td>\n",
              "      <td>42</td>\n",
              "      <td>200</td>\n",
              "      <td>mps</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>105</th>\n",
              "      <td>0.500000</td>\n",
              "      <td>0.550000</td>\n",
              "      <td>1.000000</td>\n",
              "      <td>0.550000</td>\n",
              "      <td>0.709677</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 0.0001220703125, 'hidden_layer_sizes...</td>\n",
              "      <td>1405.755020</td>\n",
              "      <td>2495.791879</td>\n",
              "      <td>marker_WT2D.txt</td>\n",
              "      <td>cae</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>106</th>\n",
              "      <td>0.595960</td>\n",
              "      <td>0.600000</td>\n",
              "      <td>0.545455</td>\n",
              "      <td>0.666667</td>\n",
              "      <td>0.600000</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 0.03125, 'kernel': 'linear'}</td>\n",
              "      <td>834.236857</td>\n",
              "      <td>2495.791879</td>\n",
              "      <td>marker_WT2D.txt</td>\n",
              "      <td>cae</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>107</th>\n",
              "      <td>0.626263</td>\n",
              "      <td>0.500000</td>\n",
              "      <td>0.636364</td>\n",
              "      <td>0.538462</td>\n",
              "      <td>0.583333</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'entropy', 'max_features': 'log2...</td>\n",
              "      <td>254.800342</td>\n",
              "      <td>2495.791879</td>\n",
              "      <td>marker_WT2D.txt</td>\n",
              "      <td>cae</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>108</th>\n",
              "      <td>0.666667</td>\n",
              "      <td>0.600000</td>\n",
              "      <td>0.363636</td>\n",
              "      <td>0.800000</td>\n",
              "      <td>0.500000</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 0.001953125, 'hidden_layer_sizes': (...</td>\n",
              "      <td>2478.668654</td>\n",
              "      <td>2730.213370</td>\n",
              "      <td>marker_WT2D.txt</td>\n",
              "      <td>plain</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>109</th>\n",
              "      <td>0.636364</td>\n",
              "      <td>0.500000</td>\n",
              "      <td>0.545455</td>\n",
              "      <td>0.545455</td>\n",
              "      <td>0.545455</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 2, 'gamma': 0.0001220703125, 'kernel': '...</td>\n",
              "      <td>137.955329</td>\n",
              "      <td>2730.213370</td>\n",
              "      <td>marker_WT2D.txt</td>\n",
              "      <td>plain</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>110</th>\n",
              "      <td>0.525253</td>\n",
              "      <td>0.450000</td>\n",
              "      <td>0.636364</td>\n",
              "      <td>0.500000</td>\n",
              "      <td>0.560000</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'entropy', 'max_features': 'log2...</td>\n",
              "      <td>113.584896</td>\n",
              "      <td>2730.213370</td>\n",
              "      <td>marker_WT2D.txt</td>\n",
              "      <td>plain</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>111</th>\n",
              "      <td>0.500000</td>\n",
              "      <td>0.550000</td>\n",
              "      <td>1.000000</td>\n",
              "      <td>0.550000</td>\n",
              "      <td>0.709677</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 0.001953125, 'hidden_layer_sizes': (...</td>\n",
              "      <td>13.699150</td>\n",
              "      <td>92.950624</td>\n",
              "      <td>marker_WT2D.txt</td>\n",
              "      <td>shallow_ae</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>112</th>\n",
              "      <td>0.500000</td>\n",
              "      <td>0.550000</td>\n",
              "      <td>1.000000</td>\n",
              "      <td>0.550000</td>\n",
              "      <td>0.709677</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 8, 'gamma': 0.5, 'kernel': 'rbf'}</td>\n",
              "      <td>0.413084</td>\n",
              "      <td>92.950624</td>\n",
              "      <td>marker_WT2D.txt</td>\n",
              "      <td>shallow_ae</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>113</th>\n",
              "      <td>0.500000</td>\n",
              "      <td>0.550000</td>\n",
              "      <td>1.000000</td>\n",
              "      <td>0.550000</td>\n",
              "      <td>0.709677</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'gini', 'max_features': 'sqrt', ...</td>\n",
              "      <td>78.685357</td>\n",
              "      <td>92.950624</td>\n",
              "      <td>marker_WT2D.txt</td>\n",
              "      <td>shallow_ae</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>114</th>\n",
              "      <td>0.515152</td>\n",
              "      <td>0.550000</td>\n",
              "      <td>1.000000</td>\n",
              "      <td>0.550000</td>\n",
              "      <td>0.709677</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 0.125, 'hidden_layer_sizes': (64, 64)}</td>\n",
              "      <td>22.683855</td>\n",
              "      <td>107.277794</td>\n",
              "      <td>marker_WT2D.txt</td>\n",
              "      <td>ae</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>115</th>\n",
              "      <td>0.717172</td>\n",
              "      <td>0.550000</td>\n",
              "      <td>1.000000</td>\n",
              "      <td>0.550000</td>\n",
              "      <td>0.709677</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 0.03125, 'gamma': 0.001953125, 'kernel':...</td>\n",
              "      <td>0.406561</td>\n",
              "      <td>107.277794</td>\n",
              "      <td>marker_WT2D.txt</td>\n",
              "      <td>ae</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>116</th>\n",
              "      <td>0.393939</td>\n",
              "      <td>0.500000</td>\n",
              "      <td>0.818182</td>\n",
              "      <td>0.529412</td>\n",
              "      <td>0.642857</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'entropy', 'max_features': 'log2...</td>\n",
              "      <td>80.390760</td>\n",
              "      <td>107.277794</td>\n",
              "      <td>marker_WT2D.txt</td>\n",
              "      <td>ae</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>117</th>\n",
              "      <td>0.237374</td>\n",
              "      <td>0.550000</td>\n",
              "      <td>1.000000</td>\n",
              "      <td>0.550000</td>\n",
              "      <td>0.709677</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 0.03125, 'hidden_layer_sizes': (64, ...</td>\n",
              "      <td>16.322586</td>\n",
              "      <td>95.779552</td>\n",
              "      <td>marker_WT2D.txt</td>\n",
              "      <td>vae</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>118</th>\n",
              "      <td>0.500000</td>\n",
              "      <td>0.550000</td>\n",
              "      <td>1.000000</td>\n",
              "      <td>0.550000</td>\n",
              "      <td>0.709677</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 0.03125, 'gamma': 0.00048828125, 'kernel...</td>\n",
              "      <td>0.387627</td>\n",
              "      <td>95.779552</td>\n",
              "      <td>marker_WT2D.txt</td>\n",
              "      <td>vae</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>119</th>\n",
              "      <td>0.500000</td>\n",
              "      <td>0.550000</td>\n",
              "      <td>1.000000</td>\n",
              "      <td>0.550000</td>\n",
              "      <td>0.709677</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'gini', 'max_features': 'sqrt', ...</td>\n",
              "      <td>78.629580</td>\n",
              "      <td>95.779552</td>\n",
              "      <td>marker_WT2D.txt</td>\n",
              "      <td>vae</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>120</th>\n",
              "      <td>0.626263</td>\n",
              "      <td>0.600000</td>\n",
              "      <td>0.636364</td>\n",
              "      <td>0.636364</td>\n",
              "      <td>0.636364</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 3.0517578125e-05, 'hidden_layer_size...</td>\n",
              "      <td>25.624079</td>\n",
              "      <td>120.230908</td>\n",
              "      <td>marker_WT2D.txt</td>\n",
              "      <td>pca</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>121</th>\n",
              "      <td>0.545455</td>\n",
              "      <td>0.550000</td>\n",
              "      <td>0.454545</td>\n",
              "      <td>0.625000</td>\n",
              "      <td>0.526316</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 8, 'gamma': 0.0001220703125, 'kernel': '...</td>\n",
              "      <td>12.671258</td>\n",
              "      <td>120.230908</td>\n",
              "      <td>marker_WT2D.txt</td>\n",
              "      <td>pca</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>122</th>\n",
              "      <td>0.479798</td>\n",
              "      <td>0.400000</td>\n",
              "      <td>0.363636</td>\n",
              "      <td>0.444444</td>\n",
              "      <td>0.400000</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'entropy', 'max_features': 'sqrt...</td>\n",
              "      <td>81.428668</td>\n",
              "      <td>120.230908</td>\n",
              "      <td>marker_WT2D.txt</td>\n",
              "      <td>pca</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>123</th>\n",
              "      <td>0.474747</td>\n",
              "      <td>0.450000</td>\n",
              "      <td>0.454545</td>\n",
              "      <td>0.500000</td>\n",
              "      <td>0.476190</td>\n",
              "      <td>MLP</td>\n",
              "      <td>{'alpha': 0.00048828125, 'hidden_layer_sizes':...</td>\n",
              "      <td>38.448539</td>\n",
              "      <td>126.877143</td>\n",
              "      <td>marker_WT2D.txt</td>\n",
              "      <td>rp</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>124</th>\n",
              "      <td>0.303030</td>\n",
              "      <td>0.550000</td>\n",
              "      <td>1.000000</td>\n",
              "      <td>0.550000</td>\n",
              "      <td>0.709677</td>\n",
              "      <td>SVM</td>\n",
              "      <td>{'C': 0.03125, 'gamma': 0.00048828125, 'kernel...</td>\n",
              "      <td>0.625493</td>\n",
              "      <td>126.877143</td>\n",
              "      <td>marker_WT2D.txt</td>\n",
              "      <td>rp</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>125</th>\n",
              "      <td>0.737374</td>\n",
              "      <td>0.650000</td>\n",
              "      <td>0.909091</td>\n",
              "      <td>0.625000</td>\n",
              "      <td>0.740741</td>\n",
              "      <td>Random Forest</td>\n",
              "      <td>{'criterion': 'gini', 'max_features': 'log2', ...</td>\n",
              "      <td>86.934503</td>\n",
              "      <td>126.877143</td>\n",
              "      <td>marker_WT2D.txt</td>\n",
              "      <td>rp</td>\n",
              "      <td>42</td>\n",
              "      <td>2000</td>\n",
              "      <td>cpu</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "          AUC       ACC    Recall  Precision  F1_score classification_method  \\\n",
              "0    0.947464  0.829787  0.791667   0.863636  0.826087                   MLP   \n",
              "1    0.916667  0.872340  0.875000   0.875000  0.875000                   SVM   \n",
              "2    0.943841  0.829787  0.791667   0.863636  0.826087         Random Forest   \n",
              "3    0.926630  0.829787  0.750000   0.900000  0.818182                   MLP   \n",
              "4    0.926630  0.851064  0.750000   0.947368  0.837209                   SVM   \n",
              "5    0.902174  0.829787  0.916667   0.785714  0.846154         Random Forest   \n",
              "6    0.927536  0.829787  0.666667   1.000000  0.800000                   MLP   \n",
              "7    0.981884  0.893617  0.875000   0.913043  0.893617                   SVM   \n",
              "8    0.956522  0.851064  0.750000   0.947368  0.837209         Random Forest   \n",
              "9    0.934783  0.808511  0.750000   0.857143  0.800000                   MLP   \n",
              "10   0.951087  0.829787  0.708333   0.944444  0.809524                   SVM   \n",
              "11   0.936594  0.851064  0.791667   0.904762  0.844444         Random Forest   \n",
              "12   0.922101  0.829787  0.708333   0.944444  0.809524                   MLP   \n",
              "13   0.971014  0.936170  0.875000   1.000000  0.933333                   SVM   \n",
              "14   0.934783  0.872340  0.791667   0.950000  0.863636         Random Forest   \n",
              "15   0.980072  0.872340  0.875000   0.875000  0.875000                   MLP   \n",
              "16   0.974638  0.914894  0.833333   1.000000  0.909091                   SVM   \n",
              "17   0.965580  0.914894  0.875000   0.954545  0.913043         Random Forest   \n",
              "18   0.989130  0.829787  1.000000   0.750000  0.857143                   MLP   \n",
              "19   0.983696  0.914894  0.875000   0.954545  0.913043                   SVM   \n",
              "20   0.992754  0.914894  0.833333   1.000000  0.909091         Random Forest   \n",
              "21   0.500000  0.352941  0.000000   0.000000  0.000000                   MLP   \n",
              "22   0.592593  0.607843  0.757576   0.675676  0.714286                   SVM   \n",
              "23   0.640572  0.627451  0.969697   0.640000  0.771084         Random Forest   \n",
              "24   0.560606  0.450980  0.424242   0.608696  0.500000                   MLP   \n",
              "25   0.681818  0.647059  1.000000   0.647059  0.785714                   SVM   \n",
              "26   0.739057  0.647059  1.000000   0.647059  0.785714         Random Forest   \n",
              "27   0.500000  0.647059  1.000000   0.647059  0.785714                   MLP   \n",
              "28   0.500000  0.647059  1.000000   0.647059  0.785714                   SVM   \n",
              "29   0.500000  0.647059  1.000000   0.647059  0.785714         Random Forest   \n",
              "30   0.439394  0.647059  1.000000   0.647059  0.785714                   MLP   \n",
              "31   0.524411  0.647059  1.000000   0.647059  0.785714                   SVM   \n",
              "32   0.524411  0.627451  0.969697   0.640000  0.771084         Random Forest   \n",
              "33   0.370370  0.647059  1.000000   0.647059  0.785714                   MLP   \n",
              "34   0.526936  0.647059  1.000000   0.647059  0.785714                   SVM   \n",
              "35   0.501684  0.607843  0.878788   0.644444  0.743590         Random Forest   \n",
              "36   0.673401  0.627451  0.727273   0.705882  0.716418                   MLP   \n",
              "37   0.654882  0.647059  0.969697   0.653061  0.780488                   SVM   \n",
              "38   0.675084  0.686275  0.969697   0.680851  0.800000         Random Forest   \n",
              "39   0.646465  0.686275  0.909091   0.697674  0.789474                   MLP   \n",
              "40   0.326599  0.607843  0.666667   0.709677  0.687500                   SVM   \n",
              "41   0.685185  0.647059  1.000000   0.647059  0.785714         Random Forest   \n",
              "42   0.500000  0.507246  0.000000   0.000000  0.000000                   MLP   \n",
              "43   0.801681  0.753623  0.794118   0.729730  0.760563                   SVM   \n",
              "44   0.755462  0.710145  0.794118   0.675000  0.729730         Random Forest   \n",
              "45   0.807563  0.739130  0.764706   0.722222  0.742857                   MLP   \n",
              "46   0.771429  0.753623  0.823529   0.717949  0.767123                   SVM   \n",
              "47   0.727731  0.623188  0.617647   0.617647  0.617647         Random Forest   \n",
              "48   0.546639  0.507246  0.000000   0.000000  0.000000                   MLP   \n",
              "49   0.485294  0.507246  0.000000   0.000000  0.000000                   SVM   \n",
              "50   0.500000  0.507246  0.000000   0.000000  0.000000         Random Forest   \n",
              "51   0.452521  0.492754  1.000000   0.492754  0.660194                   MLP   \n",
              "52   0.397479  0.507246  0.000000   0.000000  0.000000                   SVM   \n",
              "53   0.697899  0.608696  0.264706   0.818182  0.400000         Random Forest   \n",
              "54   0.502521  0.507246  0.000000   0.000000  0.000000                   MLP   \n",
              "55   0.500840  0.507246  0.000000   0.000000  0.000000                   SVM   \n",
              "56   0.500840  0.507246  0.029412   0.500000  0.055556         Random Forest   \n",
              "57   0.608403  0.579710  0.617647   0.567568  0.591549                   MLP   \n",
              "58   0.622689  0.579710  0.588235   0.571429  0.579710                   SVM   \n",
              "59   0.625630  0.594203  0.558824   0.593750  0.575758         Random Forest   \n",
              "60   0.567227  0.536232  0.470588   0.533333  0.500000                   MLP   \n",
              "61   0.610084  0.579710  0.529412   0.580645  0.553846                   SVM   \n",
              "62   0.640336  0.623188  0.500000   0.653846  0.566667         Random Forest   \n",
              "63   0.516667  0.600000  0.000000   0.000000  0.000000                   MLP   \n",
              "64   0.550000  0.600000  0.000000   0.000000  0.000000                   SVM   \n",
              "65   0.583333  0.600000  0.000000   0.000000  0.000000         Random Forest   \n",
              "66   0.500000  0.600000  0.000000   0.000000  0.000000                   MLP   \n",
              "67   0.500000  0.600000  0.000000   0.000000  0.000000                   SVM   \n",
              "68   0.500000  0.600000  0.000000   0.000000  0.000000         Random Forest   \n",
              "69   0.740000  0.800000  0.500000   1.000000  0.666667                   MLP   \n",
              "70   0.773333  0.720000  0.500000   0.714286  0.588235                   SVM   \n",
              "71   0.753333  0.680000  0.400000   0.666667  0.500000         Random Forest   \n",
              "72   0.640000  0.640000  0.500000   0.555556  0.526316                   MLP   \n",
              "73   0.653333  0.640000  0.100000   1.000000  0.181818                   SVM   \n",
              "74   0.540000  0.640000  0.100000   1.000000  0.181818         Random Forest   \n",
              "75   0.846667  0.760000  0.800000   0.666667  0.727273                   MLP   \n",
              "76   0.793333  0.760000  0.500000   0.833333  0.625000                   SVM   \n",
              "77   0.766667  0.600000  0.300000   0.500000  0.375000         Random Forest   \n",
              "78   0.793333  0.720000  0.400000   0.800000  0.533333                   MLP   \n",
              "79   0.793333  0.600000  0.000000   0.000000  0.000000                   SVM   \n",
              "80   0.733333  0.640000  0.200000   0.666667  0.307692         Random Forest   \n",
              "81   0.806667  0.720000  0.400000   0.800000  0.533333                   MLP   \n",
              "82   0.780000  0.680000  0.400000   0.666667  0.500000                   SVM   \n",
              "83   0.820000  0.840000  0.600000   1.000000  0.750000         Random Forest   \n",
              "84   0.535294  0.772727  0.000000   0.000000  0.000000                   MLP   \n",
              "85   0.523529  0.772727  0.000000   0.000000  0.000000                   SVM   \n",
              "86   0.488235  0.772727  0.000000   0.000000  0.000000         Random Forest   \n",
              "87   0.558824  0.772727  0.000000   0.000000  0.000000                   MLP   \n",
              "88   0.588235  0.772727  0.000000   0.000000  0.000000                   SVM   \n",
              "89   0.511765  0.727273  0.000000   0.000000  0.000000         Random Forest   \n",
              "90   0.547059  0.727273  0.200000   0.333333  0.250000                   MLP   \n",
              "91   0.800000  0.727273  0.200000   0.333333  0.250000                   SVM   \n",
              "92   0.929412  0.727273  0.000000   0.000000  0.000000         Random Forest   \n",
              "93   0.858824  0.818182  0.400000   0.666667  0.500000                   MLP   \n",
              "94   0.764706  0.772727  0.200000   0.500000  0.285714                   SVM   \n",
              "95   0.847059  0.772727  0.000000   0.000000  0.000000         Random Forest   \n",
              "96   0.800000  0.772727  0.600000   0.500000  0.545455                   MLP   \n",
              "97   0.764706  0.772727  0.000000   0.000000  0.000000                   SVM   \n",
              "98   0.811765  0.772727  0.000000   0.000000  0.000000         Random Forest   \n",
              "99   0.705882  0.636364  0.200000   0.200000  0.200000                   MLP   \n",
              "100  0.788235  0.727273  0.000000   0.000000  0.000000                   SVM   \n",
              "101  0.823529  0.772727  0.000000   0.000000  0.000000         Random Forest   \n",
              "102  0.776471  0.727273  0.600000   0.428571  0.500000                   MLP   \n",
              "103  0.811765  0.727273  0.200000   0.333333  0.250000                   SVM   \n",
              "104  0.823529  0.772727  0.000000   0.000000  0.000000         Random Forest   \n",
              "105  0.500000  0.550000  1.000000   0.550000  0.709677                   MLP   \n",
              "106  0.595960  0.600000  0.545455   0.666667  0.600000                   SVM   \n",
              "107  0.626263  0.500000  0.636364   0.538462  0.583333         Random Forest   \n",
              "108  0.666667  0.600000  0.363636   0.800000  0.500000                   MLP   \n",
              "109  0.636364  0.500000  0.545455   0.545455  0.545455                   SVM   \n",
              "110  0.525253  0.450000  0.636364   0.500000  0.560000         Random Forest   \n",
              "111  0.500000  0.550000  1.000000   0.550000  0.709677                   MLP   \n",
              "112  0.500000  0.550000  1.000000   0.550000  0.709677                   SVM   \n",
              "113  0.500000  0.550000  1.000000   0.550000  0.709677         Random Forest   \n",
              "114  0.515152  0.550000  1.000000   0.550000  0.709677                   MLP   \n",
              "115  0.717172  0.550000  1.000000   0.550000  0.709677                   SVM   \n",
              "116  0.393939  0.500000  0.818182   0.529412  0.642857         Random Forest   \n",
              "117  0.237374  0.550000  1.000000   0.550000  0.709677                   MLP   \n",
              "118  0.500000  0.550000  1.000000   0.550000  0.709677                   SVM   \n",
              "119  0.500000  0.550000  1.000000   0.550000  0.709677         Random Forest   \n",
              "120  0.626263  0.600000  0.636364   0.636364  0.636364                   MLP   \n",
              "121  0.545455  0.550000  0.454545   0.625000  0.526316                   SVM   \n",
              "122  0.479798  0.400000  0.363636   0.444444  0.400000         Random Forest   \n",
              "123  0.474747  0.450000  0.454545   0.500000  0.476190                   MLP   \n",
              "124  0.303030  0.550000  1.000000   0.550000  0.709677                   SVM   \n",
              "125  0.737374  0.650000  0.909091   0.625000  0.740741         Random Forest   \n",
              "\n",
              "                                      hyper_parameters  classification_time  \\\n",
              "0    {'alpha': 0.0001220703125, 'hidden_layer_sizes...            66.259109   \n",
              "1       {'C': 32, 'gamma': 0.0078125, 'kernel': 'rbf'}             1.120851   \n",
              "2    {'criterion': 'entropy', 'max_features': 'sqrt...           111.178497   \n",
              "3    {'alpha': 3.0517578125e-05, 'hidden_layer_size...            64.426869   \n",
              "4                         {'C': 2, 'kernel': 'linear'}             1.863737   \n",
              "5    {'criterion': 'gini', 'max_features': 'log2', ...           119.985000   \n",
              "6    {'alpha': 0.0001220703125, 'hidden_layer_sizes...          2337.573239   \n",
              "7                   {'C': 0.03125, 'kernel': 'linear'}          8910.858763   \n",
              "8    {'criterion': 'entropy', 'max_features': 'sqrt...          2144.996647   \n",
              "9    {'alpha': 0.0078125, 'hidden_layer_sizes': (25...           111.703098   \n",
              "10   {'C': 2, 'gamma': 0.0001220703125, 'kernel': '...             1.306651   \n",
              "11   {'criterion': 'entropy', 'max_features': 'sqrt...           124.495656   \n",
              "12   {'alpha': 3.0517578125e-05, 'hidden_layer_size...            81.842762   \n",
              "13   {'C': 8, 'gamma': 0.0001220703125, 'kernel': '...            30.035022   \n",
              "14   {'criterion': 'gini', 'max_features': 'log2', ...           131.445379   \n",
              "15    {'alpha': 0.5, 'hidden_layer_sizes': (256, 256)}           123.187775   \n",
              "16   {'C': 0.5, 'gamma': 0.0001220703125, 'kernel':...             4.674169   \n",
              "17   {'criterion': 'gini', 'max_features': 'log2', ...           186.715222   \n",
              "18   {'alpha': 0.0001220703125, 'hidden_layer_sizes...          5058.035164   \n",
              "19                  {'C': 0.03125, 'kernel': 'linear'}          2277.429844   \n",
              "20   {'criterion': 'gini', 'max_features': 'log2', ...           310.007509   \n",
              "21   {'alpha': 3.0517578125e-05, 'hidden_layer_size...          1486.891025   \n",
              "22                  {'C': 0.03125, 'kernel': 'linear'}          7005.718553   \n",
              "23   {'criterion': 'entropy', 'max_features': 'sqrt...           918.833811   \n",
              "24   {'alpha': 0.03125, 'hidden_layer_sizes': (128,...          2762.416396   \n",
              "25   {'C': 0.125, 'gamma': 0.0001220703125, 'kernel...          1190.073292   \n",
              "26   {'criterion': 'gini', 'max_features': 'log2', ...           194.442154   \n",
              "27   {'alpha': 3.0517578125e-05, 'hidden_layer_size...            16.176070   \n",
              "28       {'C': 0.125, 'gamma': 0.125, 'kernel': 'rbf'}             0.634451   \n",
              "29   {'criterion': 'gini', 'max_features': 'sqrt', ...            83.237592   \n",
              "30   {'alpha': 0.00048828125, 'hidden_layer_sizes':...            23.258092   \n",
              "31                  {'C': 0.03125, 'kernel': 'linear'}             0.865882   \n",
              "32   {'criterion': 'gini', 'max_features': 'sqrt', ...            89.596979   \n",
              "33   {'alpha': 0.001953125, 'hidden_layer_sizes': (...            25.970036   \n",
              "34                    {'C': 0.125, 'kernel': 'linear'}             3.912843   \n",
              "35   {'criterion': 'entropy', 'max_features': 'sqrt...            60.502643   \n",
              "36   {'alpha': 0.03125, 'hidden_layer_sizes': (64, ...            45.460609   \n",
              "37                       {'C': 32, 'kernel': 'linear'}           292.242113   \n",
              "38   {'criterion': 'gini', 'max_features': 'log2', ...            74.022311   \n",
              "39      {'alpha': 0.5, 'hidden_layer_sizes': (64, 64)}            62.771949   \n",
              "40                  {'C': 0.03125, 'kernel': 'linear'}             5.681792   \n",
              "41   {'criterion': 'gini', 'max_features': 'sqrt', ...           121.447971   \n",
              "42   {'alpha': 0.0001220703125, 'hidden_layer_sizes...          2492.601276   \n",
              "43                  {'C': 0.03125, 'kernel': 'linear'}         10209.280280   \n",
              "44   {'criterion': 'entropy', 'max_features': 'sqrt...          1409.363473   \n",
              "45   {'alpha': 0.00048828125, 'hidden_layer_sizes':...          5410.766285   \n",
              "46                  {'C': 0.03125, 'kernel': 'linear'}          2465.565559   \n",
              "47   {'criterion': 'gini', 'max_features': 'log2', ...           254.427690   \n",
              "48   {'alpha': 0.0078125, 'hidden_layer_sizes': (64...            18.924324   \n",
              "49   {'C': 0.03125, 'gamma': 0.001953125, 'kernel':...             1.156454   \n",
              "50   {'criterion': 'gini', 'max_features': 'sqrt', ...            80.019774   \n",
              "51   {'alpha': 3.0517578125e-05, 'hidden_layer_size...            28.255768   \n",
              "52                  {'C': 0.03125, 'kernel': 'linear'}             1.663867   \n",
              "53   {'criterion': 'gini', 'max_features': 'sqrt', ...            96.230592   \n",
              "54   {'alpha': 3.0517578125e-05, 'hidden_layer_size...            17.820294   \n",
              "55               {'C': 8, 'gamma': 8, 'kernel': 'rbf'}             1.023353   \n",
              "56   {'criterion': 'gini', 'max_features': 'sqrt', ...            80.574382   \n",
              "57   {'alpha': 0.00048828125, 'hidden_layer_sizes':...            61.699997   \n",
              "58   {'C': 0.5, 'gamma': 0.001953125, 'kernel': 'rbf'}            71.348738   \n",
              "59   {'criterion': 'entropy', 'max_features': 'log2...            94.713534   \n",
              "60   {'alpha': 0.0001220703125, 'hidden_layer_sizes...            80.663704   \n",
              "61   {'C': 2, 'gamma': 0.0001220703125, 'kernel': '...             3.265329   \n",
              "62   {'criterion': 'gini', 'max_features': 'log2', ...           134.022436   \n",
              "63   {'alpha': 0.0078125, 'hidden_layer_sizes': (25...            22.348105   \n",
              "64   {'C': 32, 'gamma': 0.0001220703125, 'kernel': ...             0.305911   \n",
              "65   {'criterion': 'gini', 'max_features': 'sqrt', ...            54.457502   \n",
              "66   {'alpha': 0.001953125, 'hidden_layer_sizes': (...            21.643693   \n",
              "67   {'C': 0.03125, 'gamma': 0.00048828125, 'kernel...             0.401798   \n",
              "68   {'criterion': 'gini', 'max_features': 'sqrt', ...            55.509926   \n",
              "69   {'alpha': 0.0001220703125, 'hidden_layer_sizes...          1359.514105   \n",
              "70                  {'C': 0.03125, 'kernel': 'linear'}          1354.391862   \n",
              "71   {'criterion': 'gini', 'max_features': 'sqrt', ...           521.247247   \n",
              "72    {'alpha': 0.5, 'hidden_layer_sizes': (256, 256)}            40.422624   \n",
              "73       {'C': 0.5, 'gamma': 0.03125, 'kernel': 'rbf'}             0.399136   \n",
              "74   {'criterion': 'entropy', 'max_features': 'sqrt...            59.198460   \n",
              "75   {'alpha': 3.0517578125e-05, 'hidden_layer_size...            30.598888   \n",
              "76   {'C': 32, 'gamma': 0.00048828125, 'kernel': 'r...            11.931073   \n",
              "77   {'criterion': 'entropy', 'max_features': 'sqrt...            59.363168   \n",
              "78   {'alpha': 0.125, 'hidden_layer_sizes': (512, 5...            49.870998   \n",
              "79   {'C': 0.125, 'gamma': 0.0001220703125, 'kernel...             0.744758   \n",
              "80   {'criterion': 'entropy', 'max_features': 'log2...            69.876427   \n",
              "81   {'alpha': 3.0517578125e-05, 'hidden_layer_size...          1673.659658   \n",
              "82                  {'C': 0.03125, 'kernel': 'linear'}           288.527356   \n",
              "83   {'criterion': 'gini', 'max_features': 'sqrt', ...           118.617998   \n",
              "84   {'alpha': 3.0517578125e-05, 'hidden_layer_size...            25.468925   \n",
              "85             {'C': 2, 'gamma': 0.5, 'kernel': 'rbf'}             0.228910   \n",
              "86   {'criterion': 'gini', 'max_features': 'log2', ...            55.099191   \n",
              "87   {'alpha': 0.0078125, 'hidden_layer_sizes': (25...            24.641263   \n",
              "88              {'C': 32, 'gamma': 2, 'kernel': 'rbf'}             0.297684   \n",
              "89   {'criterion': 'entropy', 'max_features': 'sqrt...            58.754943   \n",
              "90   {'alpha': 3.0517578125e-05, 'hidden_layer_size...          1456.672904   \n",
              "91                  {'C': 0.03125, 'kernel': 'linear'}          1137.848792   \n",
              "92   {'criterion': 'entropy', 'max_features': 'log2...           331.307379   \n",
              "93   {'alpha': 0.00048828125, 'hidden_layer_sizes':...            40.435831   \n",
              "94        {'C': 32, 'gamma': 0.03125, 'kernel': 'rbf'}             0.346800   \n",
              "95   {'criterion': 'gini', 'max_features': 'sqrt', ...            58.136487   \n",
              "96    {'alpha': 0.5, 'hidden_layer_sizes': (128, 128)}            27.264308   \n",
              "97   {'C': 0.03125, 'gamma': 0.001953125, 'kernel':...             1.467895   \n",
              "98   {'criterion': 'entropy', 'max_features': 'sqrt...            58.264915   \n",
              "99   {'alpha': 0.00048828125, 'hidden_layer_sizes':...            45.109154   \n",
              "100  {'C': 2, 'gamma': 0.0001220703125, 'kernel': '...             0.618138   \n",
              "101  {'criterion': 'entropy', 'max_features': 'log2...            66.569057   \n",
              "102     {'alpha': 0.5, 'hidden_layer_sizes': (64, 64)}          2195.381789   \n",
              "103                 {'C': 0.03125, 'kernel': 'linear'}           207.017738   \n",
              "104  {'criterion': 'gini', 'max_features': 'log2', ...            94.945693   \n",
              "105  {'alpha': 0.0001220703125, 'hidden_layer_sizes...          1405.755020   \n",
              "106                 {'C': 0.03125, 'kernel': 'linear'}           834.236857   \n",
              "107  {'criterion': 'entropy', 'max_features': 'log2...           254.800342   \n",
              "108  {'alpha': 0.001953125, 'hidden_layer_sizes': (...          2478.668654   \n",
              "109  {'C': 2, 'gamma': 0.0001220703125, 'kernel': '...           137.955329   \n",
              "110  {'criterion': 'entropy', 'max_features': 'log2...           113.584896   \n",
              "111  {'alpha': 0.001953125, 'hidden_layer_sizes': (...            13.699150   \n",
              "112            {'C': 8, 'gamma': 0.5, 'kernel': 'rbf'}             0.413084   \n",
              "113  {'criterion': 'gini', 'max_features': 'sqrt', ...            78.685357   \n",
              "114   {'alpha': 0.125, 'hidden_layer_sizes': (64, 64)}            22.683855   \n",
              "115  {'C': 0.03125, 'gamma': 0.001953125, 'kernel':...             0.406561   \n",
              "116  {'criterion': 'entropy', 'max_features': 'log2...            80.390760   \n",
              "117  {'alpha': 0.03125, 'hidden_layer_sizes': (64, ...            16.322586   \n",
              "118  {'C': 0.03125, 'gamma': 0.00048828125, 'kernel...             0.387627   \n",
              "119  {'criterion': 'gini', 'max_features': 'sqrt', ...            78.629580   \n",
              "120  {'alpha': 3.0517578125e-05, 'hidden_layer_size...            25.624079   \n",
              "121  {'C': 8, 'gamma': 0.0001220703125, 'kernel': '...            12.671258   \n",
              "122  {'criterion': 'entropy', 'max_features': 'sqrt...            81.428668   \n",
              "123  {'alpha': 0.00048828125, 'hidden_layer_sizes':...            38.448539   \n",
              "124  {'C': 0.03125, 'gamma': 0.00048828125, 'kernel...             0.625493   \n",
              "125  {'criterion': 'gini', 'max_features': 'log2', ...            86.934503   \n",
              "\n",
              "     experiment_time                   data       model  seed  max_epoch_num  \\\n",
              "0         238.745647   marker_Cirrhosis.txt  shallow_ae    42            200   \n",
              "1         238.745647   marker_Cirrhosis.txt  shallow_ae    42            200   \n",
              "2         238.745647   marker_Cirrhosis.txt  shallow_ae    42            200   \n",
              "3         730.488681   marker_Cirrhosis.txt          ae    42            200   \n",
              "4         730.488681   marker_Cirrhosis.txt          ae    42            200   \n",
              "5         730.488681   marker_Cirrhosis.txt          ae    42            200   \n",
              "6       13683.493345   marker_Cirrhosis.txt         cae    42            200   \n",
              "7       13683.493345   marker_Cirrhosis.txt         cae    42            200   \n",
              "8       13683.493345   marker_Cirrhosis.txt         cae    42            200   \n",
              "9         433.025449   marker_Cirrhosis.txt         vae    42            200   \n",
              "10        433.025449   marker_Cirrhosis.txt         vae    42            200   \n",
              "11        433.025449   marker_Cirrhosis.txt         vae    42            200   \n",
              "12        243.937610   marker_Cirrhosis.txt         pca    42            200   \n",
              "13        243.937610   marker_Cirrhosis.txt         pca    42            200   \n",
              "14        243.937610   marker_Cirrhosis.txt         pca    42            200   \n",
              "15        315.226107   marker_Cirrhosis.txt          rp    42            200   \n",
              "16        315.226107   marker_Cirrhosis.txt          rp    42            200   \n",
              "17        315.226107   marker_Cirrhosis.txt          rp    42            200   \n",
              "18       7645.472544   marker_Cirrhosis.txt       plain    42            200   \n",
              "19       7645.472544   marker_Cirrhosis.txt       plain    42            200   \n",
              "20       7645.472544   marker_Cirrhosis.txt       plain    42            200   \n",
              "21       9414.390030     marker_Obesity.txt         cae    42           2000   \n",
              "22       9414.390030     marker_Obesity.txt         cae    42           2000   \n",
              "23       9414.390030     marker_Obesity.txt         cae    42           2000   \n",
              "24       4146.935735     marker_Obesity.txt       plain    42           2000   \n",
              "25       4146.935735     marker_Obesity.txt       plain    42           2000   \n",
              "26       4146.935735     marker_Obesity.txt       plain    42           2000   \n",
              "27        100.520304     marker_Obesity.txt  shallow_ae    42           2000   \n",
              "28        100.520304     marker_Obesity.txt  shallow_ae    42           2000   \n",
              "29        100.520304     marker_Obesity.txt  shallow_ae    42           2000   \n",
              "30        118.358175     marker_Obesity.txt          ae    42           2000   \n",
              "31        118.358175     marker_Obesity.txt          ae    42           2000   \n",
              "32        118.358175     marker_Obesity.txt          ae    42           2000   \n",
              "33         90.790034     marker_Obesity.txt         vae    42           2000   \n",
              "34         90.790034     marker_Obesity.txt         vae    42           2000   \n",
              "35         90.790034     marker_Obesity.txt         vae    42           2000   \n",
              "36        412.816045     marker_Obesity.txt         pca    42           2000   \n",
              "37        412.816045     marker_Obesity.txt         pca    42           2000   \n",
              "38        412.816045     marker_Obesity.txt         pca    42           2000   \n",
              "39        191.006667     marker_Obesity.txt          rp    42           2000   \n",
              "40        191.006667     marker_Obesity.txt          rp    42           2000   \n",
              "41        191.006667     marker_Obesity.txt          rp    42           2000   \n",
              "42      14115.473664         marker_T2D.txt         cae    42           2000   \n",
              "43      14115.473664         marker_T2D.txt         cae    42           2000   \n",
              "44      14115.473664         marker_T2D.txt         cae    42           2000   \n",
              "45       8130.763001         marker_T2D.txt       plain    42           2000   \n",
              "46       8130.763001         marker_T2D.txt       plain    42           2000   \n",
              "47       8130.763001         marker_T2D.txt       plain    42           2000   \n",
              "48        100.779192         marker_T2D.txt  shallow_ae    42           2000   \n",
              "49        100.779192         marker_T2D.txt  shallow_ae    42           2000   \n",
              "50        100.779192         marker_T2D.txt  shallow_ae    42           2000   \n",
              "51        132.101581         marker_T2D.txt          ae    42           2000   \n",
              "52        132.101581         marker_T2D.txt          ae    42           2000   \n",
              "53        132.101581         marker_T2D.txt          ae    42           2000   \n",
              "54        100.809295         marker_T2D.txt         vae    42           2000   \n",
              "55        100.809295         marker_T2D.txt         vae    42           2000   \n",
              "56        100.809295         marker_T2D.txt         vae    42           2000   \n",
              "57        229.635357         marker_T2D.txt         pca    42           2000   \n",
              "58        229.635357         marker_T2D.txt         pca    42           2000   \n",
              "59        229.635357         marker_T2D.txt         pca    42           2000   \n",
              "60        219.272567         marker_T2D.txt          rp    42           2000   \n",
              "61        219.272567         marker_T2D.txt          rp    42           2000   \n",
              "62        219.272567         marker_T2D.txt          rp    42           2000   \n",
              "63        102.883997  marker_Colorectal.txt  shallow_ae    42            200   \n",
              "64        102.883997  marker_Colorectal.txt  shallow_ae    42            200   \n",
              "65        102.883997  marker_Colorectal.txt  shallow_ae    42            200   \n",
              "66        316.989440  marker_Colorectal.txt          ae    42            200   \n",
              "67        316.989440  marker_Colorectal.txt          ae    42            200   \n",
              "68        316.989440  marker_Colorectal.txt          ae    42            200   \n",
              "69       3378.567152  marker_Colorectal.txt         cae    42            200   \n",
              "70       3378.567152  marker_Colorectal.txt         cae    42            200   \n",
              "71       3378.567152  marker_Colorectal.txt         cae    42            200   \n",
              "72        189.599260  marker_Colorectal.txt         vae    42            200   \n",
              "73        189.599260  marker_Colorectal.txt         vae    42            200   \n",
              "74        189.599260  marker_Colorectal.txt         vae    42            200   \n",
              "75        102.141932  marker_Colorectal.txt         pca    42            200   \n",
              "76        102.141932  marker_Colorectal.txt         pca    42            200   \n",
              "77        102.141932  marker_Colorectal.txt         pca    42            200   \n",
              "78        120.945065  marker_Colorectal.txt          rp    42            200   \n",
              "79        120.945065  marker_Colorectal.txt          rp    42            200   \n",
              "80        120.945065  marker_Colorectal.txt          rp    42            200   \n",
              "81       2080.805042  marker_Colorectal.txt       plain    42            200   \n",
              "82       2080.805042  marker_Colorectal.txt       plain    42            200   \n",
              "83       2080.805042  marker_Colorectal.txt       plain    42            200   \n",
              "84        103.051002         marker_IBD.txt  shallow_ae    42            200   \n",
              "85        103.051002         marker_IBD.txt  shallow_ae    42            200   \n",
              "86        103.051002         marker_IBD.txt  shallow_ae    42            200   \n",
              "87        285.467308         marker_IBD.txt          ae    42            200   \n",
              "88        285.467308         marker_IBD.txt          ae    42            200   \n",
              "89        285.467308         marker_IBD.txt          ae    42            200   \n",
              "90       3058.399564         marker_IBD.txt         cae    42            200   \n",
              "91       3058.399564         marker_IBD.txt         cae    42            200   \n",
              "92       3058.399564         marker_IBD.txt         cae    42            200   \n",
              "93        183.918113         marker_IBD.txt         vae    42            200   \n",
              "94        183.918113         marker_IBD.txt         vae    42            200   \n",
              "95        183.918113         marker_IBD.txt         vae    42            200   \n",
              "96         87.200318         marker_IBD.txt         pca    42            200   \n",
              "97         87.200318         marker_IBD.txt         pca    42            200   \n",
              "98         87.200318         marker_IBD.txt         pca    42            200   \n",
              "99        112.673155         marker_IBD.txt          rp    42            200   \n",
              "100       112.673155         marker_IBD.txt          rp    42            200   \n",
              "101       112.673155         marker_IBD.txt          rp    42            200   \n",
              "102      2497.345249         marker_IBD.txt       plain    42            200   \n",
              "103      2497.345249         marker_IBD.txt       plain    42            200   \n",
              "104      2497.345249         marker_IBD.txt       plain    42            200   \n",
              "105      2495.791879        marker_WT2D.txt         cae    42           2000   \n",
              "106      2495.791879        marker_WT2D.txt         cae    42           2000   \n",
              "107      2495.791879        marker_WT2D.txt         cae    42           2000   \n",
              "108      2730.213370        marker_WT2D.txt       plain    42           2000   \n",
              "109      2730.213370        marker_WT2D.txt       plain    42           2000   \n",
              "110      2730.213370        marker_WT2D.txt       plain    42           2000   \n",
              "111        92.950624        marker_WT2D.txt  shallow_ae    42           2000   \n",
              "112        92.950624        marker_WT2D.txt  shallow_ae    42           2000   \n",
              "113        92.950624        marker_WT2D.txt  shallow_ae    42           2000   \n",
              "114       107.277794        marker_WT2D.txt          ae    42           2000   \n",
              "115       107.277794        marker_WT2D.txt          ae    42           2000   \n",
              "116       107.277794        marker_WT2D.txt          ae    42           2000   \n",
              "117        95.779552        marker_WT2D.txt         vae    42           2000   \n",
              "118        95.779552        marker_WT2D.txt         vae    42           2000   \n",
              "119        95.779552        marker_WT2D.txt         vae    42           2000   \n",
              "120       120.230908        marker_WT2D.txt         pca    42           2000   \n",
              "121       120.230908        marker_WT2D.txt         pca    42           2000   \n",
              "122       120.230908        marker_WT2D.txt         pca    42           2000   \n",
              "123       126.877143        marker_WT2D.txt          rp    42           2000   \n",
              "124       126.877143        marker_WT2D.txt          rp    42           2000   \n",
              "125       126.877143        marker_WT2D.txt          rp    42           2000   \n",
              "\n",
              "    device  \n",
              "0      mps  \n",
              "1      mps  \n",
              "2      mps  \n",
              "3      mps  \n",
              "4      mps  \n",
              "5      mps  \n",
              "6      mps  \n",
              "7      mps  \n",
              "8      mps  \n",
              "9      mps  \n",
              "10     mps  \n",
              "11     mps  \n",
              "12     mps  \n",
              "13     mps  \n",
              "14     mps  \n",
              "15     mps  \n",
              "16     mps  \n",
              "17     mps  \n",
              "18     mps  \n",
              "19     mps  \n",
              "20     mps  \n",
              "21     cpu  \n",
              "22     cpu  \n",
              "23     cpu  \n",
              "24     cpu  \n",
              "25     cpu  \n",
              "26     cpu  \n",
              "27     cpu  \n",
              "28     cpu  \n",
              "29     cpu  \n",
              "30     cpu  \n",
              "31     cpu  \n",
              "32     cpu  \n",
              "33     cpu  \n",
              "34     cpu  \n",
              "35     cpu  \n",
              "36     cpu  \n",
              "37     cpu  \n",
              "38     cpu  \n",
              "39     cpu  \n",
              "40     cpu  \n",
              "41     cpu  \n",
              "42     cpu  \n",
              "43     cpu  \n",
              "44     cpu  \n",
              "45     cpu  \n",
              "46     cpu  \n",
              "47     cpu  \n",
              "48     cpu  \n",
              "49     cpu  \n",
              "50     cpu  \n",
              "51     cpu  \n",
              "52     cpu  \n",
              "53     cpu  \n",
              "54     cpu  \n",
              "55     cpu  \n",
              "56     cpu  \n",
              "57     cpu  \n",
              "58     cpu  \n",
              "59     cpu  \n",
              "60     cpu  \n",
              "61     cpu  \n",
              "62     cpu  \n",
              "63     mps  \n",
              "64     mps  \n",
              "65     mps  \n",
              "66     mps  \n",
              "67     mps  \n",
              "68     mps  \n",
              "69     mps  \n",
              "70     mps  \n",
              "71     mps  \n",
              "72     mps  \n",
              "73     mps  \n",
              "74     mps  \n",
              "75     mps  \n",
              "76     mps  \n",
              "77     mps  \n",
              "78     mps  \n",
              "79     mps  \n",
              "80     mps  \n",
              "81     mps  \n",
              "82     mps  \n",
              "83     mps  \n",
              "84     mps  \n",
              "85     mps  \n",
              "86     mps  \n",
              "87     mps  \n",
              "88     mps  \n",
              "89     mps  \n",
              "90     mps  \n",
              "91     mps  \n",
              "92     mps  \n",
              "93     mps  \n",
              "94     mps  \n",
              "95     mps  \n",
              "96     mps  \n",
              "97     mps  \n",
              "98     mps  \n",
              "99     mps  \n",
              "100    mps  \n",
              "101    mps  \n",
              "102    mps  \n",
              "103    mps  \n",
              "104    mps  \n",
              "105    cpu  \n",
              "106    cpu  \n",
              "107    cpu  \n",
              "108    cpu  \n",
              "109    cpu  \n",
              "110    cpu  \n",
              "111    cpu  \n",
              "112    cpu  \n",
              "113    cpu  \n",
              "114    cpu  \n",
              "115    cpu  \n",
              "116    cpu  \n",
              "117    cpu  \n",
              "118    cpu  \n",
              "119    cpu  \n",
              "120    cpu  \n",
              "121    cpu  \n",
              "122    cpu  \n",
              "123    cpu  \n",
              "124    cpu  \n",
              "125    cpu  "
            ]
          },
          "execution_count": 16,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "url = 'https://raw.githubusercontent.com/ZKingQ/CS598-DLH-SP24/main/result/marker_merged_metrics.csv'\n",
        "df = pd.read_csv(url)\n",
        "# show all rows\n",
        "pd.set_option('display.max_rows', None)\n",
        "df"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "8EAWAy_LwHlV"
      },
      "source": [
        "## Model comparison"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "n_30uovDnQxs"
      },
      "source": [
        "**To-Do:**\n",
        "\n",
        "Train model in full epoches and compare results. (By Apr 28)\n",
        "- we can compare the results of the different models in terms of accuracy, AUC, etc. And,\n",
        "- the time taken to train each model\n",
        "- the best hyperparameters found for each model"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "qH75TNU71eRH"
      },
      "source": [
        "# Discussion\n",
        "\n",
        "In this section,you should discuss your work and make future plan. The discussion should address the following questions:\n",
        "  * Make assessment that the paper is reproducible or not.\n",
        "  * Explain why it is not reproducible if your results are kind negative.\n",
        "  * Describe “What was easy” and “What was difficult” during the reproduction.\n",
        "  * Make suggestions to the author or other reproducers on how to improve the reproducibility.\n",
        "  * What will you do in next phase.\n",
        "\n",
        "---\n",
        "\n",
        "**To-Do:**\n",
        "\n",
        "Add discussion. (By Apr 28)\n",
        "\n",
        "- discuss the reproducibility of the paper and the ease of reproducing the results.\n",
        "- discuss what was easy and what was difficult during the reproduction.\n",
        "- make suggestions to the author or other reproducers on how to improve the reproducibility.\n",
        "- discuss the results of the model comparison and the implications of the results.\n",
        "- discuss the limitations of the study and the potential future work.\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "SHMI2chl9omn"
      },
      "source": [
        "# References\n",
        "\n",
        "1. Cho, I., & Blaser, M. J. (2012). The human microbiome: at the interface of health and disease. Nature Reviews Genetics, 13(4), 260-270.\n",
        "2. Huttenhower, C. et al. Structure, function and diversity of the healthy human microbiome. nature 486, 207 (2012).\n",
        "3. McQuade, J. L., Daniel, C. R., Helmink, B. A. & Wargo, J. A. Modulating the microbiome to improve therapeutic response in cancer. The Lancet Oncology 20, e77–e91 (2019).\n",
        "4. Eloe-Fadrosh, E. A. & Rasko, D. A. The human microbiome: from symbiosis to pathogenesis. Annual review of medicine 64, 145–163 (2013).\n",
        "5. Scholz, M. et al. Strain-level microbial epidemiology and population genomics from shotgun metagenomics. Nature methods 13, 435 (2016).\n",
        "6. Truong, D. T. et al. MetaPhlAn2 for enhanced metagenomic taxonomic profiling. Nature methods 12, 902 (2015).\n",
        "7. Oh, M., & Zhang, L. (2020). DeepMicro: deep representation learning for disease prediction based on microbiome data. Scientific reports, 10(1), 6026.\n",
        "8. Pasolli, E., Truong, D. T., Malik, F., Waldron, L. & Segata, N. Machine learning meta-analysis of large metagenomic datasets: tools and biological insights. PLoS computational biology 12, e1004977 (2016).\n",
        "9. Cawley, G. C. & Talbot, N. L. On over-fitting in model selection and subsequent selection bias in performance evaluation. Journal of Machine Learning Research 11, 2079–2107 (2010).\n",
        "10. Varma, S. & Simon, R. Bias in error estimation when using cross-validation for model selection. BMC bioinformatics 7, 91 (2006).\n"
      ]
    }
  ],
  "metadata": {
    "accelerator": "GPU",
    "colab": {
      "gpuType": "T4",
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3 (ipykernel)",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.6"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
